{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "007494cd",
   "metadata": {},
   "source": [
    "# Командная часть проекта"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "id": "d67a6cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_config = {\n",
    "    'spindalis' : {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/946/903/255/GCA_946903255.1_tdSchSpin1.1b/GCA_946903255.1_tdSchSpin1.1b_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/946/903/255/GCA_946903255.1_tdSchSpin1.1b/GCA_946903255.1_tdSchSpin1.1b_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/946/903/255/GCA_946903255.1_tdSchSpin1.1b/GCA_946903255.1_tdSchSpin1.1b_protein.faa.gz'\n",
    "    },\n",
    "    'japonicum' : {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/021/461/655/GCA_021461655.1_ASM2146165v1/GCA_021461655.1_ASM2146165v1_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/021/461/655/GCA_021461655.1_ASM2146165v1/GCA_021461655.1_ASM2146165v1_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/021/461/655/GCA_021461655.1_ASM2146165v1/GCA_021461655.1_ASM2146165v1_protein.faa.gz'\n",
    "    },\n",
    "    'bovis' : {\n",
    "    },\n",
    "    'mattheei' : {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/405/GCA_944470405.2_tdSchMatt1.2/GCA_944470405.2_tdSchMatt1.2_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/405/GCA_944470405.2_tdSchMatt1.2/GCA_944470405.2_tdSchMatt1.2_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/405/GCA_944470405.2_tdSchMatt1.2/GCA_944470405.2_tdSchMatt1.2_protein.faa.gz'\n",
    "    },\n",
    "    'haematobium' : {\n",
    "        'gff_url' : 'https://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS19/species/schistosoma_haematobium/PRJNA78265/schistosoma_haematobium.PRJNA78265.WBPS19.annotations.gff3.gz',\n",
    "        'fna_url' : 'https://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS19/species/schistosoma_haematobium/PRJNA78265/schistosoma_haematobium.PRJNA78265.WBPS19.genomic.fa.gz',\n",
    "        'proteom_url' : 'https://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS19/species/schistosoma_haematobium/PRJNA78265/schistosoma_haematobium.PRJNA78265.WBPS19.protein.fa.gz'\n",
    "    },\n",
    "    'mansoni' : {\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/237/925/GCF_000237925.1_ASM23792v2/GCF_000237925.1_ASM23792v2_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/237/925/GCF_000237925.1_ASM23792v2/GCF_000237925.1_ASM23792v2_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/237/925/GCF_000237925.1_ASM23792v2/GCF_000237925.1_ASM23792v2_protein.faa.gz'\n",
    "    },\n",
    "    'intercalatum': {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'g4_file' : 'g4_segments.csv',\n",
    "        'zdna_file' : 'zdna_segments.csv',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/385/GCA_944470385.2_tdSchInte2.2/GCA_944470385.2_tdSchInte2.2_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/385/GCA_944470385.2_tdSchInte2.2/GCA_944470385.2_tdSchInte2.2_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/385/GCA_944470385.2_tdSchInte2.2/GCA_944470385.2_tdSchInte2.2_protein.faa.gz'\n",
    "    },\n",
    "    'mekongi' : {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/034/768/735/GCA_034768735.1_ASM3476873v1/GCA_034768735.1_ASM3476873v1_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/034/768/735/GCA_034768735.1_ASM3476873v1/GCA_034768735.1_ASM3476873v1_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/034/768/735/GCA_034768735.1_ASM3476873v1/GCA_034768735.1_ASM3476873v1_protein.faa.gz'\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "id": "886783cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:file data/GCA_946903255.1_tdSchSpin1.1b_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_946903255.1_tdSchSpin1.1b_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_946903255.1_tdSchSpin1.1b_protein.faa.gz already exists\n",
      "INFO:root:file data/GCA_021461655.1_ASM2146165v1_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_021461655.1_ASM2146165v1_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_021461655.1_ASM2146165v1_protein.faa.gz already exists\n",
      "WARNING:root:No gff url provided for bovis\n",
      "WARNING:root:No fna url provided for bovis\n",
      "WARNING:root:No proteom url provided for bovis\n",
      "INFO:root:file data/GCA_944470405.2_tdSchMatt1.2_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_944470405.2_tdSchMatt1.2_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_944470405.2_tdSchMatt1.2_protein.faa.gz already exists\n",
      "INFO:root:file data/schistosoma_haematobium.PRJNA78265.WBPS19.annotations.gff3.gz already exists\n",
      "INFO:root:file data/schistosoma_haematobium.PRJNA78265.WBPS19.genomic.fa.gz already exists\n",
      "INFO:root:file data/schistosoma_haematobium.PRJNA78265.WBPS19.protein.fa.gz already exists\n",
      "INFO:root:file data/GCF_000237925.1_ASM23792v2_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCF_000237925.1_ASM23792v2_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCF_000237925.1_ASM23792v2_protein.faa.gz already exists\n",
      "INFO:root:file data/GCA_944470385.2_tdSchInte2.2_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_944470385.2_tdSchInte2.2_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_944470385.2_tdSchInte2.2_protein.faa.gz already exists\n",
      "INFO:root:file data/GCA_034768735.1_ASM3476873v1_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_034768735.1_ASM3476873v1_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_034768735.1_ASM3476873v1_protein.faa.gz already exists\n",
      "\n",
      "INFO:root:data/GCA_946903255.1_tdSchSpin1.1b_genomic.gff already exists\n",
      "INFO:root:data/GCA_021461655.1_ASM2146165v1_genomic.gff already exists\n",
      "INFO:root:data/GCA_944470405.2_tdSchMatt1.2_genomic.gff already exists\n",
      "INFO:root:data/schistosoma_haematobium.PRJNA78265.WBPS19.annotations.gff3 already exists\n",
      "INFO:root:data/GCF_000237925.1_ASM23792v2_genomic.gff already exists\n",
      "INFO:root:data/GCA_944470385.2_tdSchInte2.2_genomic.gff already exists\n",
      "INFO:root:data/GCA_034768735.1_ASM3476873v1_genomic.gff already exists\n",
      "INFO:root:data/GCA_946903255.1_tdSchSpin1.1b_genomic.fna already exists\n",
      "INFO:root:data/GCA_021461655.1_ASM2146165v1_genomic.fna already exists\n",
      "INFO:root:data/GCA_944470405.2_tdSchMatt1.2_genomic.fna already exists\n",
      "INFO:root:data/schistosoma_haematobium.PRJNA78265.WBPS19.genomic.fa already exists\n",
      "INFO:root:data/GCF_000237925.1_ASM23792v2_genomic.fna already exists\n",
      "INFO:root:data/GCA_944470385.2_tdSchInte2.2_genomic.fna already exists\n",
      "INFO:root:data/GCA_034768735.1_ASM3476873v1_genomic.fna already exists\n",
      "INFO:root:data/GCA_946903255.1_tdSchSpin1.1b_protein.faa already exists\n",
      "INFO:root:data/GCA_021461655.1_ASM2146165v1_protein.faa already exists\n",
      "INFO:root:data/GCA_944470405.2_tdSchMatt1.2_protein.faa already exists\n",
      "INFO:root:data/schistosoma_haematobium.PRJNA78265.WBPS19.protein.fa already exists\n",
      "INFO:root:data/GCF_000237925.1_ASM23792v2_protein.faa already exists\n",
      "INFO:root:data/GCA_944470385.2_tdSchInte2.2_protein.faa already exists\n",
      "INFO:root:data/GCA_034768735.1_ASM3476873v1_protein.faa already exists\n"
     ]
    }
   ],
   "source": [
    "# скачаем аннотации\n",
    "\n",
    "# скачивание файлов генома и аннотации\n",
    "\n",
    "import requests\n",
    "import logging\n",
    "import pathlib\n",
    "import glob\n",
    "import os\n",
    "import sys\n",
    "import itertools\n",
    "\n",
    "logging.basicConfig(level=logging.INFO, force = True)\n",
    "\n",
    "def download_file(url: str, local_file_name: pathlib.Path):\n",
    "    if pathlib.Path(local_file_name).is_file():\n",
    "        logging.info(f'file {local_file_name} already exists')\n",
    "        return local_file_name\n",
    "    logging.info(f'downloading {local_file_name}')\n",
    "    with requests.get(url, stream=True) as r:\n",
    "        r.raise_for_status()\n",
    "        with open(local_file_name, 'wb') as out_file:\n",
    "            for chunk in r.iter_content(chunk_size=8192): \n",
    "                out_file.write(chunk)\n",
    "\n",
    "    return local_file_name\n",
    "\n",
    "gff_files = dict()\n",
    "fna_files = dict()\n",
    "proteom_files = dict()\n",
    "\n",
    "# скачиваем аннотации\n",
    "for current_organism, current_data_config in data_config.items():\n",
    "    if 'gff_url' in current_data_config:\n",
    "        gff_files[current_organism] = download_file(current_data_config['gff_url'], f'data/{current_data_config['gff_url'].split('/')[-1]}')\n",
    "    else:\n",
    "        logging.warning(f'No gff url provided for {current_organism}')\n",
    "    \n",
    "    if 'fna_url' in current_data_config:\n",
    "        fna_files[current_organism] = download_file(current_data_config['fna_url'], f'data/{current_data_config['fna_url'].split('/')[-1]}')\n",
    "    else:\n",
    "        logging.warning(f'No fna url provided for {current_organism}')\n",
    "\n",
    "    if 'proteom_url' in current_data_config:\n",
    "        proteom_files[current_organism] = download_file(current_data_config['proteom_url'], f'data/{current_data_config['proteom_url'].split('/')[-1]}')\n",
    "    else:\n",
    "        logging.warning(f'No proteom url provided for {current_organism}')\n",
    "\n",
    "print(file=sys.stderr)\n",
    "\n",
    "# распакуем сжатые файлы для удобства\n",
    "for organism, filepath in itertools.chain(gff_files.items(), fna_files.items(), proteom_files.items()):\n",
    "    current_file_path = pathlib.Path(filepath)\n",
    "    assert current_file_path.is_file()\n",
    "    if current_file_path.suffix == '.gz':\n",
    "        current_extracted_file = current_file_path.parent / pathlib.Path(filepath).stem\n",
    "        if current_extracted_file.is_file():\n",
    "            logging.info(f'{current_extracted_file} already exists')\n",
    "        else:\n",
    "            logging.info(f'extracting {current_extracted_file}')\n",
    "            os.system(f'gunzip -k {current_extracted_file}')\n",
    "\n",
    "for organism, filepath in gff_files.items():\n",
    "    if pathlib.Path(filepath).suffix == '.gz':\n",
    "        gff_files[organism] = current_file_path.parent / pathlib.Path(filepath).stem\n",
    "\n",
    "for organism, filepath in fna_files.items():\n",
    "    if pathlib.Path(filepath).suffix == '.gz':\n",
    "        fna_files[organism] = current_file_path.parent / pathlib.Path(filepath).stem\n",
    "\n",
    "for organism, filepath in proteom_files.items():\n",
    "    if pathlib.Path(filepath).suffix == '.gz':\n",
    "        proteom_files[organism] = current_file_path.parent / pathlib.Path(filepath).stem\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "id": "d9e06e08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# функции для чтения данных из результатов индивидуальных проектов\n",
    "\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "\n",
    "def get_epigenetic_genes(organism: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "\n",
    "    if 'epigenetic_file' in data_config[organism]:\n",
    "        file_name = data_config[organism]['epigenetic_file']\n",
    "        if pathlib.Path(file_name).suffix == '.csv':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['epigenetic_file']}')\n",
    "            res = res[['Family', 'Gene_ID']]\n",
    "            res.columns = ['family', 'gene']\n",
    "            return res\n",
    "        else:\n",
    "            assert False, \"Please, add support for your file type\" # TODO\n",
    "    else:\n",
    "        logging.warning(f'No epigenetic_file specified for {organism}')\n",
    "        return None\n",
    "\n",
    "def get_g4(organism: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "\n",
    "    if 'g4_file' in data_config[organism]:\n",
    "        file_name = data_config[organism]['g4_file']\n",
    "        if pathlib.Path(file_name).suffix == '.csv':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['g4_file']}')\n",
    "            return res\n",
    "        else:\n",
    "            assert False, \"Please, add support for your file type\" # TODO\n",
    "\n",
    "    else:\n",
    "        logging.warning(f'No epigenetic_file specified for {organism}')\n",
    "        return None\n",
    "\n",
    "def get_zdna(organism: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "\n",
    "    if 'zdna_file' in data_config[organism]:\n",
    "        file_name = data_config[organism]['zdna_file']\n",
    "        if pathlib.Path(file_name).suffix == '.csv':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['zdna_file']}')\n",
    "            return res\n",
    "        else:\n",
    "            assert None, \"Please, add support for your file type\" # TODO\n",
    "\n",
    "    else:\n",
    "        logging.warning(f'No epigenetic_file specified for {organism}')\n",
    "        return None\n",
    "\n",
    "# need_contain one of 'g4' or 'zdna'\n",
    "def get_genes_with_promoters_contains(organism: str, need_contain: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "    assert need_contain in ['g4', 'zdna']\n",
    "\n",
    "    segments = get_g4(organism) if need_contain == 'g4' else get_zdna(organism)\n",
    "    if segments is None:\n",
    "        return None\n",
    "\n",
    "    if organism in gff_files:\n",
    "        res = []\n",
    "        with open(gff_files[organism]) as gff_file:\n",
    "            for row in gff_file:\n",
    "                if row[0] != '#':\n",
    "                    row_splitted = row.strip().split()\n",
    "                    if row_splitted[2] == 'gene':\n",
    "                        gene_coords = (int(row_splitted[3]), int(row_splitted[4]))\n",
    "                        promoter_coords = (gene_coords[0] - 1000, gene_coords[0] - 1)\n",
    "                        if row_splitted[5] == '-':\n",
    "                            promoter_coords = (gene_coords[1] + 1, gene_coords[1] + 1000)\n",
    "\n",
    "                        # поиск квадруплексов или zdna\n",
    "                        found_segments = []\n",
    "                        for _, segment_row in segments.iterrows():\n",
    "                            if segment_row['record_id'] == row_splitted[0]:\n",
    "                                cur_interval = (segment_row['start'], segment_row['end'])\n",
    "                                if pd.Interval(*cur_interval, closed='both').overlaps(pd.Interval(*promoter_coords, closed='both')):\n",
    "                                    found_segments.append(cur_interval)\n",
    "                        \n",
    "                        if len(found_segments) != 0:\n",
    "                            gene_id = row_splitted[8].split('ID=')[1].split(';')[0]\n",
    "                            res.append([row_splitted[0], gene_id, promoter_coords[0], promoter_coords[1], found_segments])\n",
    "        res = pd.DataFrame(res)\n",
    "        res.columns = ['record_id', 'gene', 'start', 'end', 'matches']\n",
    "        return res\n",
    "    else:\n",
    "        logging.warning(f'No gff specified for {organism}')\n",
    "        return None\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "id": "71a4aad9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:local_programs/mega.tar.gz already exists\n",
      "INFO:root:already exists\n"
     ]
    }
   ],
   "source": [
    "# установка меги\n",
    "\n",
    "# https://drive.google.com/file/d/1beRdRYvyRw38orLmFiXE-p2paAILFvVm/view?usp=sharing\n",
    "\n",
    "\n",
    "mega_file_id = \"1beRdRYvyRw38orLmFiXE-p2paAILFvVm\"\n",
    "mega_url = f\"https://drive.google.com/uc?export=download&id={mega_file_id}\"\n",
    "\n",
    "# скачивание меги\n",
    "\n",
    "import requests\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "import logging\n",
    "\n",
    "mega_gz = pathlib.Path('local_programs/mega.tar.gz')\n",
    "mega_dir = pathlib.Path('local_programs/mega')\n",
    "\n",
    "if mega_gz.is_file():\n",
    "    logging.info(f'{mega_gz} already exists')\n",
    "else:\n",
    "    logging.info(f'downloading {mega_gz}')\n",
    "    response = requests.get(mega_url)\n",
    "    assert response.status_code == 200, 'Wrong status code'\n",
    "    with open(mega_gz, 'wb') as out_file:\n",
    "        out_file.write(response.content)\n",
    "\n",
    "if not mega_dir.is_dir():\n",
    "    logging.info(f'extracting {mega_gz}')\n",
    "    ! cd local_programs/ && mkdir -p mega && tar -xzf mega.tar.gz -C mega\n",
    "else:\n",
    "    logging.info(f'already exists')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "594378b3",
   "metadata": {},
   "source": [
    "### 1. От каждого организма в группе Выбрать из генов отвечающих за эпигенетику 4-5 (как минимум 1 на члена группы) и построить для них выравнивания и деревья."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "id": "f6da3ebd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Adding genes for spindalis\n",
      "INFO:root:Adding genes for japonicum\n",
      "WARNING:root:No epigenetic_file specified for bovis\n",
      "WARNING:root:current_epigenetic_genes(bovis) is None\n",
      "INFO:root:Adding genes for mattheei\n",
      "WARNING:root:No epigenetic_file specified for haematobium\n",
      "WARNING:root:current_epigenetic_genes(haematobium) is None\n",
      "WARNING:root:No epigenetic_file specified for mansoni\n",
      "WARNING:root:current_epigenetic_genes(mansoni) is None\n",
      "INFO:root:Adding genes for intercalatum\n",
      "INFO:root:Adding genes for mekongi\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family</th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SSPN1_LOCUS6014</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SMTH1_LOCUS6929</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-KSF78_0008119</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SINT2_LOCUS6163</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-MN116_008246</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-MN116_001510</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SMTH1_LOCUS10615</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SSPN1_LOCUS9217</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-KSF78_0005440</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SINT2_LOCUS8012</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-MN116_002201</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SINT2_LOCUS6554</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SMTH1_LOCUS7396</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-KSF78_0007878</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SSPN1_LOCUS6463</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SSPN1_LOCUS8765</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SINT2_LOCUS8890</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-MN116_002383</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SMTH1_LOCUS10084</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-KSF78_0004211</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-MN116_001389</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-KSF78_0005543</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SMTH1_LOCUS10167</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SINT2_LOCUS5228</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SSPN1_LOCUS9160</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-KSF78_0005343</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SINT2_LOCUS956</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SSPN1_LOCUS3856</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SMTH1_LOCUS4562</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-MN116_003045</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SINT2_LOCUS4989</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-MN116_001474</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SMTH1_LOCUS10358</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-KSF78_0005530</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SSPN1_LOCUS9194</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-MN116_001315</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-SSPN1_LOCUS8531</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-SMTH1_LOCUS447</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-KSF78_0001822</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-SINT2_LOCUS6061</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    family                   gene      organism\n",
       "0    ACINU   gene-SSPN1_LOCUS6014     spindalis\n",
       "1    ACINU   gene-SMTH1_LOCUS6929      mattheei\n",
       "2    ACINU     gene-KSF78_0008119     japonicum\n",
       "3    ACINU   gene-SINT2_LOCUS6163  intercalatum\n",
       "4    ACINU      gene-MN116_008246       mekongi\n",
       "5    AEBP2      gene-MN116_001510       mekongi\n",
       "6    AEBP2  gene-SMTH1_LOCUS10615      mattheei\n",
       "7    AEBP2   gene-SSPN1_LOCUS9217     spindalis\n",
       "8    AEBP2     gene-KSF78_0005440     japonicum\n",
       "9    AEBP2   gene-SINT2_LOCUS8012  intercalatum\n",
       "10  ARID1A      gene-MN116_002201       mekongi\n",
       "11  ARID1A   gene-SINT2_LOCUS6554  intercalatum\n",
       "12  ARID1A   gene-SMTH1_LOCUS7396      mattheei\n",
       "13  ARID1A     gene-KSF78_0007878     japonicum\n",
       "14  ARID1A   gene-SSPN1_LOCUS6463     spindalis\n",
       "15     ATR   gene-SSPN1_LOCUS8765     spindalis\n",
       "16     ATR   gene-SINT2_LOCUS8890  intercalatum\n",
       "17     ATR      gene-MN116_002383       mekongi\n",
       "18     ATR  gene-SMTH1_LOCUS10084      mattheei\n",
       "19     ATR     gene-KSF78_0004211     japonicum\n",
       "20    ATRX      gene-MN116_001389       mekongi\n",
       "21    ATRX     gene-KSF78_0005543     japonicum\n",
       "22    ATRX  gene-SMTH1_LOCUS10167      mattheei\n",
       "23    ATRX   gene-SINT2_LOCUS5228  intercalatum\n",
       "24    ATRX   gene-SSPN1_LOCUS9160     spindalis\n",
       "25    BAP1     gene-KSF78_0005343     japonicum\n",
       "26    BAP1    gene-SINT2_LOCUS956  intercalatum\n",
       "27    BAP1   gene-SSPN1_LOCUS3856     spindalis\n",
       "28    BAP1   gene-SMTH1_LOCUS4562      mattheei\n",
       "29    BAP1      gene-MN116_003045       mekongi\n",
       "30   BARD1   gene-SINT2_LOCUS4989  intercalatum\n",
       "31   BARD1      gene-MN116_001474       mekongi\n",
       "32   BARD1  gene-SMTH1_LOCUS10358      mattheei\n",
       "33   BARD1     gene-KSF78_0005530     japonicum\n",
       "34   BARD1   gene-SSPN1_LOCUS9194     spindalis\n",
       "35    BPTF      gene-MN116_001315       mekongi\n",
       "36    BPTF   gene-SSPN1_LOCUS8531     spindalis\n",
       "37    BPTF    gene-SMTH1_LOCUS447      mattheei\n",
       "38    BPTF     gene-KSF78_0001822     japonicum\n",
       "39    BPTF   gene-SINT2_LOCUS6061  intercalatum"
      ]
     },
     "execution_count": 330,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# выбираем гены для выравнивания\n",
    "\n",
    "import logging\n",
    "\n",
    "family_in_organisms = dict()\n",
    "epigenetic_genes_for_alignment = []\n",
    "\n",
    "for organism in data_config:\n",
    "    current_epigenetic_genes = get_epigenetic_genes(organism)\n",
    "    if current_epigenetic_genes is not None:\n",
    "        logging.info(f'Adding genes for {organism}')\n",
    "        \n",
    "        current_epigenetic_genes['organism'] = organism\n",
    "        epigenetic_genes_for_alignment.append(current_epigenetic_genes)\n",
    "\n",
    "        for _, row in current_epigenetic_genes.iterrows():\n",
    "            if row['family'] not in family_in_organisms:\n",
    "                family_in_organisms[row['family']] = set()\n",
    "            family_in_organisms[row['family']].add(organism)\n",
    "    else:\n",
    "        logging.warning(f'current_epigenetic_genes({organism}) is None')\n",
    "\n",
    "family_in_organisms_array = []\n",
    "for family, organisms in family_in_organisms.items():\n",
    "    family_in_organisms_array.append((family, len(organisms)))\n",
    "\n",
    "family_in_organisms_array = list(sorted(family_in_organisms_array, key=lambda x : -x[1]))\n",
    "family_in_organisms_array = family_in_organisms_array[:8]\n",
    "\n",
    "selected_families = [x[0] for x in family_in_organisms_array]\n",
    "\n",
    "epigenetic_genes_for_alignment = pd.concat(epigenetic_genes_for_alignment)\n",
    "epigenetic_genes_for_alignment = epigenetic_genes_for_alignment[epigenetic_genes_for_alignment['family'].apply(lambda x : x in selected_families)]\n",
    "epigenetic_genes_for_alignment.sort_values('family', inplace=True)\n",
    "epigenetic_genes_for_alignment.drop_duplicates(subset=['family', 'organism'], keep='first', inplace=True)\n",
    "epigenetic_genes_for_alignment.reset_index(inplace=True, drop=True)\n",
    "epigenetic_genes_for_alignment.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "id": "db368604",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No such gff file for bovis\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family</th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "      <th>record_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SSPN1_LOCUS6014</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336419.1</td>\n",
       "      <td>20174797</td>\n",
       "      <td>20192511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SMTH1_LOCUS6929</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103946.1</td>\n",
       "      <td>19000643</td>\n",
       "      <td>19025157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-KSF78_0008119</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038286.1</td>\n",
       "      <td>12019241</td>\n",
       "      <td>12028457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SINT2_LOCUS6163</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103880.1</td>\n",
       "      <td>17878940</td>\n",
       "      <td>17899255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-MN116_008246</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068311.1</td>\n",
       "      <td>9544071</td>\n",
       "      <td>9569391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-MN116_001510</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>41392895</td>\n",
       "      <td>41398756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SMTH1_LOCUS10615</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>85876995</td>\n",
       "      <td>85881366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SSPN1_LOCUS9217</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>86989145</td>\n",
       "      <td>86991956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-KSF78_0005440</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>42441718</td>\n",
       "      <td>42464061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SINT2_LOCUS8012</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>28294149</td>\n",
       "      <td>28315298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-MN116_002201</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>71954612</td>\n",
       "      <td>71997418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SINT2_LOCUS6554</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103881.1</td>\n",
       "      <td>11733954</td>\n",
       "      <td>11780504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SMTH1_LOCUS7396</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103948.1</td>\n",
       "      <td>12246346</td>\n",
       "      <td>12286004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-KSF78_0007878</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038286.1</td>\n",
       "      <td>1806863</td>\n",
       "      <td>1849218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SSPN1_LOCUS6463</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336421.1</td>\n",
       "      <td>16572420</td>\n",
       "      <td>16601429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SSPN1_LOCUS8765</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>69844855</td>\n",
       "      <td>69912424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SINT2_LOCUS8890</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>61543600</td>\n",
       "      <td>61607504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-MN116_002383</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>80531845</td>\n",
       "      <td>80615702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SMTH1_LOCUS10084</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>65509319</td>\n",
       "      <td>65549942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-KSF78_0004211</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038289.1</td>\n",
       "      <td>78055480</td>\n",
       "      <td>78138330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-MN116_001389</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>36996685</td>\n",
       "      <td>37017934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-KSF78_0005543</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>47227496</td>\n",
       "      <td>47266275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SMTH1_LOCUS10167</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>68689821</td>\n",
       "      <td>68736678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SINT2_LOCUS5228</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>25744885</td>\n",
       "      <td>25784263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SSPN1_LOCUS9160</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>84866275</td>\n",
       "      <td>84901726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-KSF78_0005343</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>38289344</td>\n",
       "      <td>38308558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SINT2_LOCUS956</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103876.1</td>\n",
       "      <td>35851198</td>\n",
       "      <td>35858909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SSPN1_LOCUS3856</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336418.1</td>\n",
       "      <td>21839072</td>\n",
       "      <td>21850802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SMTH1_LOCUS4562</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103945.1</td>\n",
       "      <td>26934207</td>\n",
       "      <td>26937693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-MN116_003045</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068306.1</td>\n",
       "      <td>25384891</td>\n",
       "      <td>25413756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SINT2_LOCUS4989</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>15489966</td>\n",
       "      <td>15504708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-MN116_001474</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>40142102</td>\n",
       "      <td>40205167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SMTH1_LOCUS10358</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>76286811</td>\n",
       "      <td>76315711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-KSF78_0005530</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>46803245</td>\n",
       "      <td>46826009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SSPN1_LOCUS9194</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>85921765</td>\n",
       "      <td>85934532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-MN116_001315</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>34110633</td>\n",
       "      <td>34126820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-SSPN1_LOCUS8531</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>59903589</td>\n",
       "      <td>59904911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-SMTH1_LOCUS447</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103947.1</td>\n",
       "      <td>16213366</td>\n",
       "      <td>16243093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-KSF78_0001822</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038282.1</td>\n",
       "      <td>70636542</td>\n",
       "      <td>70670039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-SINT2_LOCUS6061</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103880.1</td>\n",
       "      <td>13976397</td>\n",
       "      <td>14037494</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    family                   gene      organism   record_id     start  \\\n",
       "0    ACINU   gene-SSPN1_LOCUS6014     spindalis  OX336419.1  20174797   \n",
       "1    ACINU   gene-SMTH1_LOCUS6929      mattheei  OX103946.1  19000643   \n",
       "2    ACINU     gene-KSF78_0008119     japonicum  CM038286.1  12019241   \n",
       "3    ACINU   gene-SINT2_LOCUS6163  intercalatum  OX103880.1  17878940   \n",
       "4    ACINU      gene-MN116_008246       mekongi  CM068311.1   9544071   \n",
       "5    AEBP2      gene-MN116_001510       mekongi  CM068305.1  41392895   \n",
       "6    AEBP2  gene-SMTH1_LOCUS10615      mattheei  OX103943.1  85876995   \n",
       "7    AEBP2   gene-SSPN1_LOCUS9217     spindalis  OX336416.1  86989145   \n",
       "8    AEBP2     gene-KSF78_0005440     japonicum  CM038283.1  42441718   \n",
       "9    AEBP2   gene-SINT2_LOCUS8012  intercalatum  OX103883.1  28294149   \n",
       "10  ARID1A      gene-MN116_002201       mekongi  CM068305.1  71954612   \n",
       "11  ARID1A   gene-SINT2_LOCUS6554  intercalatum  OX103881.1  11733954   \n",
       "12  ARID1A   gene-SMTH1_LOCUS7396      mattheei  OX103948.1  12246346   \n",
       "13  ARID1A     gene-KSF78_0007878     japonicum  CM038286.1   1806863   \n",
       "14  ARID1A   gene-SSPN1_LOCUS6463     spindalis  OX336421.1  16572420   \n",
       "15     ATR   gene-SSPN1_LOCUS8765     spindalis  OX336416.1  69844855   \n",
       "16     ATR   gene-SINT2_LOCUS8890  intercalatum  OX103883.1  61543600   \n",
       "17     ATR      gene-MN116_002383       mekongi  CM068305.1  80531845   \n",
       "18     ATR  gene-SMTH1_LOCUS10084      mattheei  OX103943.1  65509319   \n",
       "19     ATR     gene-KSF78_0004211     japonicum  CM038289.1  78055480   \n",
       "20    ATRX      gene-MN116_001389       mekongi  CM068305.1  36996685   \n",
       "21    ATRX     gene-KSF78_0005543     japonicum  CM038283.1  47227496   \n",
       "22    ATRX  gene-SMTH1_LOCUS10167      mattheei  OX103943.1  68689821   \n",
       "23    ATRX   gene-SINT2_LOCUS5228  intercalatum  OX103879.1  25744885   \n",
       "24    ATRX   gene-SSPN1_LOCUS9160     spindalis  OX336416.1  84866275   \n",
       "25    BAP1     gene-KSF78_0005343     japonicum  CM038283.1  38289344   \n",
       "26    BAP1    gene-SINT2_LOCUS956  intercalatum  OX103876.1  35851198   \n",
       "27    BAP1   gene-SSPN1_LOCUS3856     spindalis  OX336418.1  21839072   \n",
       "28    BAP1   gene-SMTH1_LOCUS4562      mattheei  OX103945.1  26934207   \n",
       "29    BAP1      gene-MN116_003045       mekongi  CM068306.1  25384891   \n",
       "30   BARD1   gene-SINT2_LOCUS4989  intercalatum  OX103879.1  15489966   \n",
       "31   BARD1      gene-MN116_001474       mekongi  CM068305.1  40142102   \n",
       "32   BARD1  gene-SMTH1_LOCUS10358      mattheei  OX103943.1  76286811   \n",
       "33   BARD1     gene-KSF78_0005530     japonicum  CM038283.1  46803245   \n",
       "34   BARD1   gene-SSPN1_LOCUS9194     spindalis  OX336416.1  85921765   \n",
       "35    BPTF      gene-MN116_001315       mekongi  CM068305.1  34110633   \n",
       "36    BPTF   gene-SSPN1_LOCUS8531     spindalis  OX336416.1  59903589   \n",
       "37    BPTF    gene-SMTH1_LOCUS447      mattheei  OX103947.1  16213366   \n",
       "38    BPTF     gene-KSF78_0001822     japonicum  CM038282.1  70636542   \n",
       "39    BPTF   gene-SINT2_LOCUS6061  intercalatum  OX103880.1  13976397   \n",
       "\n",
       "         end  \n",
       "0   20192511  \n",
       "1   19025157  \n",
       "2   12028457  \n",
       "3   17899255  \n",
       "4    9569391  \n",
       "5   41398756  \n",
       "6   85881366  \n",
       "7   86991956  \n",
       "8   42464061  \n",
       "9   28315298  \n",
       "10  71997418  \n",
       "11  11780504  \n",
       "12  12286004  \n",
       "13   1849218  \n",
       "14  16601429  \n",
       "15  69912424  \n",
       "16  61607504  \n",
       "17  80615702  \n",
       "18  65549942  \n",
       "19  78138330  \n",
       "20  37017934  \n",
       "21  47266275  \n",
       "22  68736678  \n",
       "23  25784263  \n",
       "24  84901726  \n",
       "25  38308558  \n",
       "26  35858909  \n",
       "27  21850802  \n",
       "28  26937693  \n",
       "29  25413756  \n",
       "30  15504708  \n",
       "31  40205167  \n",
       "32  76315711  \n",
       "33  46826009  \n",
       "34  85934532  \n",
       "35  34126820  \n",
       "36  59904911  \n",
       "37  16243093  \n",
       "38  70670039  \n",
       "39  14037494  "
      ]
     },
     "execution_count": 331,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Добавим в табличку хромосому координаты каждого\n",
    "\n",
    "import logging\n",
    "\n",
    "gene_organism_to_position = dict()\n",
    "\n",
    "for _, row in epigenetic_genes_for_alignment.iterrows():\n",
    "    gene_organism_to_position[(row['organism'], row['gene'])] = None\n",
    "\n",
    "for organism in data_config:\n",
    "    if organism in gff_files:\n",
    "        with open(gff_files[organism], 'r') as gff_file:\n",
    "            for row in gff_file:\n",
    "                current_row = row.strip()\n",
    "                if len(current_row) == 0 or current_row[0] == '#':\n",
    "                    continue\n",
    "                current_row_splitted = row.split()\n",
    "\n",
    "                if current_row_splitted[2] == 'gene':\n",
    "                    current_gene_id = current_row_splitted[8].split('ID=')[1].split(';')[0]\n",
    "                    if (organism, current_gene_id) in gene_organism_to_position:\n",
    "                        gene_organism_to_position[(organism, current_gene_id)] = (current_row_splitted[0], int(current_row_splitted[3]), int(current_row_splitted[4]))\n",
    "    else:\n",
    "        logging.warning(f'No such gff file for {organism}')\n",
    "\n",
    "def get_position(row):\n",
    "    res = gene_organism_to_position[(row['organism'], row['gene'])]\n",
    "    return res\n",
    "\n",
    "epigenetic_genes_for_alignment[['record_id', 'start', 'end']] = \\\n",
    "    epigenetic_genes_for_alignment.apply(get_position, axis=1, result_type='expand')\n",
    "\n",
    "epigenetic_genes_for_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "id": "6555a129",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No such fna file for bovis\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family</th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "      <th>record_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SSPN1_LOCUS6014</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336419.1</td>\n",
       "      <td>20174797</td>\n",
       "      <td>20192511</td>\n",
       "      <td>atgtctactccggagggtgttctgttgccagatggccggagaattg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SMTH1_LOCUS6929</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103946.1</td>\n",
       "      <td>19000643</td>\n",
       "      <td>19025157</td>\n",
       "      <td>tcaacgtcgtgatcgataacgatttgatgaggaacgatctctacgt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-KSF78_0008119</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038286.1</td>\n",
       "      <td>12019241</td>\n",
       "      <td>12028457</td>\n",
       "      <td>gcatttgaatcaagtcaccgcggatagtttaagaatagcttattta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SINT2_LOCUS6163</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103880.1</td>\n",
       "      <td>17878940</td>\n",
       "      <td>17899255</td>\n",
       "      <td>ttaacgtcgtgatcgataacgatttgatgaggaacgatctctacgt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-MN116_008246</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068311.1</td>\n",
       "      <td>9544071</td>\n",
       "      <td>9569391</td>\n",
       "      <td>atgtctactccagaggacgttattttacctgatggtcgtagaattg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-MN116_001510</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>41392895</td>\n",
       "      <td>41398756</td>\n",
       "      <td>atgaatcgcacaagttctccgttggatgaagcatatgggacataca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SMTH1_LOCUS10615</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>85876995</td>\n",
       "      <td>85881366</td>\n",
       "      <td>gttcaccatttctagtcatgtttgggagatgaaaacctgtttatct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SSPN1_LOCUS9217</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>86989145</td>\n",
       "      <td>86991956</td>\n",
       "      <td>atgagtatattattcaatccatcattttttgatatacacttctttt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-KSF78_0005440</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>42441718</td>\n",
       "      <td>42464061</td>\n",
       "      <td>aatttatttggctgattcatcggattgtgtccttctattacgatct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SINT2_LOCUS8012</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>28294149</td>\n",
       "      <td>28315298</td>\n",
       "      <td>atgtcttctgggatagaggtgattgagtctggtttccatgttactc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-MN116_002201</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>71954612</td>\n",
       "      <td>71997418</td>\n",
       "      <td>atgaatatggtaactttcttttgtcttggtaaagttgttcgagaaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SINT2_LOCUS6554</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103881.1</td>\n",
       "      <td>11733954</td>\n",
       "      <td>11780504</td>\n",
       "      <td>gagagagagagagagagagagattaactcttgctacatatgaacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SMTH1_LOCUS7396</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103948.1</td>\n",
       "      <td>12246346</td>\n",
       "      <td>12286004</td>\n",
       "      <td>ctaatgcatgagcttttcttctacaaccttccatgtatccttttca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-KSF78_0007878</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038286.1</td>\n",
       "      <td>1806863</td>\n",
       "      <td>1849218</td>\n",
       "      <td>cgaaacaacgattcatttattaaccctgtaaaatatttatatcctc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SSPN1_LOCUS6463</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336421.1</td>\n",
       "      <td>16572420</td>\n",
       "      <td>16601429</td>\n",
       "      <td>agaagaagaagaagaagaagaagaagaagaagaagaagaatgcatg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SSPN1_LOCUS8765</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>69844855</td>\n",
       "      <td>69912424</td>\n",
       "      <td>tcaatcaaataatttctgaagttgatcgaatgccaattaataacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SINT2_LOCUS8890</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>61543600</td>\n",
       "      <td>61607504</td>\n",
       "      <td>ttataagaatggcatccatcctttatacatttgacagagtcggtca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-MN116_002383</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>80531845</td>\n",
       "      <td>80615702</td>\n",
       "      <td>ttacaaccaaggcaataaatttggattcatatgaattaaattggca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SMTH1_LOCUS10084</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>65509319</td>\n",
       "      <td>65549942</td>\n",
       "      <td>ctatgttcgtaggcagcagcagccatatctagctgtccttcgtttt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-KSF78_0004211</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038289.1</td>\n",
       "      <td>78055480</td>\n",
       "      <td>78138330</td>\n",
       "      <td>ccgattagcgtttcttcaagcatgttgttttttacgggatggggtt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-MN116_001389</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>36996685</td>\n",
       "      <td>37017934</td>\n",
       "      <td>ctaaagatataggcttgcagagaagataatatctcttttgatgagg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-KSF78_0005543</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>47227496</td>\n",
       "      <td>47266275</td>\n",
       "      <td>ttaaaattgcggaaaatgaaaatttggaaaactttgtgaatgtaaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SMTH1_LOCUS10167</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>68689821</td>\n",
       "      <td>68736678</td>\n",
       "      <td>atggaacatcagtaccgatcattcggtgtcgagtcatctagcttcc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SINT2_LOCUS5228</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>25744885</td>\n",
       "      <td>25784263</td>\n",
       "      <td>cagtacacttcgctgtaatatctggtaatggtcatttcagtaaaga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SSPN1_LOCUS9160</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>84866275</td>\n",
       "      <td>84901726</td>\n",
       "      <td>ttaaagatataggctcgcacagaaaatgatatctcttttgattagg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-KSF78_0005343</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>38289344</td>\n",
       "      <td>38308558</td>\n",
       "      <td>ctcagtttatttgaaattatcaatgttaaccttggtaatgatatat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SINT2_LOCUS956</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103876.1</td>\n",
       "      <td>35851198</td>\n",
       "      <td>35858909</td>\n",
       "      <td>cgctagatatattcatgtttacacttggctcatctcatcaatttcc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SSPN1_LOCUS3856</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336418.1</td>\n",
       "      <td>21839072</td>\n",
       "      <td>21850802</td>\n",
       "      <td>agtttatataggctagcctcctgttgtcgaaaattttggatattgc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SMTH1_LOCUS4562</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103945.1</td>\n",
       "      <td>26934207</td>\n",
       "      <td>26937693</td>\n",
       "      <td>tcagttattatcagaacataatcccatcaaactgaattctacactt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-MN116_003045</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068306.1</td>\n",
       "      <td>25384891</td>\n",
       "      <td>25413756</td>\n",
       "      <td>tcaaagacgccgtttccgaaatcgttgagaattggatgaagtagac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SINT2_LOCUS4989</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>15489966</td>\n",
       "      <td>15504708</td>\n",
       "      <td>taaagatctcaaaaattatattgaaattgactgaagcaaattaaaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-MN116_001474</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>40142102</td>\n",
       "      <td>40205167</td>\n",
       "      <td>ttacagtattgtgcatttccgacaacattcttgacgaaataaatta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SMTH1_LOCUS10358</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>76286811</td>\n",
       "      <td>76315711</td>\n",
       "      <td>gggttaaatagccaacaagtagctgcacaaaggatcaaacccaaaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-KSF78_0005530</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>46803245</td>\n",
       "      <td>46826009</td>\n",
       "      <td>cttcttcaggatgtgtttctttcatttttcaggcataggtagagac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SSPN1_LOCUS9194</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>85921765</td>\n",
       "      <td>85934532</td>\n",
       "      <td>atgaaggttgggtcatgtttttgctgcaatacagtctgcctttatt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-MN116_001315</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>34110633</td>\n",
       "      <td>34126820</td>\n",
       "      <td>ctagaagtcttcatcatcagggtcatcatcggctacatcattgtca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-SSPN1_LOCUS8531</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>59903589</td>\n",
       "      <td>59904911</td>\n",
       "      <td>ttgaccttcaagtgctcgcaacccccatcagtggacgatacctctg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-SMTH1_LOCUS447</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103947.1</td>\n",
       "      <td>16213366</td>\n",
       "      <td>16243093</td>\n",
       "      <td>ccgcatatttcaattaacataaaatcatcatcattaccgtttgcgt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-KSF78_0001822</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038282.1</td>\n",
       "      <td>70636542</td>\n",
       "      <td>70670039</td>\n",
       "      <td>gctctttattctctgcctctaccttttttttaataaatgtaactgt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>gene-SINT2_LOCUS6061</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103880.1</td>\n",
       "      <td>13976397</td>\n",
       "      <td>14037494</td>\n",
       "      <td>ttataaattctcttctttgaactgtttcattttctgaacaaataat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    family                   gene      organism   record_id     start  \\\n",
       "0    ACINU   gene-SSPN1_LOCUS6014     spindalis  OX336419.1  20174797   \n",
       "1    ACINU   gene-SMTH1_LOCUS6929      mattheei  OX103946.1  19000643   \n",
       "2    ACINU     gene-KSF78_0008119     japonicum  CM038286.1  12019241   \n",
       "3    ACINU   gene-SINT2_LOCUS6163  intercalatum  OX103880.1  17878940   \n",
       "4    ACINU      gene-MN116_008246       mekongi  CM068311.1   9544071   \n",
       "5    AEBP2      gene-MN116_001510       mekongi  CM068305.1  41392895   \n",
       "6    AEBP2  gene-SMTH1_LOCUS10615      mattheei  OX103943.1  85876995   \n",
       "7    AEBP2   gene-SSPN1_LOCUS9217     spindalis  OX336416.1  86989145   \n",
       "8    AEBP2     gene-KSF78_0005440     japonicum  CM038283.1  42441718   \n",
       "9    AEBP2   gene-SINT2_LOCUS8012  intercalatum  OX103883.1  28294149   \n",
       "10  ARID1A      gene-MN116_002201       mekongi  CM068305.1  71954612   \n",
       "11  ARID1A   gene-SINT2_LOCUS6554  intercalatum  OX103881.1  11733954   \n",
       "12  ARID1A   gene-SMTH1_LOCUS7396      mattheei  OX103948.1  12246346   \n",
       "13  ARID1A     gene-KSF78_0007878     japonicum  CM038286.1   1806863   \n",
       "14  ARID1A   gene-SSPN1_LOCUS6463     spindalis  OX336421.1  16572420   \n",
       "15     ATR   gene-SSPN1_LOCUS8765     spindalis  OX336416.1  69844855   \n",
       "16     ATR   gene-SINT2_LOCUS8890  intercalatum  OX103883.1  61543600   \n",
       "17     ATR      gene-MN116_002383       mekongi  CM068305.1  80531845   \n",
       "18     ATR  gene-SMTH1_LOCUS10084      mattheei  OX103943.1  65509319   \n",
       "19     ATR     gene-KSF78_0004211     japonicum  CM038289.1  78055480   \n",
       "20    ATRX      gene-MN116_001389       mekongi  CM068305.1  36996685   \n",
       "21    ATRX     gene-KSF78_0005543     japonicum  CM038283.1  47227496   \n",
       "22    ATRX  gene-SMTH1_LOCUS10167      mattheei  OX103943.1  68689821   \n",
       "23    ATRX   gene-SINT2_LOCUS5228  intercalatum  OX103879.1  25744885   \n",
       "24    ATRX   gene-SSPN1_LOCUS9160     spindalis  OX336416.1  84866275   \n",
       "25    BAP1     gene-KSF78_0005343     japonicum  CM038283.1  38289344   \n",
       "26    BAP1    gene-SINT2_LOCUS956  intercalatum  OX103876.1  35851198   \n",
       "27    BAP1   gene-SSPN1_LOCUS3856     spindalis  OX336418.1  21839072   \n",
       "28    BAP1   gene-SMTH1_LOCUS4562      mattheei  OX103945.1  26934207   \n",
       "29    BAP1      gene-MN116_003045       mekongi  CM068306.1  25384891   \n",
       "30   BARD1   gene-SINT2_LOCUS4989  intercalatum  OX103879.1  15489966   \n",
       "31   BARD1      gene-MN116_001474       mekongi  CM068305.1  40142102   \n",
       "32   BARD1  gene-SMTH1_LOCUS10358      mattheei  OX103943.1  76286811   \n",
       "33   BARD1     gene-KSF78_0005530     japonicum  CM038283.1  46803245   \n",
       "34   BARD1   gene-SSPN1_LOCUS9194     spindalis  OX336416.1  85921765   \n",
       "35    BPTF      gene-MN116_001315       mekongi  CM068305.1  34110633   \n",
       "36    BPTF   gene-SSPN1_LOCUS8531     spindalis  OX336416.1  59903589   \n",
       "37    BPTF    gene-SMTH1_LOCUS447      mattheei  OX103947.1  16213366   \n",
       "38    BPTF     gene-KSF78_0001822     japonicum  CM038282.1  70636542   \n",
       "39    BPTF   gene-SINT2_LOCUS6061  intercalatum  OX103880.1  13976397   \n",
       "\n",
       "         end                                               data  \n",
       "0   20192511  atgtctactccggagggtgttctgttgccagatggccggagaattg...  \n",
       "1   19025157  tcaacgtcgtgatcgataacgatttgatgaggaacgatctctacgt...  \n",
       "2   12028457  gcatttgaatcaagtcaccgcggatagtttaagaatagcttattta...  \n",
       "3   17899255  ttaacgtcgtgatcgataacgatttgatgaggaacgatctctacgt...  \n",
       "4    9569391  atgtctactccagaggacgttattttacctgatggtcgtagaattg...  \n",
       "5   41398756  atgaatcgcacaagttctccgttggatgaagcatatgggacataca...  \n",
       "6   85881366  gttcaccatttctagtcatgtttgggagatgaaaacctgtttatct...  \n",
       "7   86991956  atgagtatattattcaatccatcattttttgatatacacttctttt...  \n",
       "8   42464061  aatttatttggctgattcatcggattgtgtccttctattacgatct...  \n",
       "9   28315298  atgtcttctgggatagaggtgattgagtctggtttccatgttactc...  \n",
       "10  71997418  atgaatatggtaactttcttttgtcttggtaaagttgttcgagaaa...  \n",
       "11  11780504  gagagagagagagagagagagattaactcttgctacatatgaacat...  \n",
       "12  12286004  ctaatgcatgagcttttcttctacaaccttccatgtatccttttca...  \n",
       "13   1849218  cgaaacaacgattcatttattaaccctgtaaaatatttatatcctc...  \n",
       "14  16601429  agaagaagaagaagaagaagaagaagaagaagaagaagaatgcatg...  \n",
       "15  69912424  tcaatcaaataatttctgaagttgatcgaatgccaattaataacat...  \n",
       "16  61607504  ttataagaatggcatccatcctttatacatttgacagagtcggtca...  \n",
       "17  80615702  ttacaaccaaggcaataaatttggattcatatgaattaaattggca...  \n",
       "18  65549942  ctatgttcgtaggcagcagcagccatatctagctgtccttcgtttt...  \n",
       "19  78138330  ccgattagcgtttcttcaagcatgttgttttttacgggatggggtt...  \n",
       "20  37017934  ctaaagatataggcttgcagagaagataatatctcttttgatgagg...  \n",
       "21  47266275  ttaaaattgcggaaaatgaaaatttggaaaactttgtgaatgtaaa...  \n",
       "22  68736678  atggaacatcagtaccgatcattcggtgtcgagtcatctagcttcc...  \n",
       "23  25784263  cagtacacttcgctgtaatatctggtaatggtcatttcagtaaaga...  \n",
       "24  84901726  ttaaagatataggctcgcacagaaaatgatatctcttttgattagg...  \n",
       "25  38308558  ctcagtttatttgaaattatcaatgttaaccttggtaatgatatat...  \n",
       "26  35858909  cgctagatatattcatgtttacacttggctcatctcatcaatttcc...  \n",
       "27  21850802  agtttatataggctagcctcctgttgtcgaaaattttggatattgc...  \n",
       "28  26937693  tcagttattatcagaacataatcccatcaaactgaattctacactt...  \n",
       "29  25413756  tcaaagacgccgtttccgaaatcgttgagaattggatgaagtagac...  \n",
       "30  15504708  taaagatctcaaaaattatattgaaattgactgaagcaaattaaaa...  \n",
       "31  40205167  ttacagtattgtgcatttccgacaacattcttgacgaaataaatta...  \n",
       "32  76315711  gggttaaatagccaacaagtagctgcacaaaggatcaaacccaaaa...  \n",
       "33  46826009  cttcttcaggatgtgtttctttcatttttcaggcataggtagagac...  \n",
       "34  85934532  atgaaggttgggtcatgtttttgctgcaatacagtctgcctttatt...  \n",
       "35  34126820  ctagaagtcttcatcatcagggtcatcatcggctacatcattgtca...  \n",
       "36  59904911  ttgaccttcaagtgctcgcaacccccatcagtggacgatacctctg...  \n",
       "37  16243093  ccgcatatttcaattaacataaaatcatcatcattaccgtttgcgt...  \n",
       "38  70670039  gctctttattctctgcctctaccttttttttaataaatgtaactgt...  \n",
       "39  14037494  ttataaattctcttctttgaactgtttcattttctgaacaaataat...  "
      ]
     },
     "execution_count": 332,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Добавим в табличку последовательность гена\n",
    "\n",
    "import pandas as pd\n",
    "import logging\n",
    "from Bio import SeqIO\n",
    "\n",
    "record_to_position = dict()\n",
    "\n",
    "for _, row in epigenetic_genes_for_alignment.iterrows():\n",
    "    if row['record_id'] not in record_to_position:\n",
    "        record_to_position[row['record_id']] = dict()\n",
    "    record_to_position[row['record_id']][(row['start'], row['end'])] = None\n",
    "\n",
    "record_to_position\n",
    "\n",
    "for organism in data_config:\n",
    "    if organism in fna_files:\n",
    "        records = list(SeqIO.parse(fna_files[organism], \"fasta\"))\n",
    "        for record in records:\n",
    "            if record.id in record_to_position:\n",
    "                for requested_range in record_to_position[record.id]:\n",
    "                    record_to_position[record.id][requested_range] = record.seq[requested_range[0]-1:requested_range[1]]\n",
    "\n",
    "    else:\n",
    "        logging.warning(f'No such fna file for {organism}')\n",
    "\n",
    "def get_sequence(row):\n",
    "    return [record_to_position[row['record_id']][(row['start'], row['end'])]]\n",
    "        \n",
    "epigenetic_genes_for_alignment[['data']] = \\\n",
    "    epigenetic_genes_for_alignment.apply(get_sequence, axis=1, result_type='expand')\n",
    "\n",
    "epigenetic_genes_for_alignment['data'] = epigenetic_genes_for_alignment['data'].apply(lambda s : str(s).lower())\n",
    "epigenetic_genes_for_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "id": "0609546e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Running allignment for ACINU\n",
      "INFO:root:Allignment for ACINU done\n",
      "INFO:root:Running allignment for AEBP2\n",
      "INFO:root:Allignment for AEBP2 done\n",
      "INFO:root:Running allignment for ARID1A\n",
      "INFO:root:Allignment for ARID1A done\n",
      "INFO:root:Running allignment for ATR\n",
      "INFO:root:Allignment for ATR done\n",
      "INFO:root:Running allignment for ATRX\n",
      "INFO:root:Allignment for ATRX done\n",
      "INFO:root:Running allignment for BAP1\n",
      "INFO:root:Allignment for BAP1 done\n",
      "INFO:root:Running allignment for BARD1\n",
      "INFO:root:Allignment for BARD1 done\n",
      "INFO:root:Running allignment for BPTF\n",
      "INFO:root:Allignment for BPTF done\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import logging\n",
    "\n",
    "# настройки меги для выравнивания\n",
    "with open(\"data/muscle_align.mao\", \"w\") as f:\n",
    "    f.write(\"\"\"; Please do not edit this file! If this file is modified, results are unpredictable.\n",
    "; Instead of modifying this file, simply create a new MEGA Analysis Options file by using the MEGA Prototyper.\n",
    "[ DataSettings ]\n",
    "datatype                 = snNucleotide            \n",
    "containsCodingNuc        = False                   \n",
    "missingBaseSymbol        = ?                       \n",
    "identicalBaseSymbol      = .                       \n",
    "gapSymbol                = -                       \n",
    "[ ProcessTypes ]\n",
    "ppAlign                  = true                    \n",
    "ppClustalW               = true                    \n",
    "[ AnalysisSettings ]\n",
    "Pairwise Alignment       = ====================    \n",
    "DNAPWGapOpeningPenalty   = 15.00                   \n",
    "DNAPWGapExtensionPenalty = 6.66                    \n",
    "Multiple Alignment       = ====================    \n",
    "DNAMAGapOpeningPenalty   = 15.00                   \n",
    "DNAMAGapExtensionPenalty = 6.66                    \n",
    "Global Options           = ====================    \n",
    "DNA Weight Matrix        = IUB                     \n",
    "TransitionWeightNEdit    = 0.50                    \n",
    "UseNegativeMatrix        = ON                      \n",
    "DelayDivergentCutoff     = 30                      \n",
    "KeepPredefinedGaps       = True\"\"\")\n",
    "\n",
    "epigenetec_alignment_results = []\n",
    "\n",
    "for family, subdf in epigenetic_genes_for_alignment.groupby('family'):\n",
    "    # запишем последовательности в fasta файл\n",
    "    current_fasta_file = f'data/epigenetic_genes_for_alignment_{family}.fasta'\n",
    "    current_meg_file = f'data/epigenetic_genes_alignment_{family}.meg'\n",
    "    with open(current_fasta_file, 'w') as fasta_file:\n",
    "        for _, row in subdf.iterrows():\n",
    "            seq_id, seq = f'{row['organism']}:{row['gene']}', row['data']\n",
    "            fasta_file.write(f\">{seq_id}\\n{seq}\\n\")\n",
    "    # запуск меги\n",
    "    logging.info(f'Running allignment for {family}')\n",
    "    os.system(f'./local_programs/mega/megacc -a data/muscle_align.mao -d {current_fasta_file} -o {current_meg_file} > /dev/null 2>&1')\n",
    "    epigenetec_alignment_results.append((family, current_meg_file))\n",
    "    logging.info(f'Allignment for {family} done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df453ec4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Running tree building for ACINU\n",
      "INFO:root:Tree building for ACINU done\n",
      "INFO:root:Running tree building for AEBP2\n",
      "INFO:root:Tree building for AEBP2 done\n",
      "INFO:root:Running tree building for ARID1A\n",
      "INFO:root:Tree building for ARID1A done\n",
      "INFO:root:Running tree building for ATR\n",
      "INFO:root:Tree building for ATR done\n",
      "INFO:root:Running tree building for ATRX\n",
      "INFO:root:Tree building for ATRX done\n",
      "INFO:root:Running tree building for BAP1\n",
      "INFO:root:Tree building for BAP1 done\n",
      "INFO:root:Running tree building for BARD1\n",
      "INFO:root:Tree building for BARD1 done\n",
      "INFO:root:Running tree building for BPTF\n",
      "INFO:root:Tree building for BPTF done\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# настройки меги для построения дерева\n",
    "with open(\"data/nj_tree.mao\", \"w\") as f:\n",
    "    f.write(\"\"\"[ DataSettings ]\n",
    "datatype                             = snNucleotide                        \n",
    "containsCodingNuc                    = True                                \n",
    "MissingBaseSymbol                    = ?                                   \n",
    "IdenticalBaseSymbol                  = .                                   \n",
    "GapSymbol                            = -                                   \n",
    "[ ProcessTypes ]\n",
    "ppInfer                              = true                                \n",
    "ppML                                 = true                                \n",
    "[ AnalysisSettings ]\n",
    "Analysis                             = Phylogeny Reconstruction            \n",
    "Statistical Method                   = Maximum Likelihood                  \n",
    "Phylogeny Test                       = ====================                \n",
    "Test of Phylogeny                    = Bootstrap method                    \n",
    "No. of Bootstrap Replications        = 500                                 \n",
    "Substitution Model                   = ====================                \n",
    "Substitutions Type                   = Nucleotide                          \n",
    "Model/Method                         = Kimura 2-parameter model            \n",
    "Rates and Patterns                   = ====================                \n",
    "Rates among Sites                    = Uniform Rates                       \n",
    "No of Discrete Gamma Categories      = Not Applicable                      \n",
    "Data Subset to Use                   = ====================                \n",
    "Gaps/Missing Data Treatment          = Complete deletion                   \n",
    "Site Coverage Cutoff (%)             = Not Applicable                      \n",
    "Select Codon Positions               = 1st, 2nd, 3rd, Non-Coding           \n",
    "Tree Inference Options               = ====================                \n",
    "ML Heuristic Method                  = Nearest-Neighbor-Interchange (NNI)  \n",
    "Initial Tree for ML                  = Make initial tree automatically (Neighbor Joining)\n",
    "Branch Swap Filter                   = Strong                              \n",
    "System Resource Usage                = ====================                \n",
    "Number of Threads                    = 1                                   \n",
    "Genetic Code Table                   = Not Applicable                      \n",
    "Genetic Code                         = Not Applicable                      \n",
    "Has Time Limit                       = False                               \n",
    "Maximum Execution Time               = -1  \"\"\")\n",
    "\n",
    "epigenetec_trees_results = []\n",
    "\n",
    "# запуск меги\n",
    "for family, current_alignment_file in epigenetec_alignment_results:\n",
    "    current_nwk_file = f'data/epigenetic_genes_tree_{family}.nwk'\n",
    "    logging.info(f'Running tree building for {family}')\n",
    "    os.system(f'./local_programs/mega/megacc -a data/nj_tree.mao -d {current_alignment_file} -o {current_nwk_file} > /dev/null 2>&1')\n",
    "    epigenetec_trees_results.append((family, current_nwk_file))\n",
    "    logging.info(f'Tree building for {family} done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "id": "1b82213b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                 FigTree v1.4.4, 2006-2018\n",
      "                  Tree Figure Drawing Tool\n",
      "                       Andrew Rambaut\n",
      "\n",
      "             Institute of Evolutionary Biology\n",
      "                  University of Edinburgh\n",
      "                     a.rambaut@ed.ac.uk\n",
      "\n",
      "                 http://tree.bio.ed.ac.uk/\n",
      "    Uses the Java Evolutionary Biology 2 Library (JEBL2)\n",
      "                http://jebl2.googlecode.com/\n",
      "                 Uses the iText PDF Library\n",
      "                    http://itextpdf.com/\n",
      "               Uses the Apache Batik Library\n",
      "            http://xmlgraphics.apache.org/batik/\n",
      "                 Uses the JDOM XML Library\n",
      "                    http://www.jdom.org/\n",
      " Thanks to Alexei Drummond, Joseph Heled, Philippe Lemey, \n",
      "Tulio de Oliveira, Oliver Pybus, Beth Shapiro & Marc Suchard\n",
      "\n",
      "Creating SVG graphic: data/epigenetic_genes_tree_ACINU.svg\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'ACINU:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" style=\"fill-opacity:1; color-rendering:auto; color-interpolation:auto; text-rendering:auto; stroke:black; stroke-linecap:square; stroke-miterlimit:10; shape-rendering:auto; stroke-opacity:1; fill:black; stroke-dasharray:none; font-weight:normal; stroke-width:1; font-family:'Dialog'; font-style:normal; stroke-linejoin:miter; font-size:12; stroke-dashoffset:0; image-rendering:auto;\" width=\"800.0\" viewBox=\"0 0 800.0 600.0\" height=\"600.0\"><!--Generated by the Batik Graphics2D SVG Generator--><defs id=\"genericDefs\"/><g><g transform=\"translate(6,6)\" style=\"text-rendering:optimizeLegibility; font-size:10; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><line y2=\"570\" style=\"fill:none;\" x1=\"276.1492\" x2=\"365.363\" y1=\"570\"/><text x=\"313.8061\" xml:space=\"preserve\" y=\"583\" style=\"stroke:none;\">0.2</text><path style=\"fill:none;\" d=\"M641.5122 141.25 L632.8729 141.25 L632.8729 209.375\"/><path style=\"fill:none;\" d=\"M595.1341 413.75 L377.0738 413.75 L377.0738 311.5625\"/><path style=\"fill:none;\" d=\"M24.8464 550 L6.3516 550 L6.3516 288.8542\"/><path style=\"fill:none;\" d=\"M632.8729 209.375 L377.0738 209.375 L377.0738 311.5625\"/><path style=\"fill:none;\" d=\"M377.0738 311.5625 L6.3516 311.5625 L6.3516 288.8542\"/><path style=\"fill:none;\" d=\"M360.9984 5 L6.3516 5 L6.3516 288.8542\"/><path style=\"fill:none;\" d=\"M638.6769 277.5 L632.8729 277.5 L632.8729 209.375\"/><path style=\"fill:none;\" d=\"M0 288.8542 L6.3516 288.8542\"/><text x=\"0\" y=\"6\" transform=\"matrix(1,0,0,1,651.5122,136.25)\" style=\"font-size:8; stroke:none;\" xml:space=\"preserve\">mattheei:gene-SMTH1 LOCUS6929</text></g><g transform=\"matrix(1,0,0,1,611.1341,414.75)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">japonicum:gene-KSF78 0008119</text></g><g transform=\"matrix(1,0,0,1,40.8464,551)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mekongi:gene-MN116 008246</text></g><g transform=\"matrix(1,0,0,1,376.9985,6)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">spindalis:gene-SSPN1 LOCUS6014</text></g><g transform=\"matrix(1,0,0,1,654.6769,278.5)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS6163</text></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                 FigTree v1.4.4, 2006-2018\n",
      "                  Tree Figure Drawing Tool\n",
      "                       Andrew Rambaut\n",
      "\n",
      "             Institute of Evolutionary Biology\n",
      "                  University of Edinburgh\n",
      "                     a.rambaut@ed.ac.uk\n",
      "\n",
      "                 http://tree.bio.ed.ac.uk/\n",
      "    Uses the Java Evolutionary Biology 2 Library (JEBL2)\n",
      "                http://jebl2.googlecode.com/\n",
      "                 Uses the iText PDF Library\n",
      "                    http://itextpdf.com/\n",
      "               Uses the Apache Batik Library\n",
      "            http://xmlgraphics.apache.org/batik/\n",
      "                 Uses the JDOM XML Library\n",
      "                    http://www.jdom.org/\n",
      " Thanks to Alexei Drummond, Joseph Heled, Philippe Lemey, \n",
      "Tulio de Oliveira, Oliver Pybus, Beth Shapiro & Marc Suchard\n",
      "\n",
      "Creating SVG graphic: data/epigenetic_genes_tree_AEBP2.svg\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'AEBP2:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" style=\"fill-opacity:1; color-rendering:auto; color-interpolation:auto; text-rendering:auto; stroke:black; stroke-linecap:square; stroke-miterlimit:10; shape-rendering:auto; stroke-opacity:1; fill:black; stroke-dasharray:none; font-weight:normal; stroke-width:1; font-family:'Dialog'; font-style:normal; stroke-linejoin:miter; font-size:12; stroke-dashoffset:0; image-rendering:auto;\" width=\"800.0\" viewBox=\"0 0 800.0 600.0\" height=\"600.0\"><!--Generated by the Batik Graphics2D SVG Generator--><defs id=\"genericDefs\"/><g><g transform=\"translate(6,6)\" style=\"text-rendering:optimizeLegibility; font-size:10; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><line y2=\"570\" style=\"fill:none;\" x1=\"286.6257\" x2=\"354.8802\" y1=\"570\"/><text x=\"313.8029\" xml:space=\"preserve\" y=\"583\" style=\"stroke:none;\">0.1</text><path style=\"fill:none;\" d=\"M641.5059 141.25 L286.0699 141.25 L286.0699 73.125\"/><path style=\"fill:none;\" d=\"M113.6654 175.3125 L6.3515 175.3125 L6.3515 379.6875\"/><path style=\"fill:none;\" d=\"M566.6751 550 L6.3515 550 L6.3515 379.6875\"/><path style=\"fill:none;\" d=\"M606.9082 277.5 L113.6654 277.5 L113.6654 175.3125\"/><path style=\"fill:none;\" d=\"M468.457 413.75 L6.3515 413.75 L6.3515 379.6875\"/><path style=\"fill:none;\" d=\"M511.9337 5 L286.0699 5 L286.0699 73.125\"/><path style=\"fill:none;\" d=\"M286.0699 73.125 L113.6654 73.125 L113.6654 175.3125\"/><path style=\"fill:none;\" d=\"M0 379.6875 L6.3515 379.6875\"/><text x=\"0\" y=\"6\" transform=\"matrix(1,0,0,1,651.5058,136.25)\" style=\"font-size:8; stroke:none;\" xml:space=\"preserve\">intercalatum:gene-SINT2 LOCUS8012</text></g><g transform=\"matrix(1,0,0,1,582.6751,551)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mekongi:gene-MN116 001510</text></g><g transform=\"matrix(1,0,0,1,622.9082,278.5)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">japonicum:gene-KSF78 0005440</text></g><g transform=\"matrix(1,0,0,1,484.457,414.75)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mattheei:gene-SMTH1 LOCUS10615</text></g><g transform=\"matrix(1,0,0,1,527.9336,6)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">spindalis:gene-SSPN1 LOCUS9217</text></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                 FigTree v1.4.4, 2006-2018\n",
      "                  Tree Figure Drawing Tool\n",
      "                       Andrew Rambaut\n",
      "\n",
      "             Institute of Evolutionary Biology\n",
      "                  University of Edinburgh\n",
      "                     a.rambaut@ed.ac.uk\n",
      "\n",
      "                 http://tree.bio.ed.ac.uk/\n",
      "    Uses the Java Evolutionary Biology 2 Library (JEBL2)\n",
      "                http://jebl2.googlecode.com/\n",
      "                 Uses the iText PDF Library\n",
      "                    http://itextpdf.com/\n",
      "               Uses the Apache Batik Library\n",
      "            http://xmlgraphics.apache.org/batik/\n",
      "                 Uses the JDOM XML Library\n",
      "                    http://www.jdom.org/\n",
      " Thanks to Alexei Drummond, Joseph Heled, Philippe Lemey, \n",
      "Tulio de Oliveira, Oliver Pybus, Beth Shapiro & Marc Suchard\n",
      "\n",
      "Creating SVG graphic: data/epigenetic_genes_tree_ARID1A.svg\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'ARID1A:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" style=\"fill-opacity:1; color-rendering:auto; color-interpolation:auto; text-rendering:auto; stroke:black; stroke-linecap:square; stroke-miterlimit:10; shape-rendering:auto; stroke-opacity:1; fill:black; stroke-dasharray:none; font-weight:normal; stroke-width:1; font-family:'Dialog'; font-style:normal; stroke-linejoin:miter; font-size:12; stroke-dashoffset:0; image-rendering:auto;\" width=\"800.0\" viewBox=\"0 0 800.0 600.0\" height=\"600.0\"><!--Generated by the Batik Graphics2D SVG Generator--><defs id=\"genericDefs\"/><g><g transform=\"translate(6,6)\" style=\"text-rendering:optimizeLegibility; font-size:10; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><line y2=\"570\" style=\"fill:none;\" x1=\"285.6173\" x2=\"356.2251\" y1=\"570\"/><text x=\"313.9712\" xml:space=\"preserve\" y=\"583\" style=\"stroke:none;\">0.1</text><path style=\"fill:none;\" d=\"M399.5356 141.25 L387.6466 141.25 L387.6466 73.125\"/><path style=\"fill:none;\" d=\"M143.2411 175.3125 L6.3549 175.3125 L6.3549 379.6875\"/><path style=\"fill:none;\" d=\"M640.1741 550 L6.3549 550 L6.3549 379.6875\"/><path style=\"fill:none;\" d=\"M641.8423 277.5 L143.2411 277.5 L143.2411 175.3125\"/><path style=\"fill:none;\" d=\"M504.509 413.75 L6.3549 413.75 L6.3549 379.6875\"/><path style=\"fill:none;\" d=\"M404.2453 5 L387.6466 5 L387.6466 73.125\"/><path style=\"fill:none;\" d=\"M387.6466 73.125 L143.2411 73.125 L143.2411 175.3125\"/><path style=\"fill:none;\" d=\"M0 379.6875 L6.3549 379.6875\"/><text x=\"0\" y=\"6\" transform=\"matrix(1,0,0,1,409.5356,136.25)\" style=\"font-size:8; stroke:none;\" xml:space=\"preserve\">mattheei:gene-SMTH1 LOCUS7396</text></g><g transform=\"matrix(1,0,0,1,656.1742,551)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mekongi:gene-MN116 002201</text></g><g transform=\"matrix(1,0,0,1,657.8424,278.5)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">japonicum:gene-KSF78 0007878</text></g><g transform=\"matrix(1,0,0,1,520.5089,414.75)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">spindalis:gene-SSPN1 LOCUS6463</text></g><g transform=\"matrix(1,0,0,1,420.2453,6)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS6554</text></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                 FigTree v1.4.4, 2006-2018\n",
      "                  Tree Figure Drawing Tool\n",
      "                       Andrew Rambaut\n",
      "\n",
      "             Institute of Evolutionary Biology\n",
      "                  University of Edinburgh\n",
      "                     a.rambaut@ed.ac.uk\n",
      "\n",
      "                 http://tree.bio.ed.ac.uk/\n",
      "    Uses the Java Evolutionary Biology 2 Library (JEBL2)\n",
      "                http://jebl2.googlecode.com/\n",
      "                 Uses the iText PDF Library\n",
      "                    http://itextpdf.com/\n",
      "               Uses the Apache Batik Library\n",
      "            http://xmlgraphics.apache.org/batik/\n",
      "                 Uses the JDOM XML Library\n",
      "                    http://www.jdom.org/\n",
      " Thanks to Alexei Drummond, Joseph Heled, Philippe Lemey, \n",
      "Tulio de Oliveira, Oliver Pybus, Beth Shapiro & Marc Suchard\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread \"main\" java.lang.RuntimeException: Error writing graphic file: data/epigenetic_genes_tree_ATR.nwk (No such file or directory)\n",
      "\tat figtree.application.FigTreeApplication.createGraphic(Unknown Source)\n",
      "\tat figtree.application.FigTreeApplication.main(Unknown Source)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'ATR:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'data/epigenetic_genes_tree_ATR.svg'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mFileNotFoundError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[335]\u001b[39m\u001b[32m, line 9\u001b[39m\n\u001b[32m      7\u001b[39m     os.system(\u001b[33mf\u001b[39m\u001b[33m'\u001b[39m\u001b[33mfigtree -graphic SVG \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mepigenetec_trees_result\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcurrent_image_file\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m)\n\u001b[32m      8\u001b[39m     display(\u001b[33mf\u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfamily\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m:\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m----> \u001b[39m\u001b[32m9\u001b[39m     display(\u001b[43mSVG\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcurrent_image_file\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[32m     11\u001b[39m \u001b[38;5;66;03m# более красивые картинки в отчете (тоже figtree, только через gui с настройками)\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/code/hse/bio/project_common/bioinf-project-common/.venv/lib/python3.13/site-packages/IPython/core/display.py:343\u001b[39m, in \u001b[36mDisplayObject.__init__\u001b[39m\u001b[34m(self, data, url, filename, metadata)\u001b[39m\n\u001b[32m    340\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.metadata \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    341\u001b[39m     \u001b[38;5;28mself\u001b[39m.metadata = {}\n\u001b[32m--> \u001b[39m\u001b[32m343\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mreload\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    344\u001b[39m \u001b[38;5;28mself\u001b[39m._check_data()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/code/hse/bio/project_common/bioinf-project-common/.venv/lib/python3.13/site-packages/IPython/core/display.py:369\u001b[39m, in \u001b[36mDisplayObject.reload\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m    367\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.filename \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    368\u001b[39m     encoding = \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01mif\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m._read_flags \u001b[38;5;28;01melse\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mutf-8\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m--> \u001b[39m\u001b[32m369\u001b[39m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_read_flags\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[32m    370\u001b[39m         \u001b[38;5;28mself\u001b[39m.data = f.read()\n\u001b[32m    371\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.url \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    372\u001b[39m     \u001b[38;5;66;03m# Deferred import\u001b[39;00m\n",
      "\u001b[31mFileNotFoundError\u001b[39m: [Errno 2] No such file or directory: 'data/epigenetic_genes_tree_ATR.svg'"
     ]
    }
   ],
   "source": [
    "# картинки деревьев\n",
    "from IPython.display import SVG, display\n",
    "import os\n",
    "\n",
    "for family, epigenetec_trees_result in epigenetec_trees_results:\n",
    "    current_image_file = f'data/epigenetic_genes_tree_{family}.svg'\n",
    "    os.system(f'figtree -graphic SVG {epigenetec_trees_result} {current_image_file}')\n",
    "    display(f'{family}:')\n",
    "    display(SVG(filename=current_image_file))\n",
    "\n",
    "# более красивые картинки в отчете (тоже figtree, только через gui с настройками)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "392245fa",
   "metadata": {},
   "source": [
    "### 2 и 3: Ищем ортологи"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "id": "e815320d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:local_programs/OrthoFinder_source.tar.gz alreasy exists\n",
      "INFO:root:local_programs/OrthoFinder_source alreasy exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OrthoFinder version 3.0.1b1 Copyright (C) 2014 David Emms\n",
      "\n",
      "SIMPLE USAGE:\n",
      "Run full OrthoFinder analysis on FASTA format proteomes in <dir>\n",
      "  orthofinder [options] -f <dir>\n",
      "\n",
      "To assign species from <dir1> to existing OrthoFinder orthogroups in <dir2>\n",
      "  orthofinder [options] --assign <dir1> --core <dir2>\n",
      "\n",
      "OPTIONS:\n",
      " -t <int>        Number of parallel sequence search threads [Default = 12]\n",
      " -a <int>        Number of parallel analysis threads\n",
      " -d              Input is DNA sequences\n",
      " -M <txt>        Method for gene tree inference. Options 'dendroblast' & 'msa'\n",
      "                 [Default = msa]\n",
      " -S <txt>        Sequence search program [Default = diamond]\n",
      "                 Options: blast, diamond, diamond_ultra_sens, blast_gz, mmseqs, blast_nucl\n",
      " -A <txt>        MSA program, requires '-M msa' [Default = mafft]\n",
      "                 Options: mafft, muscle, mafft_memsave\n",
      " -T <txt>        Tree inference method, requires '-M msa' [Default = fasttree]\n",
      "                 Options: fasttree, fasttree_fastest, raxml, raxml-ng, iqtree\n",
      " -s <file>       User-specified rooted species tree\n",
      " -I <int>        MCL inflation parameter [Default = 1.2]\n",
      " --save-space    Only create one compressed orthologs file per species\n",
      " -x <file>       Info for outputting results in OrthoXML format\n",
      " -p <dir>        Write the temporary pickle files to <dir>\n",
      " -1              Only perform one-way sequence search\n",
      " -X              Don't add species names to sequence IDs\n",
      " -y              Split paralogous clades below root of a HOG into separate HOGs\n",
      " -z              Don't trim MSAs (columns>=90% gap, min. alignment length 500)\n",
      " -n <txt>        Name to append to the results directory\n",
      " -o <txt>        Non-default results directory\n",
      " -h              Print this help text\n",
      "\n",
      "WORKFLOW STOPPING OPTIONS:\n",
      " -op             Stop after preparing input files for BLAST\n",
      " -og             Stop after inferring orthogroups\n",
      " -os             Stop after writing sequence files for orthogroups\n",
      "                 (requires '-M msa')\n",
      " -oa             Stop after inferring alignments for orthogroups\n",
      "                 (requires '-M msa')\n",
      " -ot             Stop after inferring gene trees for orthogroups \n",
      "\n",
      "WORKFLOW RESTART COMMANDS:\n",
      " -b  <dir>         Start OrthoFinder from pre-computed BLAST results in <dir>\n",
      " -fg <dir>         Start OrthoFinder from pre-computed orthogroups in <dir>\n",
      " -ft <dir>         Start OrthoFinder from pre-computed gene trees in <dir>\n",
      "\n",
      "LICENSE:\n",
      " Distributed under the GNU General Public License (GPLv3). See License.md\n",
      "\n",
      "CITATION:\n",
      " When publishing work that uses OrthoFinder please cite:\n",
      " Emms D.M. & Kelly S. (2019), Genome Biology 20:238\n",
      "\n",
      " If you use the species tree in your work then please also cite:\n",
      " Emms D.M. & Kelly S. (2017), MBE 34(12): 3267-3278\n",
      " Emms D.M. & Kelly S. (2018), bioRxiv https://doi.org/10.1101/267914\n"
     ]
    }
   ],
   "source": [
    "# установка orthofinder\n",
    "\n",
    "import logging\n",
    "import pathlib\n",
    "\n",
    "if pathlib.Path('local_programs/OrthoFinder_source.tar.gz').is_file():\n",
    "    logging.info('local_programs/OrthoFinder_source.tar.gz alreasy exists')\n",
    "else:\n",
    "    logging.info('Downloading OrthoFinder_source.tar.gz')\n",
    "    ! cd local_programs/ && wget https://github.com/davidemms/OrthoFinder/releases/download/3.0.1b1/OrthoFinder_source.tar.gz\n",
    "\n",
    "if pathlib.Path('local_programs/OrthoFinder_source').is_dir():\n",
    "    logging.info('local_programs/OrthoFinder_source alreasy exists')\n",
    "else:\n",
    "    logging.info('Extracting OrthoFinder_source.tar.gz')\n",
    "    ! cd local_programs/ && tar -xzf OrthoFinder_source.tar.gz\n",
    "\n",
    "! python local_programs/OrthoFinder_source/orthofinder.py -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2900bd95",
   "metadata": {},
   "outputs": [],
   "source": [
    "# готовим файлы для поиска ортологов\n",
    "\n",
    "import os\n",
    "\n",
    "! mkdir -p data/proteom\n",
    "for organism, proteom_file in proteom_files.items():\n",
    "    os.system(f'cp {proteom_file} data/proteom/{organism}.faa')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef9885c4",
   "metadata": {},
   "source": [
    "### 2. От каждого организма в группе выбрать 4-5 (как минимум 1 на члена группы)  генов (ортологов) в промоторах которых есть  квадруплексы и для этих квадруплексов построить выравнивание. И указать где есть вариации (в петле, числе повторов, выпетливания и т.п.) Если таких генов меньше, то для всех генов, что нашлись."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "6f4802f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No epigenetic_file specified for spindalis\n",
      "WARNING:root:No epigenetic_file specified for japonicum\n",
      "WARNING:root:No epigenetic_file specified for bovis\n",
      "WARNING:root:No epigenetic_file specified for mattheei\n",
      "WARNING:root:No epigenetic_file specified for haematobium\n",
      "WARNING:root:No epigenetic_file specified for mansoni\n",
      "WARNING:root:No epigenetic_file specified for mekongi\n"
     ]
    }
   ],
   "source": [
    "organism_to_genes_with_g4_in_promoter = dict()\n",
    "\n",
    "for organism in data_config:\n",
    "    res = get_genes_with_promoters_contains(organism, 'g4')\n",
    "    if res is None:\n",
    "        continue\n",
    "    res['organism'] = organism\n",
    "    organism_to_genes_with_g4_in_promoter[organism] = res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "8c756136",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "organism_to_genes_with_g4_in_promoter"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
