{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "007494cd",
   "metadata": {},
   "source": [
    "# Командная часть проекта"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "id": "d67a6cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_config = {\n",
    "    'spindalis' : {\n",
    "    },\n",
    "    'japonicum' : {\n",
    "    },\n",
    "    'bovis' : {\n",
    "    },\n",
    "    'mattheei' : {\n",
    "    },\n",
    "    'haematobium' : {\n",
    "    },\n",
    "    'mansoni' : {\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/237/925/GCF_000237925.1_ASM23792v2/GCF_000237925.1_ASM23792v2_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/237/925/GCF_000237925.1_ASM23792v2/GCF_000237925.1_ASM23792v2_genomic.fna.gz'\n",
    "    },\n",
    "    'intercalatum': {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'g4_file' : 'g4_segments.csv',\n",
    "        'zdna_file' : 'zdna_segments.csv',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/385/GCA_944470385.2_tdSchInte2.2/GCA_944470385.2_tdSchInte2.2_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/385/GCA_944470385.2_tdSchInte2.2/GCA_944470385.2_tdSchInte2.2_genomic.fna.gz'\n",
    "    },\n",
    "    'mekongi' : {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/034/768/735/GCA_034768735.1_ASM3476873v1/GCA_034768735.1_ASM3476873v1_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/034/768/735/GCA_034768735.1_ASM3476873v1/GCA_034768735.1_ASM3476873v1_genomic.fna.gz'\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "id": "886783cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No gff url provided for spindalis\n",
      "WARNING:root:No fna url provided for spindalis\n",
      "WARNING:root:No gff url provided for japonicum\n",
      "WARNING:root:No fna url provided for japonicum\n",
      "WARNING:root:No gff url provided for bovis\n",
      "WARNING:root:No fna url provided for bovis\n",
      "WARNING:root:No gff url provided for mattheei\n",
      "WARNING:root:No fna url provided for mattheei\n",
      "WARNING:root:No gff url provided for haematobium\n",
      "WARNING:root:No fna url provided for haematobium\n",
      "INFO:root:file data/GCF_000237925.1_ASM23792v2_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCF_000237925.1_ASM23792v2_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_944470385.2_tdSchInte2.2_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_944470385.2_tdSchInte2.2_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_034768735.1_ASM3476873v1_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_034768735.1_ASM3476873v1_genomic.fna.gz already exists\n",
      "\n",
      "INFO:root:data/GCF_000237925.1_ASM23792v2_genomic.gff already exists\n",
      "INFO:root:data/GCA_944470385.2_tdSchInte2.2_genomic.gff already exists\n",
      "INFO:root:data/GCA_034768735.1_ASM3476873v1_genomic.gff already exists\n",
      "INFO:root:data/GCF_000237925.1_ASM23792v2_genomic.fna already exists\n",
      "INFO:root:data/GCA_944470385.2_tdSchInte2.2_genomic.fna already exists\n",
      "INFO:root:data/GCA_034768735.1_ASM3476873v1_genomic.fna already exists\n"
     ]
    }
   ],
   "source": [
    "# скачаем аннотации\n",
    "\n",
    "# скачивание файлов генома и аннотации\n",
    "\n",
    "import requests\n",
    "import logging\n",
    "import pathlib\n",
    "import glob\n",
    "import os\n",
    "import sys\n",
    "import itertools\n",
    "\n",
    "logging.basicConfig(level=logging.INFO, force = True)\n",
    "\n",
    "def download_file(url: str, local_file_name: pathlib.Path):\n",
    "    if pathlib.Path(local_file_name).is_file():\n",
    "        logging.info(f'file {local_file_name} already exists')\n",
    "        return local_file_name\n",
    "    logging.info(f'downloading {local_file_name}')\n",
    "    with requests.get(url, stream=True) as r:\n",
    "        r.raise_for_status()\n",
    "        with open(local_file_name, 'wb') as out_file:\n",
    "            for chunk in r.iter_content(chunk_size=8192): \n",
    "                out_file.write(chunk)\n",
    "\n",
    "    return local_file_name\n",
    "\n",
    "gff_files = dict()\n",
    "fna_files = dict()\n",
    "\n",
    "# скачиваем аннотации\n",
    "for current_organism, current_data_config in data_config.items():\n",
    "    if 'gff_url' in current_data_config:\n",
    "        gff_files[current_organism] = download_file(current_data_config['gff_url'], f'data/{current_data_config['gff_url'].split('/')[-1]}')\n",
    "    else:\n",
    "        logging.warning(f'No gff url provided for {current_organism}')\n",
    "    \n",
    "    if 'fna_url' in current_data_config:\n",
    "        fna_files[current_organism] = download_file(current_data_config['fna_url'], f'data/{current_data_config['fna_url'].split('/')[-1]}')\n",
    "    else:\n",
    "        logging.warning(f'No fna url provided for {current_organism}')\n",
    "\n",
    "print(file=sys.stderr)\n",
    "\n",
    "# распакуем сжатые файлы для удобства\n",
    "for organism, filepath in itertools.chain(gff_files.items(), fna_files.items()):\n",
    "    current_file_path = pathlib.Path(filepath)\n",
    "    assert current_file_path.is_file()\n",
    "    if current_file_path.suffix == '.gz':\n",
    "        current_extracted_file = current_file_path.parent / pathlib.Path(filepath).stem\n",
    "        if current_extracted_file.is_file():\n",
    "            logging.info(f'{current_extracted_file} already exists')\n",
    "        else:\n",
    "            logging.info(f'extracting {current_extracted_file}')\n",
    "            os.system(f'gunzip -k {current_extracted_file}')\n",
    "\n",
    "for organism, filepath in gff_files.items():\n",
    "    if pathlib.Path(filepath).suffix == '.gz':\n",
    "        gff_files[organism] = current_file_path.parent / pathlib.Path(filepath).stem\n",
    "\n",
    "for organism, filepath in fna_files.items():\n",
    "    if pathlib.Path(filepath).suffix == '.gz':\n",
    "        fna_files[organism] = current_file_path.parent / pathlib.Path(filepath).stem\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "d9e06e08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# функции для чтения данных из результатов индивидуальных проектов\n",
    "\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "\n",
    "def get_epigenetic_genes(organism: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "\n",
    "    if 'epigenetic_file' in data_config[organism]:\n",
    "        file_name = data_config[organism]['epigenetic_file']\n",
    "        if pathlib.Path(file_name).suffix == '.csv':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['epigenetic_file']}')\n",
    "            res = res[['Family', 'Gene_ID']]\n",
    "            res.columns = ['family', 'gene']\n",
    "            return res\n",
    "        else:\n",
    "            assert False, \"Please, add support for your file type\" # TODO\n",
    "    else:\n",
    "        logging.warning(f'No epigenetic_file specified for {organism}')\n",
    "        return None\n",
    "\n",
    "def get_g4(organism: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "\n",
    "    if 'g4_file' in data_config[organism]:\n",
    "        file_name = data_config[organism]['g4_file']\n",
    "        if pathlib.Path(file_name).suffix == '.csv':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['g4_file']}')\n",
    "            return res\n",
    "        else:\n",
    "            assert False, \"Please, add support for your file type\" # TODO\n",
    "\n",
    "    else:\n",
    "        logging.warning(f'No epigenetic_file specified for {organism}')\n",
    "        return None\n",
    "\n",
    "def get_zdna(organism: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "\n",
    "    if 'zdna_file' in data_config[organism]:\n",
    "        file_name = data_config[organism]['zdna_file']\n",
    "        if pathlib.Path(file_name).suffix == '.csv':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['zdna_file']}')\n",
    "            return res\n",
    "        else:\n",
    "            assert None, \"Please, add support for your file type\" # TODO\n",
    "\n",
    "    else:\n",
    "        logging.warning(f'No epigenetic_file specified for {organism}')\n",
    "        return None\n",
    "\n",
    "# need_contain one of 'g4' or 'zdna'\n",
    "def get_genes_with_promoters_contains(organism: str, need_contain: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "    assert need_contain in ['g4', 'zdna']\n",
    "\n",
    "    segments = get_g4(organism) if need_contain == 'g4' else get_zdna(organism)\n",
    "    if segments is None:\n",
    "        return None\n",
    "\n",
    "    if organism in gff_files:\n",
    "        res = []\n",
    "        with open(gff_files[organism]) as gff_file:\n",
    "            for row in gff_file:\n",
    "                if row[0] != '#':\n",
    "                    row_splitted = row.strip().split()\n",
    "                    if row_splitted[2] == 'gene':\n",
    "                        gene_coords = (int(row_splitted[3]), int(row_splitted[4]))\n",
    "                        promoter_coords = (gene_coords[0] - 1000, gene_coords[0] - 1)\n",
    "                        if row_splitted[5] == '-':\n",
    "                            promoter_coords = (gene_coords[1] + 1, gene_coords[1] + 1000)\n",
    "\n",
    "                        # поиск квадруплексов или zdna\n",
    "                        found_segments = []\n",
    "                        for _, segment_row in segments.iterrows():\n",
    "                            if segment_row['record_id'] == row_splitted[0]:\n",
    "                                cur_interval = (segment_row['start'], segment_row['end'])\n",
    "                                if pd.Interval(*cur_interval, closed='both').overlaps(pd.Interval(*promoter_coords, closed='both')):\n",
    "                                    found_segments.append(cur_interval)\n",
    "                        \n",
    "                        if len(found_segments) != 0:\n",
    "                            gene_id = row_splitted[8].split('ID=')[1].split(';')[0]\n",
    "                            res.append([row_splitted[0], gene_id, promoter_coords[0], promoter_coords[1], found_segments])\n",
    "        res = pd.DataFrame(res)\n",
    "        res.columns = ['record_id', 'gene', 'start', 'end', 'matches']\n",
    "        return res\n",
    "    else:\n",
    "        logging.warning(f'No gff specified for {organism}')\n",
    "        return None\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "71a4aad9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:local_programs/mega.tar.gz already exists\n",
      "INFO:root:extracting local_programs/mega.tar.gz\n"
     ]
    }
   ],
   "source": [
    "# установка меги\n",
    "\n",
    "# https://drive.google.com/file/d/1beRdRYvyRw38orLmFiXE-p2paAILFvVm/view?usp=sharing\n",
    "\n",
    "\n",
    "mega_file_id = \"1beRdRYvyRw38orLmFiXE-p2paAILFvVm\"\n",
    "mega_url = f\"https://drive.google.com/uc?export=download&id={mega_file_id}\"\n",
    "\n",
    "# скачивание меги\n",
    "\n",
    "import requests\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "import logging\n",
    "\n",
    "mega_gz = pathlib.Path('local_programs/mega.tar.gz')\n",
    "mega_dir = pathlib.Path('local_programs/mega')\n",
    "\n",
    "if mega_gz.is_file():\n",
    "    logging.info(f'{mega_gz} already exists')\n",
    "else:\n",
    "    logging.info(f'downloading {mega_gz}')\n",
    "    response = requests.get(mega_url)\n",
    "    assert response.status_code == 200, 'Wrong status code'\n",
    "    with open(mega_gz, 'wb') as out_file:\n",
    "        out_file.write(response.content)\n",
    "\n",
    "if not mega_dir.is_dir():\n",
    "    logging.info(f'extracting {mega_gz}')\n",
    "    ! cd local_programs/ && mkdir -p mega && tar -xzf mega.tar.gz -C mega\n",
    "else:\n",
    "    logging.info(f'already exists')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "594378b3",
   "metadata": {},
   "source": [
    "### 1. От каждого организма в группе Выбрать из генов отвечающих за эпигенетику 4-5 (как минимум 1 на члена группы) и построить для них выравнивания и деревья."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "id": "f6da3ebd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No epigenetic_file specified for spindalis\n",
      "WARNING:root:current_epigenetic_genes(spindalis) is None\n",
      "WARNING:root:No epigenetic_file specified for japonicum\n",
      "WARNING:root:current_epigenetic_genes(japonicum) is None\n",
      "WARNING:root:No epigenetic_file specified for bovis\n",
      "WARNING:root:current_epigenetic_genes(bovis) is None\n",
      "WARNING:root:No epigenetic_file specified for mattheei\n",
      "WARNING:root:current_epigenetic_genes(mattheei) is None\n",
      "WARNING:root:No epigenetic_file specified for haematobium\n",
      "WARNING:root:current_epigenetic_genes(haematobium) is None\n",
      "WARNING:root:No epigenetic_file specified for mansoni\n",
      "WARNING:root:current_epigenetic_genes(mansoni) is None\n",
      "INFO:root:Adding genes for intercalatum\n",
      "INFO:root:Adding genes for mekongi\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family</th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDB1</td>\n",
       "      <td>gene-MN116_000900</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDB1</td>\n",
       "      <td>gene-SINT2_LOCUS7599</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HDAC1</td>\n",
       "      <td>gene-SINT2_LOCUS5878</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HDAC1</td>\n",
       "      <td>gene-MN116_003426</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HDAC10</td>\n",
       "      <td>gene-SINT2_LOCUS3442</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HDAC10</td>\n",
       "      <td>gene-MN116_003426</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>LEO1</td>\n",
       "      <td>gene-SINT2_LOCUS1768</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LEO1</td>\n",
       "      <td>gene-MN116_002968</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MAX</td>\n",
       "      <td>gene-SINT2_LOCUS8227</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MAX</td>\n",
       "      <td>gene-MN116_001502</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PRMT5</td>\n",
       "      <td>gene-SINT2_LOCUS2555</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PRMT5</td>\n",
       "      <td>gene-MN116_002471</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PTBP1</td>\n",
       "      <td>gene-MN116_001479</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PTBP1</td>\n",
       "      <td>gene-SINT2_LOCUS3559</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TTK</td>\n",
       "      <td>gene-SINT2_LOCUS7606</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TTK</td>\n",
       "      <td>gene-MN116_001475</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    family                  gene      organism\n",
       "0     DDB1     gene-MN116_000900       mekongi\n",
       "1     DDB1  gene-SINT2_LOCUS7599  intercalatum\n",
       "2    HDAC1  gene-SINT2_LOCUS5878  intercalatum\n",
       "3    HDAC1     gene-MN116_003426       mekongi\n",
       "4   HDAC10  gene-SINT2_LOCUS3442  intercalatum\n",
       "5   HDAC10     gene-MN116_003426       mekongi\n",
       "6     LEO1  gene-SINT2_LOCUS1768  intercalatum\n",
       "7     LEO1     gene-MN116_002968       mekongi\n",
       "8      MAX  gene-SINT2_LOCUS8227  intercalatum\n",
       "9      MAX     gene-MN116_001502       mekongi\n",
       "10   PRMT5  gene-SINT2_LOCUS2555  intercalatum\n",
       "11   PRMT5     gene-MN116_002471       mekongi\n",
       "12   PTBP1     gene-MN116_001479       mekongi\n",
       "13   PTBP1  gene-SINT2_LOCUS3559  intercalatum\n",
       "14     TTK  gene-SINT2_LOCUS7606  intercalatum\n",
       "15     TTK     gene-MN116_001475       mekongi"
      ]
     },
     "execution_count": 294,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# выбираем гены для выравнивания\n",
    "\n",
    "import logging\n",
    "\n",
    "family_in_organisms = dict()\n",
    "epigenetic_genes_for_alignment = []\n",
    "\n",
    "for organism in data_config:\n",
    "    current_epigenetic_genes = get_epigenetic_genes(organism)\n",
    "    if current_epigenetic_genes is not None:\n",
    "        logging.info(f'Adding genes for {organism}')\n",
    "        \n",
    "        current_epigenetic_genes['organism'] = organism\n",
    "        epigenetic_genes_for_alignment.append(current_epigenetic_genes)\n",
    "\n",
    "        for _, row in current_epigenetic_genes.iterrows():\n",
    "            if row['family'] not in family_in_organisms:\n",
    "                family_in_organisms[row['family']] = set()\n",
    "            family_in_organisms[row['family']].add(organism)\n",
    "    else:\n",
    "        logging.warning(f'current_epigenetic_genes({organism}) is None')\n",
    "\n",
    "family_in_organisms_array = []\n",
    "for family, organisms in family_in_organisms.items():\n",
    "    family_in_organisms_array.append((family, len(organisms)))\n",
    "\n",
    "family_in_organisms_array = list(sorted(family_in_organisms_array, key=lambda x : -x[1]))\n",
    "family_in_organisms_array = family_in_organisms_array[:8]\n",
    "\n",
    "selected_families = [x[0] for x in family_in_organisms_array]\n",
    "\n",
    "epigenetic_genes_for_alignment = pd.concat(epigenetic_genes_for_alignment)\n",
    "epigenetic_genes_for_alignment = epigenetic_genes_for_alignment[epigenetic_genes_for_alignment['family'].apply(lambda x : x in selected_families)]\n",
    "epigenetic_genes_for_alignment.sort_values('family', inplace=True)\n",
    "epigenetic_genes_for_alignment.drop_duplicates(subset=['family', 'organism'], keep='first', inplace=True)\n",
    "epigenetic_genes_for_alignment.reset_index(inplace=True, drop=True)\n",
    "epigenetic_genes_for_alignment.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "id": "db368604",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No such gff file for spindalis\n",
      "WARNING:root:No such gff file for japonicum\n",
      "WARNING:root:No such gff file for bovis\n",
      "WARNING:root:No such gff file for mattheei\n",
      "WARNING:root:No such gff file for haematobium\n",
      "WARNING:root:No such gff file for mansoni\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family</th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "      <th>record_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDB1</td>\n",
       "      <td>gene-MN116_000900</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>15786884</td>\n",
       "      <td>15806637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDB1</td>\n",
       "      <td>gene-SINT2_LOCUS7599</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>14043184</td>\n",
       "      <td>14073399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HDAC1</td>\n",
       "      <td>gene-SINT2_LOCUS5878</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103880.1</td>\n",
       "      <td>6400392</td>\n",
       "      <td>6417942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HDAC1</td>\n",
       "      <td>gene-MN116_003426</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068306.1</td>\n",
       "      <td>42561196</td>\n",
       "      <td>42566841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HDAC10</td>\n",
       "      <td>gene-SINT2_LOCUS3442</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103878.1</td>\n",
       "      <td>681741</td>\n",
       "      <td>747814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HDAC10</td>\n",
       "      <td>gene-MN116_003426</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068306.1</td>\n",
       "      <td>42561196</td>\n",
       "      <td>42566841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>LEO1</td>\n",
       "      <td>gene-SINT2_LOCUS1768</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103876.1</td>\n",
       "      <td>67947331</td>\n",
       "      <td>67961316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LEO1</td>\n",
       "      <td>gene-MN116_002968</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068306.1</td>\n",
       "      <td>21472172</td>\n",
       "      <td>21486037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MAX</td>\n",
       "      <td>gene-SINT2_LOCUS8227</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>35422182</td>\n",
       "      <td>35430203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MAX</td>\n",
       "      <td>gene-MN116_001502</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>41066770</td>\n",
       "      <td>41090611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PRMT5</td>\n",
       "      <td>gene-SINT2_LOCUS2555</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103877.1</td>\n",
       "      <td>13937814</td>\n",
       "      <td>13955141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PRMT5</td>\n",
       "      <td>gene-MN116_002471</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>85191758</td>\n",
       "      <td>85209253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PTBP1</td>\n",
       "      <td>gene-MN116_001479</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>40401660</td>\n",
       "      <td>40410673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PTBP1</td>\n",
       "      <td>gene-SINT2_LOCUS3559</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103878.1</td>\n",
       "      <td>4966110</td>\n",
       "      <td>4972483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TTK</td>\n",
       "      <td>gene-SINT2_LOCUS7606</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>14351506</td>\n",
       "      <td>14383750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TTK</td>\n",
       "      <td>gene-MN116_001475</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>40230330</td>\n",
       "      <td>40239008</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    family                  gene      organism   record_id     start       end\n",
       "0     DDB1     gene-MN116_000900       mekongi  CM068305.1  15786884  15806637\n",
       "1     DDB1  gene-SINT2_LOCUS7599  intercalatum  OX103883.1  14043184  14073399\n",
       "2    HDAC1  gene-SINT2_LOCUS5878  intercalatum  OX103880.1   6400392   6417942\n",
       "3    HDAC1     gene-MN116_003426       mekongi  CM068306.1  42561196  42566841\n",
       "4   HDAC10  gene-SINT2_LOCUS3442  intercalatum  OX103878.1    681741    747814\n",
       "5   HDAC10     gene-MN116_003426       mekongi  CM068306.1  42561196  42566841\n",
       "6     LEO1  gene-SINT2_LOCUS1768  intercalatum  OX103876.1  67947331  67961316\n",
       "7     LEO1     gene-MN116_002968       mekongi  CM068306.1  21472172  21486037\n",
       "8      MAX  gene-SINT2_LOCUS8227  intercalatum  OX103883.1  35422182  35430203\n",
       "9      MAX     gene-MN116_001502       mekongi  CM068305.1  41066770  41090611\n",
       "10   PRMT5  gene-SINT2_LOCUS2555  intercalatum  OX103877.1  13937814  13955141\n",
       "11   PRMT5     gene-MN116_002471       mekongi  CM068305.1  85191758  85209253\n",
       "12   PTBP1     gene-MN116_001479       mekongi  CM068305.1  40401660  40410673\n",
       "13   PTBP1  gene-SINT2_LOCUS3559  intercalatum  OX103878.1   4966110   4972483\n",
       "14     TTK  gene-SINT2_LOCUS7606  intercalatum  OX103883.1  14351506  14383750\n",
       "15     TTK     gene-MN116_001475       mekongi  CM068305.1  40230330  40239008"
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Добавим в табличку хромосому координаты каждого\n",
    "\n",
    "import logging\n",
    "\n",
    "gene_organism_to_position = dict()\n",
    "\n",
    "for _, row in epigenetic_genes_for_alignment.iterrows():\n",
    "    gene_organism_to_position[(row['organism'], row['gene'])] = None\n",
    "\n",
    "for organism in data_config:\n",
    "    if organism in gff_files:\n",
    "        with open(gff_files[organism], 'r') as gff_file:\n",
    "            for row in gff_file:\n",
    "                current_row = row.strip()\n",
    "                if len(current_row) == 0 or current_row[0] == '#':\n",
    "                    continue\n",
    "                current_row_splitted = row.split()\n",
    "\n",
    "                if current_row_splitted[2] == 'gene':\n",
    "                    current_gene_id = current_row_splitted[8].split('ID=')[1].split(';')[0]\n",
    "                    if (organism, current_gene_id) in gene_organism_to_position:\n",
    "                        gene_organism_to_position[(organism, current_gene_id)] = (current_row_splitted[0], int(current_row_splitted[3]), int(current_row_splitted[4]))\n",
    "    else:\n",
    "        logging.warning(f'No such gff file for {organism}')\n",
    "\n",
    "def get_position(row):\n",
    "    res = gene_organism_to_position[(row['organism'], row['gene'])]\n",
    "    return res\n",
    "\n",
    "epigenetic_genes_for_alignment[['record_id', 'start', 'end']] = \\\n",
    "    epigenetic_genes_for_alignment.apply(get_position, axis=1, result_type='expand')\n",
    "\n",
    "epigenetic_genes_for_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "id": "6555a129",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No such fna file for spindalis\n",
      "WARNING:root:No such fna file for japonicum\n",
      "WARNING:root:No such fna file for bovis\n",
      "WARNING:root:No such fna file for mattheei\n",
      "WARNING:root:No such fna file for haematobium\n",
      "WARNING:root:No such fna file for mansoni\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family</th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "      <th>record_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDB1</td>\n",
       "      <td>gene-MN116_000900</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>15786884</td>\n",
       "      <td>15806637</td>\n",
       "      <td>ttagaatgcaaaacgagtacgtatatcttcaagacgtttagcaata...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDB1</td>\n",
       "      <td>gene-SINT2_LOCUS7599</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>14043184</td>\n",
       "      <td>14073399</td>\n",
       "      <td>tatcgcgaagtccttgataactcattattcaggtgctgctataata...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HDAC1</td>\n",
       "      <td>gene-SINT2_LOCUS5878</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103880.1</td>\n",
       "      <td>6400392</td>\n",
       "      <td>6417942</td>\n",
       "      <td>tggaacaacaataattaatattatacatacatacgacatttgccat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HDAC1</td>\n",
       "      <td>gene-MN116_003426</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068306.1</td>\n",
       "      <td>42561196</td>\n",
       "      <td>42566841</td>\n",
       "      <td>ttaattggcagcccgctctggttctatgaggccgttggcgttttta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HDAC10</td>\n",
       "      <td>gene-SINT2_LOCUS3442</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103878.1</td>\n",
       "      <td>681741</td>\n",
       "      <td>747814</td>\n",
       "      <td>agcgccaaatcaatcgtggtgctactagcattagtaaccgtaagct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HDAC10</td>\n",
       "      <td>gene-MN116_003426</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068306.1</td>\n",
       "      <td>42561196</td>\n",
       "      <td>42566841</td>\n",
       "      <td>ttaattggcagcccgctctggttctatgaggccgttggcgttttta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>LEO1</td>\n",
       "      <td>gene-SINT2_LOCUS1768</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103876.1</td>\n",
       "      <td>67947331</td>\n",
       "      <td>67961316</td>\n",
       "      <td>gtttgaaggcggctgttccagattctggcgttacgttagaaatatg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LEO1</td>\n",
       "      <td>gene-MN116_002968</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068306.1</td>\n",
       "      <td>21472172</td>\n",
       "      <td>21486037</td>\n",
       "      <td>tcagtcactttcctcttcatcactgattttcgctcgtactttctta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MAX</td>\n",
       "      <td>gene-SINT2_LOCUS8227</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>35422182</td>\n",
       "      <td>35430203</td>\n",
       "      <td>atggctcgatgcctaaattcatacgctcagaaagtcatgcatgaag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MAX</td>\n",
       "      <td>gene-MN116_001502</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>41066770</td>\n",
       "      <td>41090611</td>\n",
       "      <td>tcaatcttttacttccctcccactttcattcccgacatcattgtat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PRMT5</td>\n",
       "      <td>gene-SINT2_LOCUS2555</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103877.1</td>\n",
       "      <td>13937814</td>\n",
       "      <td>13955141</td>\n",
       "      <td>gctcgaaagctgtttatttcattatgattaaaatataatgcaaaca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PRMT5</td>\n",
       "      <td>gene-MN116_002471</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>85191758</td>\n",
       "      <td>85209253</td>\n",
       "      <td>ttattcctgttgtatccctgttaaagtaaacagaaaaataaagaga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PTBP1</td>\n",
       "      <td>gene-MN116_001479</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>40401660</td>\n",
       "      <td>40410673</td>\n",
       "      <td>ttaaatcaattctccacaccgagaaatagtaacaggttccttgact...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PTBP1</td>\n",
       "      <td>gene-SINT2_LOCUS3559</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103878.1</td>\n",
       "      <td>4966110</td>\n",
       "      <td>4972483</td>\n",
       "      <td>actgaaaagaaaacgtttttaattaaggcaaacattttttatttac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TTK</td>\n",
       "      <td>gene-SINT2_LOCUS7606</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>14351506</td>\n",
       "      <td>14383750</td>\n",
       "      <td>tcagcgctagttgtggatttcagggtattttattgtttttgtactc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TTK</td>\n",
       "      <td>gene-MN116_001475</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>40230330</td>\n",
       "      <td>40239008</td>\n",
       "      <td>ctagaaaacacgttcatagtaagaattatcatcgggtagaggtgga...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    family                  gene      organism   record_id     start  \\\n",
       "0     DDB1     gene-MN116_000900       mekongi  CM068305.1  15786884   \n",
       "1     DDB1  gene-SINT2_LOCUS7599  intercalatum  OX103883.1  14043184   \n",
       "2    HDAC1  gene-SINT2_LOCUS5878  intercalatum  OX103880.1   6400392   \n",
       "3    HDAC1     gene-MN116_003426       mekongi  CM068306.1  42561196   \n",
       "4   HDAC10  gene-SINT2_LOCUS3442  intercalatum  OX103878.1    681741   \n",
       "5   HDAC10     gene-MN116_003426       mekongi  CM068306.1  42561196   \n",
       "6     LEO1  gene-SINT2_LOCUS1768  intercalatum  OX103876.1  67947331   \n",
       "7     LEO1     gene-MN116_002968       mekongi  CM068306.1  21472172   \n",
       "8      MAX  gene-SINT2_LOCUS8227  intercalatum  OX103883.1  35422182   \n",
       "9      MAX     gene-MN116_001502       mekongi  CM068305.1  41066770   \n",
       "10   PRMT5  gene-SINT2_LOCUS2555  intercalatum  OX103877.1  13937814   \n",
       "11   PRMT5     gene-MN116_002471       mekongi  CM068305.1  85191758   \n",
       "12   PTBP1     gene-MN116_001479       mekongi  CM068305.1  40401660   \n",
       "13   PTBP1  gene-SINT2_LOCUS3559  intercalatum  OX103878.1   4966110   \n",
       "14     TTK  gene-SINT2_LOCUS7606  intercalatum  OX103883.1  14351506   \n",
       "15     TTK     gene-MN116_001475       mekongi  CM068305.1  40230330   \n",
       "\n",
       "         end                                               data  \n",
       "0   15806637  ttagaatgcaaaacgagtacgtatatcttcaagacgtttagcaata...  \n",
       "1   14073399  tatcgcgaagtccttgataactcattattcaggtgctgctataata...  \n",
       "2    6417942  tggaacaacaataattaatattatacatacatacgacatttgccat...  \n",
       "3   42566841  ttaattggcagcccgctctggttctatgaggccgttggcgttttta...  \n",
       "4     747814  agcgccaaatcaatcgtggtgctactagcattagtaaccgtaagct...  \n",
       "5   42566841  ttaattggcagcccgctctggttctatgaggccgttggcgttttta...  \n",
       "6   67961316  gtttgaaggcggctgttccagattctggcgttacgttagaaatatg...  \n",
       "7   21486037  tcagtcactttcctcttcatcactgattttcgctcgtactttctta...  \n",
       "8   35430203  atggctcgatgcctaaattcatacgctcagaaagtcatgcatgaag...  \n",
       "9   41090611  tcaatcttttacttccctcccactttcattcccgacatcattgtat...  \n",
       "10  13955141  gctcgaaagctgtttatttcattatgattaaaatataatgcaaaca...  \n",
       "11  85209253  ttattcctgttgtatccctgttaaagtaaacagaaaaataaagaga...  \n",
       "12  40410673  ttaaatcaattctccacaccgagaaatagtaacaggttccttgact...  \n",
       "13   4972483  actgaaaagaaaacgtttttaattaaggcaaacattttttatttac...  \n",
       "14  14383750  tcagcgctagttgtggatttcagggtattttattgtttttgtactc...  \n",
       "15  40239008  ctagaaaacacgttcatagtaagaattatcatcgggtagaggtgga...  "
      ]
     },
     "execution_count": 297,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Добавим в табличку последовательность гена\n",
    "\n",
    "import pandas as pd\n",
    "import logging\n",
    "from Bio import SeqIO\n",
    "\n",
    "record_to_position = dict()\n",
    "\n",
    "for _, row in epigenetic_genes_for_alignment.iterrows():\n",
    "    if row['record_id'] not in record_to_position:\n",
    "        record_to_position[row['record_id']] = dict()\n",
    "    record_to_position[row['record_id']][(row['start'], row['end'])] = None\n",
    "\n",
    "record_to_position\n",
    "\n",
    "for organism in data_config:\n",
    "    if organism in fna_files:\n",
    "        records = list(SeqIO.parse(fna_files[organism], \"fasta\"))\n",
    "        for record in records:\n",
    "            if record.id in record_to_position:\n",
    "                for requested_range in record_to_position[record.id]:\n",
    "                    record_to_position[record.id][requested_range] = record.seq[requested_range[0]-1:requested_range[1]]\n",
    "\n",
    "    else:\n",
    "        logging.warning(f'No such fna file for {organism}')\n",
    "\n",
    "def get_sequence(row):\n",
    "    return [record_to_position[row['record_id']][(row['start'], row['end'])]]\n",
    "        \n",
    "epigenetic_genes_for_alignment[['data']] = \\\n",
    "    epigenetic_genes_for_alignment.apply(get_sequence, axis=1, result_type='expand')\n",
    "\n",
    "epigenetic_genes_for_alignment['data'] = epigenetic_genes_for_alignment['data'].apply(lambda s : str(s).lower())\n",
    "epigenetic_genes_for_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "id": "0609546e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Running allignment for DDB1\n",
      "INFO:root:Allignment for DDB1 done\n",
      "INFO:root:Running allignment for HDAC1\n",
      "INFO:root:Allignment for HDAC1 done\n",
      "INFO:root:Running allignment for HDAC10\n",
      "INFO:root:Allignment for HDAC10 done\n",
      "INFO:root:Running allignment for LEO1\n",
      "INFO:root:Allignment for LEO1 done\n",
      "INFO:root:Running allignment for MAX\n",
      "INFO:root:Allignment for MAX done\n",
      "INFO:root:Running allignment for PRMT5\n",
      "INFO:root:Allignment for PRMT5 done\n",
      "INFO:root:Running allignment for PTBP1\n",
      "INFO:root:Allignment for PTBP1 done\n",
      "INFO:root:Running allignment for TTK\n",
      "INFO:root:Allignment for TTK done\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import logging\n",
    "\n",
    "# настройки меги для выравнивания\n",
    "with open(\"data/muscle_align.mao\", \"w\") as f:\n",
    "    f.write(\"\"\"; Please do not edit this file! If this file is modified, results are unpredictable.\n",
    "; Instead of modifying this file, simply create a new MEGA Analysis Options file by using the MEGA Prototyper.\n",
    "[ DataSettings ]\n",
    "datatype                 = snNucleotide            \n",
    "containsCodingNuc        = False                   \n",
    "missingBaseSymbol        = ?                       \n",
    "identicalBaseSymbol      = .                       \n",
    "gapSymbol                = -                       \n",
    "[ ProcessTypes ]\n",
    "ppAlign                  = true                    \n",
    "ppClustalW               = true                    \n",
    "[ AnalysisSettings ]\n",
    "Pairwise Alignment       = ====================    \n",
    "DNAPWGapOpeningPenalty   = 15.00                   \n",
    "DNAPWGapExtensionPenalty = 6.66                    \n",
    "Multiple Alignment       = ====================    \n",
    "DNAMAGapOpeningPenalty   = 15.00                   \n",
    "DNAMAGapExtensionPenalty = 6.66                    \n",
    "Global Options           = ====================    \n",
    "DNA Weight Matrix        = IUB                     \n",
    "TransitionWeightNEdit    = 0.50                    \n",
    "UseNegativeMatrix        = ON                      \n",
    "DelayDivergentCutoff     = 30                      \n",
    "KeepPredefinedGaps       = True\"\"\")\n",
    "\n",
    "epigenetec_alignment_results = []\n",
    "\n",
    "for family, subdf in epigenetic_genes_for_alignment.groupby('family'):\n",
    "    # запишем последовательности в fasta файл\n",
    "    current_fasta_file = f'data/epigenetic_genes_for_alignment_{family}.fasta'\n",
    "    current_meg_file = f'data/epigenetic_genes_alignment_{family}.meg'\n",
    "    with open(current_fasta_file, 'w') as fasta_file:\n",
    "        for _, row in subdf.iterrows():\n",
    "            seq_id, seq = f'{row['organism']}:{row['gene']}', row['data']\n",
    "            fasta_file.write(f\">{seq_id}\\n{seq}\\n\")\n",
    "    # запуск меги\n",
    "    logging.info(f'Running allignment for {family}')\n",
    "    os.system(f'./local_programs/mega/megacc -a data/muscle_align.mao -d {current_fasta_file} -o {current_meg_file} > /dev/null 2>&1')\n",
    "    epigenetec_alignment_results.append((family, current_meg_file))\n",
    "    logging.info(f'Allignment for {family} done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df453ec4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Running tree building for DDB1\n",
      "INFO:root:Tree building for DDB1 done\n",
      "INFO:root:Running tree building for HDAC1\n",
      "INFO:root:Tree building for HDAC1 done\n",
      "INFO:root:Running tree building for HDAC10\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/17/2025 09:03:03 PM\n",
      "\n",
      "Data file                                         : epigenetic_genes_alignment_DDB1.meg\n",
      "Status                                            : Reading header\n",
      "Current Progress                                  : 1% \n",
      "Status                                            : Setting up genes/domains information                                      MEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:03 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_DDB1.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_DDB1.nwk\n",
      "  Message      = One or more domains not specified as having coding data were found but the mao file calls for coding data. All domains will be treated as having coding data\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "Status                                            : Organizing sequence informationMEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_DDB1.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_DDB1.nwk\n",
      "  Message      = No Code table was specified, assuming the Genetic Code Table is 'Standard'\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "MEGA-CC has logged the following error:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_DDB1.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_DDB1.nwk\n",
      "  Message      = At least 3 taxa are required\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/17/2025 09:03:04 PM\n",
      "\n",
      "Data file                                         : epigenetic_genes_alignment_HDAC1.meg\n",
      "Status                                            : Reading header\n",
      "Current Progress                                  : 1% \n",
      "Status                                            : Setting up genes/domains information                                      MEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_HDAC1.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_HDAC1.nwk\n",
      "  Message      = One or more domains not specified as having coding data were found but the mao file calls for coding data. All domains will be treated as having coding data\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "Status                                            : Organizing sequence informationMEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_HDAC1.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_HDAC1.nwk\n",
      "  Message      = No Code table was specified, assuming the Genetic Code Table is 'Standard'\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "MEGA-CC has logged the following error:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_HDAC1.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_HDAC1.nwk\n",
      "  Message      = At least 3 taxa are required\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/17/2025 09:03:04 PM\n",
      "\n",
      "Data file                                         : epigenetic_genes_alignment_HDAC10.meg\n",
      "Status                                            : Reading header\n",
      "Current Progress                                  : 1% \n",
      "Status                                            : Setting up genes/domains information                                      MEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_HDAC10.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_HDAC10.nwk\n",
      "  Message      = One or more domains not specified as having coding data were found but the mao file calls for coding data. All domains will be treated as having coding data\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Tree building for HDAC10 done\n",
      "INFO:root:Running tree building for LEO1\n",
      "INFO:root:Tree building for LEO1 done\n",
      "INFO:root:Running tree building for MAX\n",
      "INFO:root:Tree building for MAX done\n",
      "INFO:root:Running tree building for PRMT5\n",
      "INFO:root:Tree building for PRMT5 done\n",
      "INFO:root:Running tree building for PTBP1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "Status                                            : Organizing sequence informationMEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_HDAC10.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_HDAC10.nwk\n",
      "  Message      = No Code table was specified, assuming the Genetic Code Table is 'Standard'\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "MEGA-CC has logged the following error:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_HDAC10.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_HDAC10.nwk\n",
      "  Message      = At least 3 taxa are required\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/17/2025 09:03:04 PM\n",
      "\n",
      "Data file                                         : epigenetic_genes_alignment_LEO1.meg\n",
      "Status                                            : Reading header\n",
      "Current Progress                                  : 1% \n",
      "Status                                            : Setting up genes/domains information                                      MEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_LEO1.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_LEO1.nwk\n",
      "  Message      = One or more domains not specified as having coding data were found but the mao file calls for coding data. All domains will be treated as having coding data\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "Status                                            : Organizing sequence informationMEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_LEO1.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_LEO1.nwk\n",
      "  Message      = No Code table was specified, assuming the Genetic Code Table is 'Standard'\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "MEGA-CC has logged the following error:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_LEO1.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_LEO1.nwk\n",
      "  Message      = At least 3 taxa are required\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/17/2025 09:03:04 PM\n",
      "\n",
      "Data file                                         : epigenetic_genes_alignment_MAX.meg\n",
      "Status                                            : Reading header\n",
      "Current Progress                                  : 1% \n",
      "Status                                            : Setting up genes/domains information                                      MEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_MAX.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_MAX.nwk\n",
      "  Message      = One or more domains not specified as having coding data were found but the mao file calls for coding data. All domains will be treated as having coding data\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "Status                                            : Organizing sequence informationMEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_MAX.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_MAX.nwk\n",
      "  Message      = No Code table was specified, assuming the Genetic Code Table is 'Standard'\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "MEGA-CC has logged the following error:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_MAX.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_MAX.nwk\n",
      "  Message      = At least 3 taxa are required\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/17/2025 09:03:04 PM\n",
      "\n",
      "Data file                                         : epigenetic_genes_alignment_PRMT5.meg\n",
      "Status                                            : Reading header\n",
      "Current Progress                                  : 1% \n",
      "Status                                            : Setting up genes/domains information                                      MEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_PRMT5.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_PRMT5.nwk\n",
      "  Message      = One or more domains not specified as having coding data were found but the mao file calls for coding data. All domains will be treated as having coding data\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "Status                                            : Organizing sequence informationMEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_PRMT5.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_PRMT5.nwk\n",
      "  Message      = No Code table was specified, assuming the Genetic Code Table is 'Standard'\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "MEGA-CC has logged the following error:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_PRMT5.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_PRMT5.nwk\n",
      "  Message      = At least 3 taxa are required\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Tree building for PTBP1 done\n",
      "INFO:root:Running tree building for TTK\n",
      "INFO:root:Tree building for TTK done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/17/2025 09:03:04 PM\n",
      "\n",
      "Data file                                         : epigenetic_genes_alignment_PTBP1.meg\n",
      "Status                                            : Reading header\n",
      "Current Progress                                  : 1% \n",
      "Status                                            : Setting up genes/domains information                                      MEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_PTBP1.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_PTBP1.nwk\n",
      "  Message      = One or more domains not specified as having coding data were found but the mao file calls for coding data. All domains will be treated as having coding data\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "Status                                            : Organizing sequence informationMEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_PTBP1.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_PTBP1.nwk\n",
      "  Message      = No Code table was specified, assuming the Genetic Code Table is 'Standard'\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "MEGA-CC has logged the following error:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_PTBP1.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_PTBP1.nwk\n",
      "  Message      = At least 3 taxa are required\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/17/2025 09:03:04 PM\n",
      "\n",
      "Data file                                         : epigenetic_genes_alignment_TTK.meg\n",
      "Status                                            : Reading header\n",
      "Current Progress                                  : 1% \n",
      "Status                                            : Setting up genes/domains information                                      MEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_TTK.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_TTK.nwk\n",
      "  Message      = One or more domains not specified as having coding data were found but the mao file calls for coding data. All domains will be treated as having coding data\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "Status                                            : Organizing sequence informationMEGA-CC has logged the following warning:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_TTK.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_TTK.nwk\n",
      "  Message      = No Code table was specified, assuming the Genetic Code Table is 'Standard'\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "MEGA-CC has logged the following error:\n",
      "  When         = 06/17/2025 09:03:04 PM\n",
      "  Data file    = data/epigenetic_genes_alignment_TTK.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree_TTK.nwk\n",
      "  Message      = At least 3 taxa are required\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# настройки меги для построения дерева\n",
    "with open(\"data/nj_tree.mao\", \"w\") as f:\n",
    "    f.write(\"\"\"[ DataSettings ]\n",
    "datatype                             = snNucleotide                        \n",
    "containsCodingNuc                    = True                                \n",
    "MissingBaseSymbol                    = ?                                   \n",
    "IdenticalBaseSymbol                  = .                                   \n",
    "GapSymbol                            = -                                   \n",
    "[ ProcessTypes ]\n",
    "ppInfer                              = true                                \n",
    "ppML                                 = true                                \n",
    "[ AnalysisSettings ]\n",
    "Analysis                             = Phylogeny Reconstruction            \n",
    "Statistical Method                   = Maximum Likelihood                  \n",
    "Phylogeny Test                       = ====================                \n",
    "Test of Phylogeny                    = Bootstrap method                    \n",
    "No. of Bootstrap Replications        = 500                                 \n",
    "Substitution Model                   = ====================                \n",
    "Substitutions Type                   = Nucleotide                          \n",
    "Model/Method                         = Kimura 2-parameter model            \n",
    "Rates and Patterns                   = ====================                \n",
    "Rates among Sites                    = Uniform Rates                       \n",
    "No of Discrete Gamma Categories      = Not Applicable                      \n",
    "Data Subset to Use                   = ====================                \n",
    "Gaps/Missing Data Treatment          = Complete deletion                   \n",
    "Site Coverage Cutoff (%)             = Not Applicable                      \n",
    "Select Codon Positions               = 1st, 2nd, 3rd, Non-Coding           \n",
    "Tree Inference Options               = ====================                \n",
    "ML Heuristic Method                  = Nearest-Neighbor-Interchange (NNI)  \n",
    "Initial Tree for ML                  = Make initial tree automatically (Neighbor Joining)\n",
    "Branch Swap Filter                   = Strong                              \n",
    "System Resource Usage                = ====================                \n",
    "Number of Threads                    = 1                                   \n",
    "Genetic Code Table                   = Not Applicable                      \n",
    "Genetic Code                         = Not Applicable                      \n",
    "Has Time Limit                       = False                               \n",
    "Maximum Execution Time               = -1  \"\"\")\n",
    "\n",
    "epigenetec_trees_results = []\n",
    "\n",
    "# запуск меги\n",
    "for family, current_alignment_file in epigenetec_alignment_results:\n",
    "    current_nwk_file = f'data/epigenetic_genes_tree_{family}.nwk'\n",
    "    logging.info(f'Running tree building for {family}')\n",
    "    os.system(f'./local_programs/mega/megacc -a data/nj_tree.mao -d {current_alignment_file} -o {current_nwk_file} > /dev/null 2>&1')\n",
    "    epigenetec_trees_results.append((family, current_nwk_file))\n",
    "    logging.info(f'Tree building for {family} done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b82213b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                 FigTree v1.4.4, 2006-2018\n",
      "                  Tree Figure Drawing Tool\n",
      "                       Andrew Rambaut\n",
      "\n",
      "             Institute of Evolutionary Biology\n",
      "                  University of Edinburgh\n",
      "                     a.rambaut@ed.ac.uk\n",
      "\n",
      "                 http://tree.bio.ed.ac.uk/\n",
      "    Uses the Java Evolutionary Biology 2 Library (JEBL2)\n",
      "                http://jebl2.googlecode.com/\n",
      "                 Uses the iText PDF Library\n",
      "                    http://itextpdf.com/\n",
      "               Uses the Apache Batik Library\n",
      "            http://xmlgraphics.apache.org/batik/\n",
      "                 Uses the JDOM XML Library\n",
      "                    http://www.jdom.org/\n",
      " Thanks to Alexei Drummond, Joseph Heled, Philippe Lemey, \n",
      "Tulio de Oliveira, Oliver Pybus, Beth Shapiro & Marc Suchard\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread \"main\" java.lang.RuntimeException: Error writing graphic file: data/epigenetic_genes_tree_DDB1.nwk (No such file or directory)\n",
      "\tat figtree.application.FigTreeApplication.createGraphic(Unknown Source)\n",
      "\tat figtree.application.FigTreeApplication.main(Unknown Source)\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'data/epigenetic_genes_tree_DDB1.svg'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mFileNotFoundError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[314]\u001b[39m\u001b[32m, line 10\u001b[39m\n\u001b[32m      8\u001b[39m     current_image_file = \u001b[33mf\u001b[39m\u001b[33m'\u001b[39m\u001b[33mdata/epigenetic_genes_tree_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfamily\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m.svg\u001b[39m\u001b[33m'\u001b[39m\n\u001b[32m      9\u001b[39m     os.system(\u001b[33mf\u001b[39m\u001b[33m'\u001b[39m\u001b[33mfigtree -graphic SVG \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mepigenetec_trees_result\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcurrent_image_file\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m10\u001b[39m     display(\u001b[43mSVG\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcurrent_image_file\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[32m     12\u001b[39m \u001b[38;5;66;03m# более красивые картинки в отчете (тоже figtree, только через gui с настройками)\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/code/hse/bio/project_common/bioinf-project-common/.venv/lib/python3.13/site-packages/IPython/core/display.py:343\u001b[39m, in \u001b[36mDisplayObject.__init__\u001b[39m\u001b[34m(self, data, url, filename, metadata)\u001b[39m\n\u001b[32m    340\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.metadata \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    341\u001b[39m     \u001b[38;5;28mself\u001b[39m.metadata = {}\n\u001b[32m--> \u001b[39m\u001b[32m343\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mreload\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    344\u001b[39m \u001b[38;5;28mself\u001b[39m._check_data()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/code/hse/bio/project_common/bioinf-project-common/.venv/lib/python3.13/site-packages/IPython/core/display.py:369\u001b[39m, in \u001b[36mDisplayObject.reload\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m    367\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.filename \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    368\u001b[39m     encoding = \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01mif\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m._read_flags \u001b[38;5;28;01melse\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mutf-8\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m--> \u001b[39m\u001b[32m369\u001b[39m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_read_flags\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[32m    370\u001b[39m         \u001b[38;5;28mself\u001b[39m.data = f.read()\n\u001b[32m    371\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.url \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    372\u001b[39m     \u001b[38;5;66;03m# Deferred import\u001b[39;00m\n",
      "\u001b[31mFileNotFoundError\u001b[39m: [Errno 2] No such file or directory: 'data/epigenetic_genes_tree_DDB1.svg'"
     ]
    }
   ],
   "source": [
    "# картинки деревьев\n",
    "from IPython.display import SVG, display\n",
    "import os\n",
    "\n",
    "for family, epigenetec_trees_result in epigenetec_trees_results:\n",
    "    current_image_file = f'data/epigenetic_genes_tree_{family}.svg'\n",
    "    os.system(f'figtree -graphic SVG {epigenetec_trees_result} {current_image_file}')\n",
    "    display(SVG(filename=current_image_file))\n",
    "\n",
    "# более красивые картинки в отчете (тоже figtree, только через gui с настройками)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef9885c4",
   "metadata": {},
   "source": [
    "### 2. От каждого организма в группе выбрать 4-5 (как минимум 1 на члена группы)  генов (ортологов) в промоторах которых есть  квадруплексы и для этих квадруплексов построить выравнивание. И указать где есть вариации (в петле, числе повторов, выпетливания и т.п.) Если таких генов меньше, то для всех генов, что нашлись."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "6f4802f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No epigenetic_file specified for spindalis\n",
      "WARNING:root:No epigenetic_file specified for japonicum\n",
      "WARNING:root:No epigenetic_file specified for bovis\n",
      "WARNING:root:No epigenetic_file specified for mattheei\n",
      "WARNING:root:No epigenetic_file specified for haematobium\n",
      "WARNING:root:No epigenetic_file specified for mansoni\n",
      "WARNING:root:No epigenetic_file specified for mekongi\n"
     ]
    }
   ],
   "source": [
    "organism_to_genes_with_g4_in_promoter = dict()\n",
    "\n",
    "for organism in data_config:\n",
    "    res = get_genes_with_promoters_contains(organism, 'g4')\n",
    "    if res is None:\n",
    "        continue\n",
    "    res['organism'] = organism\n",
    "    organism_to_genes_with_g4_in_promoter[organism] = res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "8c756136",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "organism_to_genes_with_g4_in_promoter"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
