{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "007494cd",
   "metadata": {},
   "source": [
    "# Командная часть проекта"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "d67a6cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_config = {\n",
    "    'spindalis' : {\n",
    "    },\n",
    "    'japonicum' : {\n",
    "    },\n",
    "    'bovis' : {\n",
    "    },\n",
    "    'mattheei' : {\n",
    "    },\n",
    "    'haematobium' : {\n",
    "    },\n",
    "    'mansoni' : {\n",
    "    },\n",
    "    'intercalatum': {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'g4_file' : 'g4_segments.csv',\n",
    "        'zrnk_file' : 'zrnk_segments.csv',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/385/GCA_944470385.2_tdSchInte2.2/GCA_944470385.2_tdSchInte2.2_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/385/GCA_944470385.2_tdSchInte2.2/GCA_944470385.2_tdSchInte2.2_genomic.fna.gz'\n",
    "    },\n",
    "    'mekongi' : {\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "886783cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No gff url provided for spindalis\n",
      "WARNING:root:No fna url provided for spindalis\n",
      "WARNING:root:No gff url provided for japonicum\n",
      "WARNING:root:No fna url provided for japonicum\n",
      "WARNING:root:No gff url provided for bovis\n",
      "WARNING:root:No fna url provided for bovis\n",
      "WARNING:root:No gff url provided for mattheei\n",
      "WARNING:root:No fna url provided for mattheei\n",
      "WARNING:root:No gff url provided for haematobium\n",
      "WARNING:root:No fna url provided for haematobium\n",
      "WARNING:root:No gff url provided for mansoni\n",
      "WARNING:root:No fna url provided for mansoni\n",
      "INFO:root:file data/GCA_944470385.2_tdSchInte2.2_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_944470385.2_tdSchInte2.2_genomic.fna.gz already exists\n",
      "WARNING:root:No gff url provided for mekongi\n",
      "WARNING:root:No fna url provided for mekongi\n",
      "\n",
      "INFO:root:data/GCA_944470385.2_tdSchInte2.2_genomic.gff already exists\n",
      "INFO:root:data/GCA_944470385.2_tdSchInte2.2_genomic.fna already exists\n"
     ]
    }
   ],
   "source": [
    "# скачаем аннотации\n",
    "\n",
    "# скачивание файлов генома и аннотации\n",
    "\n",
    "import requests\n",
    "import logging\n",
    "import pathlib\n",
    "import glob\n",
    "import os\n",
    "import sys\n",
    "import itertools\n",
    "\n",
    "logging.basicConfig(level=logging.INFO, force = True)\n",
    "\n",
    "def download_file(url: str, local_file_name: pathlib.Path):\n",
    "    if pathlib.Path(local_file_name).is_file():\n",
    "        logging.info(f'file {local_file_name} already exists')\n",
    "        return local_file_name\n",
    "    logging.info(f'downloading {local_file_name}')\n",
    "    with requests.get(url, stream=True) as r:\n",
    "        r.raise_for_status()\n",
    "        with open(local_file_name, 'wb') as out_file:\n",
    "            for chunk in r.iter_content(chunk_size=8192): \n",
    "                out_file.write(chunk)\n",
    "\n",
    "    return local_file_name\n",
    "\n",
    "gff_files = dict()\n",
    "fna_files = dict()\n",
    "\n",
    "# скачиваем аннотации\n",
    "for current_organism, current_data_config in data_config.items():\n",
    "    if 'gff_url' in current_data_config:\n",
    "        gff_files[current_organism] = download_file(current_data_config['gff_url'], f'data/{current_data_config['gff_url'].split('/')[-1]}')\n",
    "    else:\n",
    "        logging.warning(f'No gff url provided for {current_organism}')\n",
    "    \n",
    "    if 'fna_url' in current_data_config:\n",
    "        fna_files[current_organism] = download_file(current_data_config['fna_url'], f'data/{current_data_config['fna_url'].split('/')[-1]}')\n",
    "    else:\n",
    "        logging.warning(f'No fna url provided for {current_organism}')\n",
    "\n",
    "print(file=sys.stderr)\n",
    "\n",
    "# распакуем сжатые файлы для удобства\n",
    "for organism, filepath in itertools.chain(gff_files.items(), fna_files.items()):\n",
    "    current_file_path = pathlib.Path(filepath)\n",
    "    assert current_file_path.is_file()\n",
    "    if current_file_path.suffix == '.gz':\n",
    "        current_extracted_file = current_file_path.parent / pathlib.Path(filepath).stem\n",
    "        if current_extracted_file.is_file():\n",
    "            logging.info(f'{current_extracted_file} already exists')\n",
    "        else:\n",
    "            logging.info(f'extracting {current_extracted_file}')\n",
    "            os.system(f'gunzip -k {current_extracted_file}')\n",
    "\n",
    "for organism, filepath in gff_files.items():\n",
    "    if pathlib.Path(filepath).suffix == '.gz':\n",
    "        gff_files[organism] = current_file_path.parent / pathlib.Path(filepath).stem\n",
    "\n",
    "for organism, filepath in fna_files.items():\n",
    "    if pathlib.Path(filepath).suffix == '.gz':\n",
    "        fna_files[organism] = current_file_path.parent / pathlib.Path(filepath).stem\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "d9e06e08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# функции для чтения данных из результатов индивидуальных проектов\n",
    "\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "\n",
    "def get_epigenetic_genes(organism: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "\n",
    "    if 'epigenetic_file' in data_config[organism]:\n",
    "        file_name = data_config[organism]['epigenetic_file']\n",
    "        if pathlib.Path(file_name).suffix == '.csv':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['epigenetic_file']}')\n",
    "            res = res[['Gene_ID']]\n",
    "            res.columns = ['gene']\n",
    "            return res\n",
    "        else:\n",
    "            assert False, \"Please, add support for your file type\" # TODO\n",
    "    else:\n",
    "        logging.warning(f'No epigenetic_file specified for {organism}')\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "594378b3",
   "metadata": {},
   "source": [
    "### 1. От каждого организма в группе Выбрать из генов отвечающих за эпигенетику 4-5 (как минимум 1 на члена группы) и построить для них выравнивания и деревья."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "f6da3ebd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No epigenetic_file specified for spindalis\n",
      "WARNING:root:current_epigenetic_genes(spindalis) is None\n",
      "WARNING:root:No epigenetic_file specified for japonicum\n",
      "WARNING:root:current_epigenetic_genes(japonicum) is None\n",
      "WARNING:root:No epigenetic_file specified for bovis\n",
      "WARNING:root:current_epigenetic_genes(bovis) is None\n",
      "WARNING:root:No epigenetic_file specified for mattheei\n",
      "WARNING:root:current_epigenetic_genes(mattheei) is None\n",
      "WARNING:root:No epigenetic_file specified for haematobium\n",
      "WARNING:root:current_epigenetic_genes(haematobium) is None\n",
      "WARNING:root:No epigenetic_file specified for mansoni\n",
      "WARNING:root:current_epigenetic_genes(mansoni) is None\n",
      "INFO:root:Adding genes for intercalatum\n",
      "WARNING:root:No epigenetic_file specified for mekongi\n",
      "WARNING:root:current_epigenetic_genes(mekongi) is None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>gene-SINT2_LOCUS1809</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>gene-SINT2_LOCUS5354</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gene-SINT2_LOCUS3201</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>gene-SINT2_LOCUS4930</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>gene-SINT2_LOCUS5557</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene      organism\n",
       "218  gene-SINT2_LOCUS1809  intercalatum\n",
       "66   gene-SINT2_LOCUS5354  intercalatum\n",
       "9    gene-SINT2_LOCUS3201  intercalatum\n",
       "170  gene-SINT2_LOCUS4930  intercalatum\n",
       "15   gene-SINT2_LOCUS5557  intercalatum"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import logging\n",
    "\n",
    "epigenetic_genes_for_alignment = []\n",
    "\n",
    "for organism in data_config:\n",
    "    current_epigenetic_genes = get_epigenetic_genes(organism)\n",
    "    if current_epigenetic_genes is not None:\n",
    "        current_epigenetic_genes_selected = current_epigenetic_genes.sample(5, random_state=42)\n",
    "        assert len(current_epigenetic_genes_selected) == 5, 'too small current_epigenetic_genes_selected size'\n",
    "        logging.info(f'Adding genes for {organism}')\n",
    "        current_epigenetic_genes_selected['organism'] = organism\n",
    "        epigenetic_genes_for_alignment.append(current_epigenetic_genes_selected)\n",
    "    else:\n",
    "        logging.warning(f'current_epigenetic_genes({organism}) is None')\n",
    "\n",
    "epigenetic_genes_for_alignment = pd.concat(epigenetic_genes_for_alignment)\n",
    "epigenetic_genes_for_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "db368604",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No such gff file for spindalis\n",
      "WARNING:root:No such gff file for japonicum\n",
      "WARNING:root:No such gff file for bovis\n",
      "WARNING:root:No such gff file for mattheei\n",
      "WARNING:root:No such gff file for haematobium\n",
      "WARNING:root:No such gff file for mansoni\n",
      "WARNING:root:No such gff file for mekongi\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "      <th>record_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>gene-SINT2_LOCUS1809</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103876.1</td>\n",
       "      <td>69497234</td>\n",
       "      <td>69537522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>gene-SINT2_LOCUS5354</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>30466976</td>\n",
       "      <td>30476049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gene-SINT2_LOCUS3201</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103877.1</td>\n",
       "      <td>38092074</td>\n",
       "      <td>38115003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>gene-SINT2_LOCUS4930</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>12765221</td>\n",
       "      <td>12771890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>gene-SINT2_LOCUS5557</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>37810964</td>\n",
       "      <td>37839258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene      organism   record_id     start       end\n",
       "218  gene-SINT2_LOCUS1809  intercalatum  OX103876.1  69497234  69537522\n",
       "66   gene-SINT2_LOCUS5354  intercalatum  OX103879.1  30466976  30476049\n",
       "9    gene-SINT2_LOCUS3201  intercalatum  OX103877.1  38092074  38115003\n",
       "170  gene-SINT2_LOCUS4930  intercalatum  OX103879.1  12765221  12771890\n",
       "15   gene-SINT2_LOCUS5557  intercalatum  OX103879.1  37810964  37839258"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Добавим в табличку хромосому координаты каждого\n",
    "\n",
    "import logging\n",
    "\n",
    "gene_organism_to_position = dict()\n",
    "\n",
    "for _, row in epigenetic_genes_for_alignment.iterrows():\n",
    "    gene_organism_to_position[(row['organism'], row['gene'])] = None\n",
    "\n",
    "for organism in data_config:\n",
    "    if organism in gff_files:\n",
    "        with open(gff_files[organism], 'r') as gff_file:\n",
    "            for row in gff_file:\n",
    "                current_row = row.strip()\n",
    "                if len(current_row) == 0 or current_row[0] == '#':\n",
    "                    continue\n",
    "                current_row_splitted = row.split()\n",
    "\n",
    "                if current_row_splitted[2] == 'gene':\n",
    "                    current_gene_id = current_row_splitted[8].split('ID=')[1].split(';')[0]\n",
    "                    if (organism, current_gene_id) in gene_organism_to_position:\n",
    "                        gene_organism_to_position[(organism, current_gene_id)] = (current_row_splitted[0], int(current_row_splitted[3]), int(current_row_splitted[4]))\n",
    "    else:\n",
    "        logging.warning(f'No such gff file for {organism}')\n",
    "\n",
    "def get_position(row):\n",
    "    res = gene_organism_to_position[(row['organism'], row['gene'])]\n",
    "    return res\n",
    "\n",
    "epigenetic_genes_for_alignment[['record_id', 'start', 'end']] = \\\n",
    "    epigenetic_genes_for_alignment.apply(get_position, axis=1, result_type='expand')\n",
    "\n",
    "epigenetic_genes_for_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "6555a129",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No such fna file for spindalis\n",
      "WARNING:root:No such fna file for japonicum\n",
      "WARNING:root:No such fna file for bovis\n",
      "WARNING:root:No such fna file for mattheei\n",
      "WARNING:root:No such fna file for haematobium\n",
      "WARNING:root:No such fna file for mansoni\n",
      "WARNING:root:No such fna file for mekongi\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "      <th>record_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>gene-SINT2_LOCUS1809</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103876.1</td>\n",
       "      <td>69497234</td>\n",
       "      <td>69537522</td>\n",
       "      <td>atggatgtatgctttatttcatttaacttctaagatgcacttcatt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>gene-SINT2_LOCUS5354</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>30466976</td>\n",
       "      <td>30476049</td>\n",
       "      <td>acagaataaaagaggtgtactgttaacaacagccagagtgcttttt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gene-SINT2_LOCUS3201</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103877.1</td>\n",
       "      <td>38092074</td>\n",
       "      <td>38115003</td>\n",
       "      <td>tctggttgtactggcggatgacttagaggtaggattttgttttcct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>gene-SINT2_LOCUS4930</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>12765221</td>\n",
       "      <td>12771890</td>\n",
       "      <td>tcataaatgtaaccgtttagcacatgtaggtggtaatgacggtgat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>gene-SINT2_LOCUS5557</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>37810964</td>\n",
       "      <td>37839258</td>\n",
       "      <td>gtttaagaagttgcgccttcttcagatcttagatggtattaacatt...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene      organism   record_id     start       end  \\\n",
       "218  gene-SINT2_LOCUS1809  intercalatum  OX103876.1  69497234  69537522   \n",
       "66   gene-SINT2_LOCUS5354  intercalatum  OX103879.1  30466976  30476049   \n",
       "9    gene-SINT2_LOCUS3201  intercalatum  OX103877.1  38092074  38115003   \n",
       "170  gene-SINT2_LOCUS4930  intercalatum  OX103879.1  12765221  12771890   \n",
       "15   gene-SINT2_LOCUS5557  intercalatum  OX103879.1  37810964  37839258   \n",
       "\n",
       "                                                  data  \n",
       "218  atggatgtatgctttatttcatttaacttctaagatgcacttcatt...  \n",
       "66   acagaataaaagaggtgtactgttaacaacagccagagtgcttttt...  \n",
       "9    tctggttgtactggcggatgacttagaggtaggattttgttttcct...  \n",
       "170  tcataaatgtaaccgtttagcacatgtaggtggtaatgacggtgat...  \n",
       "15   gtttaagaagttgcgccttcttcagatcttagatggtattaacatt...  "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Добавим в табличку последовательность гена\n",
    "\n",
    "import pandas as pd\n",
    "import logging\n",
    "from Bio import SeqIO\n",
    "\n",
    "record_to_position = dict()\n",
    "\n",
    "for _, row in epigenetic_genes_for_alignment.iterrows():\n",
    "    if row['record_id'] not in record_to_position:\n",
    "        record_to_position[row['record_id']] = dict()\n",
    "    record_to_position[row['record_id']][(row['start'], row['end'])] = None\n",
    "\n",
    "record_to_position\n",
    "\n",
    "for organism in data_config:\n",
    "    if organism in faa_files:\n",
    "        records = list(SeqIO.parse(fna_files[organism], \"fasta\"))\n",
    "        for record in records:\n",
    "            if record.id in record_to_position:\n",
    "                for requested_range in record_to_position[record.id]:\n",
    "                    record_to_position[record.id][requested_range] = record.seq[requested_range[0]-1:requested_range[1]]\n",
    "\n",
    "    else:\n",
    "        logging.warning(f'No such fna file for {organism}')\n",
    "\n",
    "def get_sequence(row):\n",
    "    return [record_to_position[row['record_id']][(row['start'], row['end'])]]\n",
    "        \n",
    "epigenetic_genes_for_alignment[['data']] = \\\n",
    "    epigenetic_genes_for_alignment.apply(get_sequence, axis=1, result_type='expand')\n",
    "\n",
    "epigenetic_genes_for_alignment['data'] = epigenetic_genes_for_alignment['data'].apply(lambda s : str(s).lower())\n",
    "epigenetic_genes_for_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0609546e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
