{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "007494cd",
   "metadata": {},
   "source": [
    "# Командная часть проекта"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "d67a6cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_config = {\n",
    "    'spindalis' : {\n",
    "    },\n",
    "    'japonicum' : {\n",
    "    },\n",
    "    'bovis' : {\n",
    "    },\n",
    "    'mattheei' : {\n",
    "    },\n",
    "    'haematobium' : {\n",
    "    },\n",
    "    'mansoni' : {\n",
    "    },\n",
    "    'intercalatum': {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'g4_file' : 'g4_segments.csv',\n",
    "        'zrnk_file' : 'zrnk_segments.csv',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/385/GCA_944470385.2_tdSchInte2.2/GCA_944470385.2_tdSchInte2.2_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/385/GCA_944470385.2_tdSchInte2.2/GCA_944470385.2_tdSchInte2.2_genomic.fna.gz'\n",
    "    },\n",
    "    'mekongi' : {\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "886783cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No gff url provided for spindalis\n",
      "WARNING:root:No fna url provided for spindalis\n",
      "WARNING:root:No gff url provided for japonicum\n",
      "WARNING:root:No fna url provided for japonicum\n",
      "WARNING:root:No gff url provided for bovis\n",
      "WARNING:root:No fna url provided for bovis\n",
      "WARNING:root:No gff url provided for mattheei\n",
      "WARNING:root:No fna url provided for mattheei\n",
      "WARNING:root:No gff url provided for haematobium\n",
      "WARNING:root:No fna url provided for haematobium\n",
      "WARNING:root:No gff url provided for mansoni\n",
      "WARNING:root:No fna url provided for mansoni\n",
      "INFO:root:file data/GCA_944470385.2_tdSchInte2.2_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_944470385.2_tdSchInte2.2_genomic.fna.gz already exists\n",
      "WARNING:root:No gff url provided for mekongi\n",
      "WARNING:root:No fna url provided for mekongi\n",
      "\n",
      "INFO:root:data/GCA_944470385.2_tdSchInte2.2_genomic.gff already exists\n",
      "INFO:root:data/GCA_944470385.2_tdSchInte2.2_genomic.fna already exists\n"
     ]
    }
   ],
   "source": [
    "# скачаем аннотации\n",
    "\n",
    "# скачивание файлов генома и аннотации\n",
    "\n",
    "import requests\n",
    "import logging\n",
    "import pathlib\n",
    "import glob\n",
    "import os\n",
    "import sys\n",
    "import itertools\n",
    "\n",
    "logging.basicConfig(level=logging.INFO, force = True)\n",
    "\n",
    "def download_file(url: str, local_file_name: pathlib.Path):\n",
    "    if pathlib.Path(local_file_name).is_file():\n",
    "        logging.info(f'file {local_file_name} already exists')\n",
    "        return local_file_name\n",
    "    logging.info(f'downloading {local_file_name}')\n",
    "    with requests.get(url, stream=True) as r:\n",
    "        r.raise_for_status()\n",
    "        with open(local_file_name, 'wb') as out_file:\n",
    "            for chunk in r.iter_content(chunk_size=8192): \n",
    "                out_file.write(chunk)\n",
    "\n",
    "    return local_file_name\n",
    "\n",
    "gff_files = dict()\n",
    "fna_files = dict()\n",
    "\n",
    "# скачиваем аннотации\n",
    "for current_organism, current_data_config in data_config.items():\n",
    "    if 'gff_url' in current_data_config:\n",
    "        gff_files[current_organism] = download_file(current_data_config['gff_url'], f'data/{current_data_config['gff_url'].split('/')[-1]}')\n",
    "    else:\n",
    "        logging.warning(f'No gff url provided for {current_organism}')\n",
    "    \n",
    "    if 'fna_url' in current_data_config:\n",
    "        fna_files[current_organism] = download_file(current_data_config['fna_url'], f'data/{current_data_config['fna_url'].split('/')[-1]}')\n",
    "    else:\n",
    "        logging.warning(f'No fna url provided for {current_organism}')\n",
    "\n",
    "print(file=sys.stderr)\n",
    "\n",
    "# распакуем сжатые файлы для удобства\n",
    "for organism, filepath in itertools.chain(gff_files.items(), fna_files.items()):\n",
    "    current_file_path = pathlib.Path(filepath)\n",
    "    assert current_file_path.is_file()\n",
    "    if current_file_path.suffix == '.gz':\n",
    "        current_extracted_file = current_file_path.parent / pathlib.Path(filepath).stem\n",
    "        if current_extracted_file.is_file():\n",
    "            logging.info(f'{current_extracted_file} already exists')\n",
    "        else:\n",
    "            logging.info(f'extracting {current_extracted_file}')\n",
    "            os.system(f'gunzip -k {current_extracted_file}')\n",
    "\n",
    "for organism, filepath in gff_files.items():\n",
    "    if pathlib.Path(filepath).suffix == '.gz':\n",
    "        gff_files[organism] = current_file_path.parent / pathlib.Path(filepath).stem\n",
    "\n",
    "for organism, filepath in fna_files.items():\n",
    "    if pathlib.Path(filepath).suffix == '.gz':\n",
    "        fna_files[organism] = current_file_path.parent / pathlib.Path(filepath).stem\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "d9e06e08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# функции для чтения данных из результатов индивидуальных проектов\n",
    "\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "\n",
    "def get_epigenetic_genes(organism: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "\n",
    "    if 'epigenetic_file' in data_config[organism]:\n",
    "        file_name = data_config[organism]['epigenetic_file']\n",
    "        if pathlib.Path(file_name).suffix == '.csv':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['epigenetic_file']}')\n",
    "            res = res[['Gene_ID']]\n",
    "            res.columns = ['gene']\n",
    "            return res\n",
    "        else:\n",
    "            assert False, \"Please, add support for your file type\" # TODO\n",
    "    else:\n",
    "        logging.warning(f'No epigenetic_file specified for {organism}')\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "71a4aad9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:local_programs/mega.tar.gz already exists\n",
      "INFO:root:extracting local_programs/mega.tar.gz\n"
     ]
    }
   ],
   "source": [
    "# установка меги\n",
    "\n",
    "# https://drive.google.com/file/d/1beRdRYvyRw38orLmFiXE-p2paAILFvVm/view?usp=sharing\n",
    "\n",
    "\n",
    "mega_file_id = \"1beRdRYvyRw38orLmFiXE-p2paAILFvVm\"\n",
    "mega_url = f\"https://drive.google.com/uc?export=download&id={mega_file_id}\"\n",
    "\n",
    "# скачивание меги\n",
    "\n",
    "import requests\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "import logging\n",
    "\n",
    "mega_gz = pathlib.Path('local_programs/mega.tar.gz')\n",
    "mega_dir = pathlib.Path('local_programs/mega')\n",
    "\n",
    "if mega_gz.is_file():\n",
    "    logging.info(f'{mega_gz} already exists')\n",
    "else:\n",
    "    logging.info(f'downloading {mega_gz}')\n",
    "    response = requests.get(mega_url)\n",
    "    assert response.status_code == 200, 'Wrong status code'\n",
    "    with open(mega_gz, 'wb') as out_file:\n",
    "        out_file.write(response.content)\n",
    "\n",
    "if not mega_dir.is_dir():\n",
    "    logging.info(f'extracting {mega_gz}')\n",
    "    ! cd local_programs/ && mkdir -p mega && tar -xzf mega.tar.gz -C mega\n",
    "else:\n",
    "    logging.info(f'already exists')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "594378b3",
   "metadata": {},
   "source": [
    "### 1. От каждого организма в группе Выбрать из генов отвечающих за эпигенетику 4-5 (как минимум 1 на члена группы) и построить для них выравнивания и деревья."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "f6da3ebd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No epigenetic_file specified for spindalis\n",
      "WARNING:root:current_epigenetic_genes(spindalis) is None\n",
      "WARNING:root:No epigenetic_file specified for japonicum\n",
      "WARNING:root:current_epigenetic_genes(japonicum) is None\n",
      "WARNING:root:No epigenetic_file specified for bovis\n",
      "WARNING:root:current_epigenetic_genes(bovis) is None\n",
      "WARNING:root:No epigenetic_file specified for mattheei\n",
      "WARNING:root:current_epigenetic_genes(mattheei) is None\n",
      "WARNING:root:No epigenetic_file specified for haematobium\n",
      "WARNING:root:current_epigenetic_genes(haematobium) is None\n",
      "WARNING:root:No epigenetic_file specified for mansoni\n",
      "WARNING:root:current_epigenetic_genes(mansoni) is None\n",
      "INFO:root:Adding genes for intercalatum\n",
      "WARNING:root:No epigenetic_file specified for mekongi\n",
      "WARNING:root:current_epigenetic_genes(mekongi) is None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>gene-SINT2_LOCUS1809</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>gene-SINT2_LOCUS5354</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gene-SINT2_LOCUS3201</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>gene-SINT2_LOCUS4930</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>gene-SINT2_LOCUS5557</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene      organism\n",
       "218  gene-SINT2_LOCUS1809  intercalatum\n",
       "66   gene-SINT2_LOCUS5354  intercalatum\n",
       "9    gene-SINT2_LOCUS3201  intercalatum\n",
       "170  gene-SINT2_LOCUS4930  intercalatum\n",
       "15   gene-SINT2_LOCUS5557  intercalatum"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import logging\n",
    "\n",
    "epigenetic_genes_for_alignment = []\n",
    "\n",
    "for organism in data_config:\n",
    "    current_epigenetic_genes = get_epigenetic_genes(organism)\n",
    "    if current_epigenetic_genes is not None:\n",
    "        current_epigenetic_genes_selected = current_epigenetic_genes.sample(5, random_state=42)\n",
    "        assert len(current_epigenetic_genes_selected) == 5, 'too small current_epigenetic_genes_selected size'\n",
    "        logging.info(f'Adding genes for {organism}')\n",
    "        current_epigenetic_genes_selected['organism'] = organism\n",
    "        epigenetic_genes_for_alignment.append(current_epigenetic_genes_selected)\n",
    "    else:\n",
    "        logging.warning(f'current_epigenetic_genes({organism}) is None')\n",
    "\n",
    "epigenetic_genes_for_alignment = pd.concat(epigenetic_genes_for_alignment)\n",
    "epigenetic_genes_for_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "db368604",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No such gff file for spindalis\n",
      "WARNING:root:No such gff file for japonicum\n",
      "WARNING:root:No such gff file for bovis\n",
      "WARNING:root:No such gff file for mattheei\n",
      "WARNING:root:No such gff file for haematobium\n",
      "WARNING:root:No such gff file for mansoni\n",
      "WARNING:root:No such gff file for mekongi\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "      <th>record_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>gene-SINT2_LOCUS1809</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103876.1</td>\n",
       "      <td>69497234</td>\n",
       "      <td>69537522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>gene-SINT2_LOCUS5354</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>30466976</td>\n",
       "      <td>30476049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gene-SINT2_LOCUS3201</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103877.1</td>\n",
       "      <td>38092074</td>\n",
       "      <td>38115003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>gene-SINT2_LOCUS4930</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>12765221</td>\n",
       "      <td>12771890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>gene-SINT2_LOCUS5557</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>37810964</td>\n",
       "      <td>37839258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene      organism   record_id     start       end\n",
       "218  gene-SINT2_LOCUS1809  intercalatum  OX103876.1  69497234  69537522\n",
       "66   gene-SINT2_LOCUS5354  intercalatum  OX103879.1  30466976  30476049\n",
       "9    gene-SINT2_LOCUS3201  intercalatum  OX103877.1  38092074  38115003\n",
       "170  gene-SINT2_LOCUS4930  intercalatum  OX103879.1  12765221  12771890\n",
       "15   gene-SINT2_LOCUS5557  intercalatum  OX103879.1  37810964  37839258"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Добавим в табличку хромосому координаты каждого\n",
    "\n",
    "import logging\n",
    "\n",
    "gene_organism_to_position = dict()\n",
    "\n",
    "for _, row in epigenetic_genes_for_alignment.iterrows():\n",
    "    gene_organism_to_position[(row['organism'], row['gene'])] = None\n",
    "\n",
    "for organism in data_config:\n",
    "    if organism in gff_files:\n",
    "        with open(gff_files[organism], 'r') as gff_file:\n",
    "            for row in gff_file:\n",
    "                current_row = row.strip()\n",
    "                if len(current_row) == 0 or current_row[0] == '#':\n",
    "                    continue\n",
    "                current_row_splitted = row.split()\n",
    "\n",
    "                if current_row_splitted[2] == 'gene':\n",
    "                    current_gene_id = current_row_splitted[8].split('ID=')[1].split(';')[0]\n",
    "                    if (organism, current_gene_id) in gene_organism_to_position:\n",
    "                        gene_organism_to_position[(organism, current_gene_id)] = (current_row_splitted[0], int(current_row_splitted[3]), int(current_row_splitted[4]))\n",
    "    else:\n",
    "        logging.warning(f'No such gff file for {organism}')\n",
    "\n",
    "def get_position(row):\n",
    "    res = gene_organism_to_position[(row['organism'], row['gene'])]\n",
    "    return res\n",
    "\n",
    "epigenetic_genes_for_alignment[['record_id', 'start', 'end']] = \\\n",
    "    epigenetic_genes_for_alignment.apply(get_position, axis=1, result_type='expand')\n",
    "\n",
    "epigenetic_genes_for_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "6555a129",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:No such fna file for spindalis\n",
      "WARNING:root:No such fna file for japonicum\n",
      "WARNING:root:No such fna file for bovis\n",
      "WARNING:root:No such fna file for mattheei\n",
      "WARNING:root:No such fna file for haematobium\n",
      "WARNING:root:No such fna file for mansoni\n",
      "WARNING:root:No such fna file for mekongi\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "      <th>record_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>gene-SINT2_LOCUS1809</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103876.1</td>\n",
       "      <td>69497234</td>\n",
       "      <td>69537522</td>\n",
       "      <td>atggatgtatgctttatttcatttaacttctaagatgcacttcatt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>gene-SINT2_LOCUS5354</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>30466976</td>\n",
       "      <td>30476049</td>\n",
       "      <td>acagaataaaagaggtgtactgttaacaacagccagagtgcttttt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gene-SINT2_LOCUS3201</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103877.1</td>\n",
       "      <td>38092074</td>\n",
       "      <td>38115003</td>\n",
       "      <td>tctggttgtactggcggatgacttagaggtaggattttgttttcct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>gene-SINT2_LOCUS4930</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>12765221</td>\n",
       "      <td>12771890</td>\n",
       "      <td>tcataaatgtaaccgtttagcacatgtaggtggtaatgacggtgat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>gene-SINT2_LOCUS5557</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>37810964</td>\n",
       "      <td>37839258</td>\n",
       "      <td>gtttaagaagttgcgccttcttcagatcttagatggtattaacatt...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene      organism   record_id     start       end  \\\n",
       "218  gene-SINT2_LOCUS1809  intercalatum  OX103876.1  69497234  69537522   \n",
       "66   gene-SINT2_LOCUS5354  intercalatum  OX103879.1  30466976  30476049   \n",
       "9    gene-SINT2_LOCUS3201  intercalatum  OX103877.1  38092074  38115003   \n",
       "170  gene-SINT2_LOCUS4930  intercalatum  OX103879.1  12765221  12771890   \n",
       "15   gene-SINT2_LOCUS5557  intercalatum  OX103879.1  37810964  37839258   \n",
       "\n",
       "                                                  data  \n",
       "218  atggatgtatgctttatttcatttaacttctaagatgcacttcatt...  \n",
       "66   acagaataaaagaggtgtactgttaacaacagccagagtgcttttt...  \n",
       "9    tctggttgtactggcggatgacttagaggtaggattttgttttcct...  \n",
       "170  tcataaatgtaaccgtttagcacatgtaggtggtaatgacggtgat...  \n",
       "15   gtttaagaagttgcgccttcttcagatcttagatggtattaacatt...  "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Добавим в табличку последовательность гена\n",
    "\n",
    "import pandas as pd\n",
    "import logging\n",
    "from Bio import SeqIO\n",
    "\n",
    "record_to_position = dict()\n",
    "\n",
    "for _, row in epigenetic_genes_for_alignment.iterrows():\n",
    "    if row['record_id'] not in record_to_position:\n",
    "        record_to_position[row['record_id']] = dict()\n",
    "    record_to_position[row['record_id']][(row['start'], row['end'])] = None\n",
    "\n",
    "record_to_position\n",
    "\n",
    "for organism in data_config:\n",
    "    if organism in faa_files:\n",
    "        records = list(SeqIO.parse(fna_files[organism], \"fasta\"))\n",
    "        for record in records:\n",
    "            if record.id in record_to_position:\n",
    "                for requested_range in record_to_position[record.id]:\n",
    "                    record_to_position[record.id][requested_range] = record.seq[requested_range[0]-1:requested_range[1]]\n",
    "\n",
    "    else:\n",
    "        logging.warning(f'No such fna file for {organism}')\n",
    "\n",
    "def get_sequence(row):\n",
    "    return [record_to_position[row['record_id']][(row['start'], row['end'])]]\n",
    "        \n",
    "epigenetic_genes_for_alignment[['data']] = \\\n",
    "    epigenetic_genes_for_alignment.apply(get_sequence, axis=1, result_type='expand')\n",
    "\n",
    "epigenetic_genes_for_alignment['data'] = epigenetic_genes_for_alignment['data'].apply(lambda s : str(s).lower())\n",
    "epigenetic_genes_for_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0609546e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/15/2025 03:26:19 PM\n",
      "Alignment settings:\n",
      "                          : Pairwise Alignment                  : ====================\n",
      "                          : DNAPWGapOpeningPenalty              : 15.00\n",
      "                          : DNAPWGapExtensionPenalty            : 6.66\n",
      "                          : Multiple Alignment                  : ====================\n",
      "                          : DNAMAGapOpeningPenalty              : 15.00\n",
      "                          : DNAMAGapExtensionPenalty            : 6.66\n",
      "                          : Global Options                      : ====================\n",
      "                          : DNA Weight Matrix                   : IUB\n",
      "                          : TransitionWeightNEdit               : 0.50\n",
      "                          : UseNegativeMatrix                   : ON\n",
      "                          : DelayDivergentCutoff                : 30\n",
      "                          : KeepPredefinedGaps                  : True\n",
      "                          : datatype                            : snNucleotide\n",
      "                          : containsCodingNuc                   : False\n",
      "                          : missingBaseSymbol                   : ?\n",
      "                          : identicalBaseSymbol                 : .\n",
      "                          : gapSymbol                           : -\n",
      " \n",
      "Executing pairwise alignment\n",
      ".\n",
      "\n",
      "Executing multiple sequence alignment\n",
      ".\n",
      "\n",
      "       100% Analysis Complete                                                               \n",
      " \n",
      "MEGA has completed the requested action\n",
      "\n",
      "Terminating the megacc process with successful exit status\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# запишем последовательности в fasta файл\n",
    "with open(\"data/epigenetic_genes_for_alignment.fasta\", \"w\") as fasta_file:\n",
    "    for _, row in epigenetic_genes_for_alignment.iterrows():\n",
    "        seq_id, seq = f'{row['organism']}:{row['gene']}', row['data']\n",
    "        fasta_file.write(f\">{seq_id}\\n{seq}\\n\")\n",
    "\n",
    "\n",
    "# настройки меги для выравнивания\n",
    "with open(\"data/muscle_align.mao\", \"w\") as f:\n",
    "    f.write(\"\"\"; Please do not edit this file! If this file is modified, results are unpredictable.\n",
    "; Instead of modifying this file, simply create a new MEGA Analysis Options file by using the MEGA Prototyper.\n",
    "[ DataSettings ]\n",
    "datatype                 = snNucleotide            \n",
    "containsCodingNuc        = False                   \n",
    "missingBaseSymbol        = ?                       \n",
    "identicalBaseSymbol      = .                       \n",
    "gapSymbol                = -                       \n",
    "[ ProcessTypes ]\n",
    "ppAlign                  = true                    \n",
    "ppClustalW               = true                    \n",
    "[ AnalysisSettings ]\n",
    "Pairwise Alignment       = ====================    \n",
    "DNAPWGapOpeningPenalty   = 15.00                   \n",
    "DNAPWGapExtensionPenalty = 6.66                    \n",
    "Multiple Alignment       = ====================    \n",
    "DNAMAGapOpeningPenalty   = 15.00                   \n",
    "DNAMAGapExtensionPenalty = 6.66                    \n",
    "Global Options           = ====================    \n",
    "DNA Weight Matrix        = IUB                     \n",
    "TransitionWeightNEdit    = 0.50                    \n",
    "UseNegativeMatrix        = ON                      \n",
    "DelayDivergentCutoff     = 30                      \n",
    "KeepPredefinedGaps       = True\"\"\")\n",
    "\n",
    "# запуск меги\n",
    "! ./local_programs/mega/megacc -a data/muscle_align.mao -d data/epigenetic_genes_for_alignment.fasta -o data/epigenetic_genes_alignment.meg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df453ec4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/15/2025 05:00:02 PM\n",
      "\n",
      "Data file                                         : epigenetic_genes_alignment.meg\n",
      "Status                                            : Reading header\n",
      "Current Progress                                  : 1% \n",
      "Status                                            : Setting up genes/domains information                                      MEGA-CC has logged the following warning:\n",
      "  When         = 06/15/2025 05:00:02 PM\n",
      "  Data file    = data/epigenetic_genes_alignment.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree.nwk\n",
      "  Message      = One or more domains not specified as having coding data were found but the mao file calls for coding data. All domains will be treated as having coding data\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "Status                                            : Organizing sequence informationMEGA-CC has logged the following warning:\n",
      "  When         = 06/15/2025 05:00:02 PM\n",
      "  Data file    = data/epigenetic_genes_alignment.meg\n",
      "  AnalysisFile = data/nj_tree.mao\n",
      "  Parameter    = -o\n",
      "  Parameter    = data/epigenetic_genes_tree.nwk\n",
      "  Message      = No Code table was specified, assuming the Genetic Code Table is 'Standard'\n",
      "\n",
      "  Please see the summary file for warnings/messages\n",
      "  Please validate all your inputs and try again. If you think this is a\n",
      "  bug, please report it at http://www.megasoftware.net\n",
      "\n",
      "\n",
      "No. of Taxa                                       : 5\n",
      "Using the following analysis options:\n",
      "No. of Taxa=5\n",
      "Analysis=Phylogeny Reconstruction\n",
      "\n",
      "Statistical Method             : Maximum Likelihood\n",
      "Test of Phylogeny              : Bootstrap method\n",
      "No. of Bootstrap Replications  : 500\n",
      "Substitutions Type             : Nucleotide\n",
      "Model/Method                   : Kimura 2-parameter model\n",
      "Rates among Sites              : Uniform Rates\n",
      "Gaps/Missing Data Treatment    : Complete deletion\n",
      "Select Codon Positions         : 1st, 2nd, 3rd, Non-Coding\n",
      "ML Heuristic Method            : Nearest-Neighbor-Interchange (NNI)\n",
      "Initial Tree for ML            : Make initial tree automatically (Neighbor Joining)\n",
      "Branch Swap Filter             : Strong\n",
      "Number of Threads              : 1\n",
      "Has Time Limit                 : False\n",
      "Maximum Execution Time         : -1\n",
      "datatype                       : snNucleotide\n",
      "containsCodingNuc              : True\n",
      "MissingBaseSymbol              : ?\n",
      "IdenticalBaseSymbol            : .\n",
      "GapSymbol                      : -Progress and Status\n",
      "\n",
      "Status                         : Preparing data\n",
      "Current Progress               : 99%                                       \n",
      "No. of Seqs                    : 5\n",
      "No. of Sites                   : 5337\n",
      "Current Progress               : 25%                                       \n",
      "Status                         : Bootstrapping ML tree\n",
      "Replicate No.                  : 5 of 500                                      \n",
      "Current Progress               : 1% \n",
      "Replicate No.                  : 10 of 500                                      \n",
      "Current Progress               : 2% \n",
      "Replicate No.                  : 15 of 500                                      \n",
      "Current Progress               : 3% \n",
      "Replicate No.                  : 20 of 500                                      \n",
      "Current Progress               : 4% \n",
      "Replicate No.                  : 25 of 500                                      \n",
      "Current Progress               : 5% \n",
      "Replicate No.                  : 30 of 500                                      \n",
      "Current Progress               : 6% \n",
      "Replicate No.                  : 35 of 500                                      \n",
      "Current Progress               : 7% \n",
      "Replicate No.                  : 40 of 500                                      \n",
      "Current Progress               : 8% \n",
      "Replicate No.                  : 45 of 500                                      \n",
      "Current Progress               : 9% \n",
      "Replicate No.                  : 50 of 500                                      \n",
      "Current Progress               : 10% \n",
      "Replicate No.                  : 55 of 500                                      \n",
      "Current Progress               : 11% \n",
      "Replicate No.                  : 60 of 500                                      \n",
      "Current Progress               : 12% \n",
      "Replicate No.                  : 65 of 500                                      \n",
      "Current Progress               : 13% \n",
      "Replicate No.                  : 70 of 500                                      \n",
      "Current Progress               : 14% \n",
      "Replicate No.                  : 75 of 500                                      \n",
      "Current Progress               : 15% \n",
      "Replicate No.                  : 80 of 500                                      \n",
      "Current Progress               : 16% \n",
      "Replicate No.                  : 85 of 500                                      \n",
      "Current Progress               : 17% \n",
      "Replicate No.                  : 90 of 500                                      \n",
      "Current Progress               : 18% \n",
      "Replicate No.                  : 95 of 500                                      \n",
      "Current Progress               : 19% \n",
      "Replicate No.                  : 100 of 500                                      \n",
      "Current Progress               : 20% \n",
      "Replicate No.                  : 105 of 500                                      \n",
      "Current Progress               : 21% \n",
      "Replicate No.                  : 110 of 500                                      \n",
      "Current Progress               : 22% \n",
      "Replicate No.                  : 115 of 500                                      \n",
      "Current Progress               : 23% \n",
      "Replicate No.                  : 120 of 500                                      \n",
      "Current Progress               : 24% \n",
      "Replicate No.                  : 125 of 500                                      \n",
      "Current Progress               : 25% \n",
      "Replicate No.                  : 130 of 500                                      \n",
      "Current Progress               : 26% \n",
      "Replicate No.                  : 135 of 500                                      \n",
      "Current Progress               : 27% \n",
      "Replicate No.                  : 140 of 500                                      \n",
      "Current Progress               : 28% \n",
      "Replicate No.                  : 146 of 500                                      \n",
      "Current Progress               : 29% \n",
      "Replicate No.                  : 150 of 500                                      \n",
      "Current Progress               : 30% \n",
      "Replicate No.                  : 155 of 500                                      \n",
      "Current Progress               : 31% \n",
      "Replicate No.                  : 160 of 500                                      \n",
      "Current Progress               : 32% \n",
      "Replicate No.                  : 165 of 500                                      \n",
      "Current Progress               : 33% \n",
      "Replicate No.                  : 170 of 500                                      \n",
      "Current Progress               : 34% \n",
      "Replicate No.                  : 175 of 500                                      \n",
      "Current Progress               : 35% \n",
      "Replicate No.                  : 180 of 500                                      \n",
      "Current Progress               : 36% \n",
      "Replicate No.                  : 185 of 500                                      \n",
      "Current Progress               : 37% \n",
      "Replicate No.                  : 190 of 500                                      \n",
      "Current Progress               : 38% \n",
      "Replicate No.                  : 195 of 500                                      \n",
      "Current Progress               : 39% \n",
      "Replicate No.                  : 200 of 500                                      \n",
      "Current Progress               : 40% \n",
      "Replicate No.                  : 205 of 500                                      \n",
      "Current Progress               : 41% \n",
      "Replicate No.                  : 210 of 500                                      \n",
      "Current Progress               : 42% \n",
      "Replicate No.                  : 215 of 500                                      \n",
      "Current Progress               : 43% \n",
      "Replicate No.                  : 220 of 500                                      \n",
      "Current Progress               : 44% \n",
      "Replicate No.                  : 225 of 500                                      \n",
      "Current Progress               : 45% \n",
      "Replicate No.                  : 230 of 500                                      \n",
      "Current Progress               : 46% \n",
      "Replicate No.                  : 235 of 500                                      \n",
      "Current Progress               : 47% \n",
      "Replicate No.                  : 240 of 500                                      \n",
      "Current Progress               : 48% \n",
      "Replicate No.                  : 245 of 500                                      \n",
      "Current Progress               : 49% \n",
      "Replicate No.                  : 250 of 500                                      \n",
      "Current Progress               : 50% \n",
      "Replicate No.                  : 255 of 500                                      \n",
      "Current Progress               : 51% \n",
      "Replicate No.                  : 260 of 500                                      \n",
      "Current Progress               : 52% \n",
      "Replicate No.                  : 265 of 500                                      \n",
      "Current Progress               : 53% \n",
      "Replicate No.                  : 270 of 500                                      \n",
      "Current Progress               : 54% \n",
      "Replicate No.                  : 275 of 500                                      \n",
      "Current Progress               : 55% \n",
      "Replicate No.                  : 280 of 500                                      \n",
      "Current Progress               : 56% \n",
      "Replicate No.                  : 286 of 500                                      \n",
      "Current Progress               : 57% \n",
      "Replicate No.                  : 291 of 500                                      \n",
      "Current Progress               : 58% \n",
      "Replicate No.                  : 295 of 500                                      \n",
      "Current Progress               : 59% \n",
      "Replicate No.                  : 300 of 500                                      \n",
      "Current Progress               : 60% \n",
      "Replicate No.                  : 305 of 500                                      \n",
      "Current Progress               : 61% \n",
      "Replicate No.                  : 310 of 500                                      \n",
      "Current Progress               : 62% \n",
      "Replicate No.                  : 315 of 500                                      \n",
      "Current Progress               : 63% \n",
      "Replicate No.                  : 320 of 500                                      \n",
      "Current Progress               : 64% \n",
      "Replicate No.                  : 325 of 500                                      \n",
      "Current Progress               : 65% \n",
      "Replicate No.                  : 330 of 500                                      \n",
      "Current Progress               : 66% \n",
      "Replicate No.                  : 335 of 500                                      \n",
      "Current Progress               : 67% \n",
      "Replicate No.                  : 340 of 500                                      \n",
      "Current Progress               : 68% \n",
      "Replicate No.                  : 345 of 500                                      \n",
      "Current Progress               : 69% \n",
      "Replicate No.                  : 350 of 500                                      \n",
      "Current Progress               : 70% \n",
      "Replicate No.                  : 355 of 500                                      \n",
      "Current Progress               : 71% \n",
      "Replicate No.                  : 360 of 500                                      \n",
      "Current Progress               : 72% \n",
      "Replicate No.                  : 365 of 500                                      \n",
      "Current Progress               : 73% \n",
      "Replicate No.                  : 370 of 500                                      \n",
      "Current Progress               : 74% \n",
      "Replicate No.                  : 375 of 500                                      \n",
      "Current Progress               : 75% \n",
      "Replicate No.                  : 380 of 500                                      \n",
      "Current Progress               : 76% \n",
      "Replicate No.                  : 385 of 500                                      \n",
      "Current Progress               : 77% \n",
      "Replicate No.                  : 390 of 500                                      \n",
      "Current Progress               : 78% \n",
      "Replicate No.                  : 395 of 500                                      \n",
      "Current Progress               : 79% \n",
      "Replicate No.                  : 400 of 500                                      \n",
      "Current Progress               : 80% \n",
      "Replicate No.                  : 405 of 500                                      \n",
      "Current Progress               : 81% \n",
      "Replicate No.                  : 410 of 500                                      \n",
      "Current Progress               : 82% \n",
      "Replicate No.                  : 415 of 500                                      \n",
      "Current Progress               : 83% \n",
      "Replicate No.                  : 420 of 500                                      \n",
      "Current Progress               : 84% \n",
      "Replicate No.                  : 425 of 500                                      \n",
      "Current Progress               : 85% \n",
      "Replicate No.                  : 430 of 500                                      \n",
      "Current Progress               : 86% \n",
      "Replicate No.                  : 435 of 500                                      \n",
      "Current Progress               : 87% \n",
      "Replicate No.                  : 440 of 500                                      \n",
      "Current Progress               : 88% \n",
      "Replicate No.                  : 445 of 500                                      \n",
      "Current Progress               : 89% \n",
      "Replicate No.                  : 450 of 500                                      \n",
      "Current Progress               : 90% \n",
      "Replicate No.                  : 455 of 500                                      \n",
      "Current Progress               : 91% \n",
      "Replicate No.                  : 460 of 500                                      \n",
      "Current Progress               : 92% \n",
      "Replicate No.                  : 465 of 500                                      \n",
      "Current Progress               : 93% \n",
      "Replicate No.                  : 470 of 500                                      \n",
      "Current Progress               : 94% \n",
      "Replicate No.                  : 475 of 500                                      \n",
      "Current Progress               : 95% \n",
      "Replicate No.                  : 480 of 500                                      \n",
      "Current Progress               : 96% \n",
      "Replicate No.                  : 485 of 500                                      \n",
      "Current Progress               : 97% \n",
      "Replicate No.                  : 490 of 500                                      \n",
      "Current Progress               : 98% \n",
      "Replicate No.                  : 495 of 500                                      \n",
      "Current Progress               : 99% \n",
      "Replicate No.                  : 500 of 500                                      \n",
      "Current Progress               : 100% \n",
      "\n",
      "       100% Analysis Complete                                                               \n",
      " \n",
      "MEGA has completed the requested action\n",
      "\n",
      "Terminating the megacc process with successful exit status\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# настройки меги для построения дерева\n",
    "with open(\"data/nj_tree.mao\", \"w\") as f:\n",
    "    f.write(\"\"\"[ DataSettings ]\n",
    "datatype                             = snNucleotide                        \n",
    "containsCodingNuc                    = True                                \n",
    "MissingBaseSymbol                    = ?                                   \n",
    "IdenticalBaseSymbol                  = .                                   \n",
    "GapSymbol                            = -                                   \n",
    "[ ProcessTypes ]\n",
    "ppInfer                              = true                                \n",
    "ppML                                 = true                                \n",
    "[ AnalysisSettings ]\n",
    "Analysis                             = Phylogeny Reconstruction            \n",
    "Statistical Method                   = Maximum Likelihood                  \n",
    "Phylogeny Test                       = ====================                \n",
    "Test of Phylogeny                    = Bootstrap method                    \n",
    "No. of Bootstrap Replications        = 500                                 \n",
    "Substitution Model                   = ====================                \n",
    "Substitutions Type                   = Nucleotide                          \n",
    "Model/Method                         = Kimura 2-parameter model            \n",
    "Rates and Patterns                   = ====================                \n",
    "Rates among Sites                    = Uniform Rates                       \n",
    "No of Discrete Gamma Categories      = Not Applicable                      \n",
    "Data Subset to Use                   = ====================                \n",
    "Gaps/Missing Data Treatment          = Complete deletion                   \n",
    "Site Coverage Cutoff (%)             = Not Applicable                      \n",
    "Select Codon Positions               = 1st, 2nd, 3rd, Non-Coding           \n",
    "Tree Inference Options               = ====================                \n",
    "ML Heuristic Method                  = Nearest-Neighbor-Interchange (NNI)  \n",
    "Initial Tree for ML                  = Make initial tree automatically (Neighbor Joining)\n",
    "Branch Swap Filter                   = Strong                              \n",
    "System Resource Usage                = ====================                \n",
    "Number of Threads                    = 1                                   \n",
    "Genetic Code Table                   = Not Applicable                      \n",
    "Genetic Code                         = Not Applicable                      \n",
    "Has Time Limit                       = False                               \n",
    "Maximum Execution Time               = -1  \"\"\")\n",
    "\n",
    "# запуск меги\n",
    "! ./local_programs/mega/megacc -a data/nj_tree.mao -d data/epigenetic_genes_alignment.meg -o data/epigenetic_genes_tree.nwk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "1b82213b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                 FigTree v1.4.4, 2006-2018\n",
      "                  Tree Figure Drawing Tool\n",
      "                       Andrew Rambaut\n",
      "\n",
      "             Institute of Evolutionary Biology\n",
      "                  University of Edinburgh\n",
      "                     a.rambaut@ed.ac.uk\n",
      "\n",
      "                 http://tree.bio.ed.ac.uk/\n",
      "    Uses the Java Evolutionary Biology 2 Library (JEBL2)\n",
      "                http://jebl2.googlecode.com/\n",
      "                 Uses the iText PDF Library\n",
      "                    http://itextpdf.com/\n",
      "               Uses the Apache Batik Library\n",
      "            http://xmlgraphics.apache.org/batik/\n",
      "                 Uses the JDOM XML Library\n",
      "                    http://www.jdom.org/\n",
      " Thanks to Alexei Drummond, Joseph Heled, Philippe Lemey, \n",
      "Tulio de Oliveira, Oliver Pybus, Beth Shapiro & Marc Suchard\n",
      "\n",
      "Creating SVG graphic: data/epigenetic_genes_tree.svg\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" style=\"fill-opacity:1; color-rendering:auto; color-interpolation:auto; text-rendering:auto; stroke:black; stroke-linecap:square; stroke-miterlimit:10; shape-rendering:auto; stroke-opacity:1; fill:black; stroke-dasharray:none; font-weight:normal; stroke-width:1; font-family:'Dialog'; font-style:normal; stroke-linejoin:miter; font-size:12; stroke-dashoffset:0; image-rendering:auto;\" width=\"800.0\" viewBox=\"0 0 800.0 600.0\" height=\"600.0\"><!--Generated by the Batik Graphics2D SVG Generator--><defs id=\"genericDefs\"/><g><g transform=\"translate(6,6)\" style=\"text-rendering:optimizeLegibility; font-size:10; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><line y2=\"570\" style=\"fill:none;\" x1=\"258.634\" x2=\"382.8718\" y1=\"570\"/><text x=\"313.8029\" xml:space=\"preserve\" y=\"583\" style=\"stroke:none;\">0.2</text><path style=\"fill:none;\" d=\"M641.5059 141.25 L189.3101 141.25 L189.3101 209.375\"/><path style=\"fill:none;\" d=\"M589.4637 413.75 L104.3934 413.75 L104.3934 311.5625\"/><path style=\"fill:none;\" d=\"M623.1093 550 L6.3515 550 L6.3515 288.8542\"/><path style=\"fill:none;\" d=\"M189.3101 209.375 L104.3934 209.375 L104.3934 311.5625\"/><path style=\"fill:none;\" d=\"M104.3934 311.5625 L6.3515 311.5625 L6.3515 288.8542\"/><path style=\"fill:none;\" d=\"M455.9677 5 L6.3515 5 L6.3515 288.8542\"/><path style=\"fill:none;\" d=\"M368.0565 277.5 L189.3101 277.5 L189.3101 209.375\"/><path style=\"fill:none;\" d=\"M0 288.8542 L6.3515 288.8542\"/><text x=\"0\" y=\"6\" transform=\"matrix(1,0,0,1,651.5059,136.25)\" style=\"font-size:8; stroke:none;\" xml:space=\"preserve\">intercalatum:gene-SINT2 LOCUS5354</text></g><g transform=\"matrix(1,0,0,1,605.4637,414.75)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS3201</text></g><g transform=\"matrix(1,0,0,1,639.1092,551)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS4930</text></g><g transform=\"matrix(1,0,0,1,471.9678,6)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS1809</text></g><g transform=\"matrix(1,0,0,1,384.0566,278.5)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS5557</text></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import SVG, display\n",
    "\n",
    "# строим визуализацию дерева\n",
    "! figtree -graphic SVG data/epigenetic_genes_tree.nwk data/epigenetic_genes_tree.svg\n",
    "display(SVG(filename='data/epigenetic_genes_tree.svg'))\n",
    "\n",
    "# более красивые картинки в отчете (тоже figtree, только через gui с настройками)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef9885c4",
   "metadata": {},
   "source": [
    "### 2. От каждого организма в группе выбрать 4-5 (как минимум 1 на члена группы)  генов (ортологов) в промоторах которых есть  квадруплексы и для этих квадруплексов построить выравнивание. И указать где есть вариации (в петле, числе повторов, выпетливания и т.п.) Если таких генов меньше, то для всех генов, что нашлись."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f4802f1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
