{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "007494cd",
   "metadata": {},
   "source": [
    "# Командная часть проекта"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 593,
   "id": "d67a6cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_config = {\n",
    "    'spindalis' : {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'g4_file' : 'g4_results.bed',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/946/903/255/GCA_946903255.1_tdSchSpin1.1b/GCA_946903255.1_tdSchSpin1.1b_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/946/903/255/GCA_946903255.1_tdSchSpin1.1b/GCA_946903255.1_tdSchSpin1.1b_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/946/903/255/GCA_946903255.1_tdSchSpin1.1b/GCA_946903255.1_tdSchSpin1.1b_protein.faa.gz'\n",
    "    },\n",
    "    'japonicum' : {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/021/461/655/GCA_021461655.1_ASM2146165v1/GCA_021461655.1_ASM2146165v1_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/021/461/655/GCA_021461655.1_ASM2146165v1/GCA_021461655.1_ASM2146165v1_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/021/461/655/GCA_021461655.1_ASM2146165v1/GCA_021461655.1_ASM2146165v1_protein.faa.gz'\n",
    "    },\n",
    "    'bovis' : {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'g4_file' : 'quadruplex.bed',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/425/GCA_944470425.2_tdSchBovi1.2/GCA_944470425.2_tdSchBovi1.2_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/425/GCA_944470425.2_tdSchBovi1.2/GCA_944470425.2_tdSchBovi1.2_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/425/GCA_944470425.2_tdSchBovi1.2/GCA_944470425.2_tdSchBovi1.2_protein.faa.gz'\n",
    "    },\n",
    "    'mattheei' : {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'g4_file' : 'quadruplex.bed',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/405/GCA_944470405.2_tdSchMatt1.2/GCA_944470405.2_tdSchMatt1.2_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/405/GCA_944470405.2_tdSchMatt1.2/GCA_944470405.2_tdSchMatt1.2_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/405/GCA_944470405.2_tdSchMatt1.2/GCA_944470405.2_tdSchMatt1.2_protein.faa.gz'\n",
    "    },\n",
    "    'haematobium' : {\n",
    "        # 'epigenetic_file' : 'epigenetic_genes.tsv',\n",
    "        'gff_url' : 'https://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS19/species/schistosoma_haematobium/PRJNA78265/schistosoma_haematobium.PRJNA78265.WBPS19.annotations.gff3.gz',\n",
    "        'fna_url' : 'https://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS19/species/schistosoma_haematobium/PRJNA78265/schistosoma_haematobium.PRJNA78265.WBPS19.genomic.fa.gz',\n",
    "        'proteom_url' : 'https://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS19/species/schistosoma_haematobium/PRJNA78265/schistosoma_haematobium.PRJNA78265.WBPS19.protein.fa.gz'\n",
    "    },\n",
    "    'mansoni' : {\n",
    "        'g4_file' : 'g4_results.bed',\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/237/925/GCF_000237925.1_ASM23792v2/GCF_000237925.1_ASM23792v2_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/237/925/GCF_000237925.1_ASM23792v2/GCF_000237925.1_ASM23792v2_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/237/925/GCF_000237925.1_ASM23792v2/GCF_000237925.1_ASM23792v2_protein.faa.gz'\n",
    "    },\n",
    "    'intercalatum': {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'g4_file' : 'g4_segments.csv',\n",
    "        'zdna_file' : 'zdna_segments.csv',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/385/GCA_944470385.2_tdSchInte2.2/GCA_944470385.2_tdSchInte2.2_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/385/GCA_944470385.2_tdSchInte2.2/GCA_944470385.2_tdSchInte2.2_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/470/385/GCA_944470385.2_tdSchInte2.2/GCA_944470385.2_tdSchInte2.2_protein.faa.gz'\n",
    "    },\n",
    "    'mekongi' : {\n",
    "        'epigenetic_file' : 'epigenetic_genes.csv',\n",
    "        'g4_file' : 'quadruplex.bed',\n",
    "        'zdna_file' : 'zdnabert.bed',\n",
    "        'gff_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/034/768/735/GCA_034768735.1_ASM3476873v1/GCA_034768735.1_ASM3476873v1_genomic.gff.gz',\n",
    "        'fna_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/034/768/735/GCA_034768735.1_ASM3476873v1/GCA_034768735.1_ASM3476873v1_genomic.fna.gz',\n",
    "        'proteom_url' : 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/034/768/735/GCA_034768735.1_ASM3476873v1/GCA_034768735.1_ASM3476873v1_protein.faa.gz'\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 481,
   "id": "886783cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:file data/GCA_946903255.1_tdSchSpin1.1b_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_946903255.1_tdSchSpin1.1b_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_946903255.1_tdSchSpin1.1b_protein.faa.gz already exists\n",
      "INFO:root:file data/GCA_021461655.1_ASM2146165v1_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_021461655.1_ASM2146165v1_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_021461655.1_ASM2146165v1_protein.faa.gz already exists\n",
      "INFO:root:file data/GCA_944470425.2_tdSchBovi1.2_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_944470425.2_tdSchBovi1.2_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_944470425.2_tdSchBovi1.2_protein.faa.gz already exists\n",
      "INFO:root:file data/GCA_944470405.2_tdSchMatt1.2_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_944470405.2_tdSchMatt1.2_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_944470405.2_tdSchMatt1.2_protein.faa.gz already exists\n",
      "INFO:root:file data/schistosoma_haematobium.PRJNA78265.WBPS19.annotations.gff3.gz already exists\n",
      "INFO:root:file data/schistosoma_haematobium.PRJNA78265.WBPS19.genomic.fa.gz already exists\n",
      "INFO:root:file data/schistosoma_haematobium.PRJNA78265.WBPS19.protein.fa.gz already exists\n",
      "INFO:root:file data/GCF_000237925.1_ASM23792v2_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCF_000237925.1_ASM23792v2_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCF_000237925.1_ASM23792v2_protein.faa.gz already exists\n",
      "INFO:root:file data/GCA_944470385.2_tdSchInte2.2_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_944470385.2_tdSchInte2.2_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_944470385.2_tdSchInte2.2_protein.faa.gz already exists\n",
      "INFO:root:file data/GCA_034768735.1_ASM3476873v1_genomic.gff.gz already exists\n",
      "INFO:root:file data/GCA_034768735.1_ASM3476873v1_genomic.fna.gz already exists\n",
      "INFO:root:file data/GCA_034768735.1_ASM3476873v1_protein.faa.gz already exists\n",
      "\n",
      "INFO:root:data/GCA_946903255.1_tdSchSpin1.1b_genomic.gff already exists\n",
      "INFO:root:data/GCA_021461655.1_ASM2146165v1_genomic.gff already exists\n",
      "INFO:root:data/GCA_944470425.2_tdSchBovi1.2_genomic.gff already exists\n",
      "INFO:root:data/GCA_944470405.2_tdSchMatt1.2_genomic.gff already exists\n",
      "INFO:root:data/schistosoma_haematobium.PRJNA78265.WBPS19.annotations.gff3 already exists\n",
      "INFO:root:data/GCF_000237925.1_ASM23792v2_genomic.gff already exists\n",
      "INFO:root:data/GCA_944470385.2_tdSchInte2.2_genomic.gff already exists\n",
      "INFO:root:data/GCA_034768735.1_ASM3476873v1_genomic.gff already exists\n",
      "INFO:root:data/GCA_946903255.1_tdSchSpin1.1b_genomic.fna already exists\n",
      "INFO:root:data/GCA_021461655.1_ASM2146165v1_genomic.fna already exists\n",
      "INFO:root:data/GCA_944470425.2_tdSchBovi1.2_genomic.fna already exists\n",
      "INFO:root:data/GCA_944470405.2_tdSchMatt1.2_genomic.fna already exists\n",
      "INFO:root:data/schistosoma_haematobium.PRJNA78265.WBPS19.genomic.fa already exists\n",
      "INFO:root:data/GCF_000237925.1_ASM23792v2_genomic.fna already exists\n",
      "INFO:root:data/GCA_944470385.2_tdSchInte2.2_genomic.fna already exists\n",
      "INFO:root:data/GCA_034768735.1_ASM3476873v1_genomic.fna already exists\n",
      "INFO:root:data/GCA_946903255.1_tdSchSpin1.1b_protein.faa already exists\n",
      "INFO:root:data/GCA_021461655.1_ASM2146165v1_protein.faa already exists\n",
      "INFO:root:data/GCA_944470425.2_tdSchBovi1.2_protein.faa already exists\n",
      "INFO:root:data/GCA_944470405.2_tdSchMatt1.2_protein.faa already exists\n",
      "INFO:root:data/schistosoma_haematobium.PRJNA78265.WBPS19.protein.fa already exists\n",
      "INFO:root:data/GCF_000237925.1_ASM23792v2_protein.faa already exists\n",
      "INFO:root:data/GCA_944470385.2_tdSchInte2.2_protein.faa already exists\n",
      "INFO:root:data/GCA_034768735.1_ASM3476873v1_protein.faa already exists\n"
     ]
    }
   ],
   "source": [
    "# скачаем аннотации\n",
    "\n",
    "# скачивание файлов генома и аннотации\n",
    "\n",
    "import requests\n",
    "import logging\n",
    "import pathlib\n",
    "import glob\n",
    "import os\n",
    "import sys\n",
    "import itertools\n",
    "\n",
    "logging.basicConfig(level=logging.INFO, force = True)\n",
    "\n",
    "def download_file(url: str, local_file_name: pathlib.Path):\n",
    "    if pathlib.Path(local_file_name).is_file():\n",
    "        logging.info(f'file {local_file_name} already exists')\n",
    "        return local_file_name\n",
    "    logging.info(f'downloading {local_file_name}')\n",
    "    with requests.get(url, stream=True) as r:\n",
    "        r.raise_for_status()\n",
    "        with open(local_file_name, 'wb') as out_file:\n",
    "            for chunk in r.iter_content(chunk_size=8192): \n",
    "                out_file.write(chunk)\n",
    "\n",
    "    return local_file_name\n",
    "\n",
    "gff_files = dict()\n",
    "fna_files = dict()\n",
    "proteom_files = dict()\n",
    "\n",
    "# скачиваем аннотации\n",
    "for current_organism, current_data_config in data_config.items():\n",
    "    if 'gff_url' in current_data_config:\n",
    "        gff_files[current_organism] = download_file(current_data_config['gff_url'], f'data/{current_data_config['gff_url'].split('/')[-1]}')\n",
    "    else:\n",
    "        logging.warning(f'No gff url provided for {current_organism}')\n",
    "    \n",
    "    if 'fna_url' in current_data_config:\n",
    "        fna_files[current_organism] = download_file(current_data_config['fna_url'], f'data/{current_data_config['fna_url'].split('/')[-1]}')\n",
    "    else:\n",
    "        logging.warning(f'No fna url provided for {current_organism}')\n",
    "\n",
    "    if 'proteom_url' in current_data_config:\n",
    "        proteom_files[current_organism] = download_file(current_data_config['proteom_url'], f'data/{current_data_config['proteom_url'].split('/')[-1]}')\n",
    "    else:\n",
    "        logging.warning(f'No proteom url provided for {current_organism}')\n",
    "\n",
    "print(file=sys.stderr)\n",
    "\n",
    "# распакуем сжатые файлы для удобства\n",
    "for organism, filepath in itertools.chain(gff_files.items(), fna_files.items(), proteom_files.items()):\n",
    "    current_file_path = pathlib.Path(filepath)\n",
    "    assert current_file_path.is_file()\n",
    "    if current_file_path.suffix == '.gz':\n",
    "        current_extracted_file = current_file_path.parent / pathlib.Path(filepath).stem\n",
    "        if current_extracted_file.is_file():\n",
    "            logging.info(f'{current_extracted_file} already exists')\n",
    "        else:\n",
    "            logging.info(f'extracting {current_extracted_file}')\n",
    "            os.system(f'gunzip -k {current_extracted_file}')\n",
    "\n",
    "for organism, filepath in gff_files.items():\n",
    "    if pathlib.Path(filepath).suffix == '.gz':\n",
    "        gff_files[organism] = current_file_path.parent / pathlib.Path(filepath).stem\n",
    "\n",
    "for organism, filepath in fna_files.items():\n",
    "    if pathlib.Path(filepath).suffix == '.gz':\n",
    "        fna_files[organism] = current_file_path.parent / pathlib.Path(filepath).stem\n",
    "\n",
    "for organism, filepath in proteom_files.items():\n",
    "    if pathlib.Path(filepath).suffix == '.gz':\n",
    "        proteom_files[organism] = current_file_path.parent / pathlib.Path(filepath).stem\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 819,
   "id": "d9e06e08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# функции для чтения данных из результатов индивидуальных проектов\n",
    "\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "\n",
    "def get_epigenetic_genes(organism: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "\n",
    "    if 'epigenetic_file' in data_config[organism]:\n",
    "        file_name = data_config[organism]['epigenetic_file']\n",
    "        if pathlib.Path(file_name).suffix == '.csv':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['epigenetic_file']}')\n",
    "            res = res[['Family', 'Gene_ID']]\n",
    "            res.columns = ['family', 'gene']\n",
    "            return res\n",
    "        elif pathlib.Path(file_name).suffix == '.tsv':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['epigenetic_file']}', sep='\\t')\n",
    "            res = res[['Family', 'Gene_ID']]\n",
    "            res.columns = ['family', 'gene']\n",
    "            return res\n",
    "        else:\n",
    "            assert False, \"Please, add support for your file type\" # TODO\n",
    "    else:\n",
    "        logging.warning(f'No epigenetic_file specified for {organism}')\n",
    "        return None\n",
    "\n",
    "def get_g4(organism: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "\n",
    "    if 'g4_file' in data_config[organism]:\n",
    "        file_name = data_config[organism]['g4_file']\n",
    "        if pathlib.Path(file_name).suffix == '.csv':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['g4_file']}')\n",
    "            return res\n",
    "        elif pathlib.Path(file_name).suffix == '.bed':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['g4_file']}', sep='\\t', header=None).iloc[:, :3]\n",
    "            res.columns=['record_id' ,'start', 'end']\n",
    "            return res\n",
    "        else:\n",
    "            assert False, \"Please, add support for your file type\" # TODO\n",
    "\n",
    "    else:\n",
    "        logging.warning(f'No g4_file specified for {organism}')\n",
    "        return None\n",
    "\n",
    "def get_zdna(organism: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "\n",
    "    if 'zdna_file' in data_config[organism]:\n",
    "        file_name = data_config[organism]['zdna_file']\n",
    "        if pathlib.Path(file_name).suffix == '.csv':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['zdna_file']}')\n",
    "            return res\n",
    "        elif pathlib.Path(file_name).suffix == '.bed':\n",
    "            res = pd.read_csv(f'data_in/{organism}/{data_config[organism]['zdna_file']}', sep='\\t', header=None).iloc[:, :3]\n",
    "            res.columns=['record_id' ,'start', 'end']\n",
    "            return res\n",
    "        else:\n",
    "            assert None, \"Please, add support for your file type\" # TODO\n",
    "\n",
    "    else:\n",
    "        logging.warning(f'No zdna_file specified for {organism}')\n",
    "        return None\n",
    "\n",
    "# need_contain one of 'g4' or 'zdna'\n",
    "def get_genes_with_promoters_contains(organism: str, need_contain: str):\n",
    "    assert organism in data_config, f'unknown organism: {organism}'\n",
    "    assert need_contain in ['g4', 'zdna']\n",
    "\n",
    "    segments = get_g4(organism) if need_contain == 'g4' else get_zdna(organism)\n",
    "    if segments is None:\n",
    "        return None\n",
    "\n",
    "    if organism in gff_files:\n",
    "        res = []\n",
    "        with open(gff_files[organism]) as gff_file:\n",
    "            for row in gff_file:\n",
    "                if row[0] != '#':\n",
    "                    row_splitted = row.strip().split()\n",
    "                    if row_splitted[2] == 'gene':\n",
    "                        gene_coords = (int(row_splitted[3]), int(row_splitted[4]))\n",
    "                        promoter_coords = (gene_coords[0] - 1000, gene_coords[0] - 1)\n",
    "                        if row_splitted[5] == '-':\n",
    "                            promoter_coords = (gene_coords[1] + 1, gene_coords[1] + 1000)\n",
    "\n",
    "                        # поиск квадруплексов или zdna\n",
    "                        found_segments = []\n",
    "                        for _, segment_row in segments.iterrows():\n",
    "                            if segment_row['record_id'] == row_splitted[0]:\n",
    "                                cur_interval = (segment_row['start'], segment_row['end'])\n",
    "                                if pd.Interval(*cur_interval, closed='both').overlaps(pd.Interval(*promoter_coords, closed='both')):\n",
    "                                    found_segments.append(cur_interval)\n",
    "                        \n",
    "                        if len(found_segments) != 0:\n",
    "                            gene_id = row_splitted[8].split('ID=')[1].split(';')[0]\n",
    "                            res.append([row_splitted[0], gene_id, promoter_coords[0], promoter_coords[1], found_segments])\n",
    "        res = pd.DataFrame(res)\n",
    "        res.columns = ['record_id', 'gene', 'start', 'end', 'matches']\n",
    "        return res\n",
    "    else:\n",
    "        logging.warning(f'No gff specified for {organism}')\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 484,
   "id": "71a4aad9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:local_programs/mega.tar.gz already exists\n",
      "INFO:root:already exists\n"
     ]
    }
   ],
   "source": [
    "# установка меги\n",
    "\n",
    "# https://drive.google.com/file/d/1beRdRYvyRw38orLmFiXE-p2paAILFvVm/view?usp=sharing\n",
    "\n",
    "\n",
    "mega_file_id = \"1beRdRYvyRw38orLmFiXE-p2paAILFvVm\"\n",
    "mega_url = f\"https://drive.google.com/uc?export=download&id={mega_file_id}\"\n",
    "\n",
    "# скачивание меги\n",
    "\n",
    "import requests\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "import logging\n",
    "\n",
    "mega_gz = pathlib.Path('local_programs/mega.tar.gz')\n",
    "mega_dir = pathlib.Path('local_programs/mega')\n",
    "\n",
    "if mega_gz.is_file():\n",
    "    logging.info(f'{mega_gz} already exists')\n",
    "else:\n",
    "    logging.info(f'downloading {mega_gz}')\n",
    "    response = requests.get(mega_url)\n",
    "    assert response.status_code == 200, 'Wrong status code'\n",
    "    with open(mega_gz, 'wb') as out_file:\n",
    "        out_file.write(response.content)\n",
    "\n",
    "if not mega_dir.is_dir():\n",
    "    logging.info(f'extracting {mega_gz}')\n",
    "    ! cd local_programs/ && mkdir -p mega && tar -xzf mega.tar.gz -C mega\n",
    "else:\n",
    "    logging.info(f'already exists')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "594378b3",
   "metadata": {},
   "source": [
    "### 1. От каждого организма в группе Выбрать из генов отвечающих за эпигенетику 4-5 (как минимум 1 на члена группы) и построить для них выравнивания и деревья."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 485,
   "id": "f6da3ebd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Adding genes for spindalis\n",
      "INFO:root:Adding genes for japonicum\n",
      "INFO:root:Adding genes for bovis\n",
      "INFO:root:Adding genes for mattheei\n",
      "WARNING:root:No epigenetic_file specified for haematobium\n",
      "WARNING:root:current_epigenetic_genes(haematobium) is None\n",
      "INFO:root:Adding genes for mansoni\n",
      "INFO:root:Adding genes for intercalatum\n",
      "INFO:root:Adding genes for mekongi\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family</th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SSPN1_LOCUS6014</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SMTH1_LOCUS6929</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-KSF78_0008119</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-Smp_039490.1</td>\n",
       "      <td>mansoni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SBOV1_LOCUS9112</td>\n",
       "      <td>bovis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-MN116_008246</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SINT2_LOCUS6163</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SINT2_LOCUS8012</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SBOV1_LOCUS9059</td>\n",
       "      <td>bovis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SSPN1_LOCUS9217</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SMTH1_LOCUS10615</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-KSF78_0005440</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-MN116_001510</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-Smp_155070</td>\n",
       "      <td>mansoni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-MN116_002201</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SMTH1_LOCUS7396</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-Smp_157150</td>\n",
       "      <td>mansoni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SBOV1_LOCUS6579</td>\n",
       "      <td>bovis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SINT2_LOCUS6554</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-KSF78_0007878</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SSPN1_LOCUS6463</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-MN116_002008</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SMTH1_LOCUS10083</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SINT2_LOCUS8890</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SSPN1_LOCUS8606</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-Smp_162340</td>\n",
       "      <td>mansoni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-KSF78_0004211</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SBOV1_LOCUS8938</td>\n",
       "      <td>bovis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SMTH1_LOCUS10333</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-Smp_163110</td>\n",
       "      <td>mansoni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SBOV1_LOCUS9010</td>\n",
       "      <td>bovis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-MN116_001315</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-KSF78_0005543</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SSPN1_LOCUS9160</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SINT2_LOCUS5228</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-Smp_168800.1</td>\n",
       "      <td>mansoni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SINT2_LOCUS956</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SMTH1_LOCUS4562</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-MN116_003045</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SBOV1_LOCUS4028</td>\n",
       "      <td>bovis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-KSF78_0005343</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SSPN1_LOCUS3856</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SINT2_LOCUS4989</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-MN116_001474</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-KSF78_0005473</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SSPN1_LOCUS8341</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-Smp_137090</td>\n",
       "      <td>mansoni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SBOV1_LOCUS9420</td>\n",
       "      <td>bovis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SMTH1_LOCUS10644</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-SSPN1_LOCUS9023</td>\n",
       "      <td>spindalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-Smp_078280</td>\n",
       "      <td>mansoni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-SBOV1_LOCUS9010</td>\n",
       "      <td>bovis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-SINT2_LOCUS6604</td>\n",
       "      <td>intercalatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-KSF78_0005240</td>\n",
       "      <td>japonicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-SMTH1_LOCUS10167</td>\n",
       "      <td>mattheei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-MN116_001223</td>\n",
       "      <td>mekongi</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    family                   gene      organism\n",
       "0    ACINU   gene-SSPN1_LOCUS6014     spindalis\n",
       "1    ACINU   gene-SMTH1_LOCUS6929      mattheei\n",
       "2    ACINU     gene-KSF78_0008119     japonicum\n",
       "3    ACINU      gene-Smp_039490.1       mansoni\n",
       "4    ACINU   gene-SBOV1_LOCUS9112         bovis\n",
       "5    ACINU      gene-MN116_008246       mekongi\n",
       "6    ACINU   gene-SINT2_LOCUS6163  intercalatum\n",
       "7    AEBP2   gene-SINT2_LOCUS8012  intercalatum\n",
       "8    AEBP2   gene-SBOV1_LOCUS9059         bovis\n",
       "9    AEBP2   gene-SSPN1_LOCUS9217     spindalis\n",
       "10   AEBP2  gene-SMTH1_LOCUS10615      mattheei\n",
       "11   AEBP2     gene-KSF78_0005440     japonicum\n",
       "12   AEBP2      gene-MN116_001510       mekongi\n",
       "13   AEBP2        gene-Smp_155070       mansoni\n",
       "14  ARID1A      gene-MN116_002201       mekongi\n",
       "15  ARID1A   gene-SMTH1_LOCUS7396      mattheei\n",
       "16  ARID1A        gene-Smp_157150       mansoni\n",
       "17  ARID1A   gene-SBOV1_LOCUS6579         bovis\n",
       "18  ARID1A   gene-SINT2_LOCUS6554  intercalatum\n",
       "19  ARID1A     gene-KSF78_0007878     japonicum\n",
       "20  ARID1A   gene-SSPN1_LOCUS6463     spindalis\n",
       "21     ATR      gene-MN116_002008       mekongi\n",
       "22     ATR  gene-SMTH1_LOCUS10083      mattheei\n",
       "23     ATR   gene-SINT2_LOCUS8890  intercalatum\n",
       "24     ATR   gene-SSPN1_LOCUS8606     spindalis\n",
       "25     ATR        gene-Smp_162340       mansoni\n",
       "26     ATR     gene-KSF78_0004211     japonicum\n",
       "27     ATR   gene-SBOV1_LOCUS8938         bovis\n",
       "28    ATRX  gene-SMTH1_LOCUS10333      mattheei\n",
       "29    ATRX        gene-Smp_163110       mansoni\n",
       "30    ATRX   gene-SBOV1_LOCUS9010         bovis\n",
       "31    ATRX      gene-MN116_001315       mekongi\n",
       "32    ATRX     gene-KSF78_0005543     japonicum\n",
       "33    ATRX   gene-SSPN1_LOCUS9160     spindalis\n",
       "34    ATRX   gene-SINT2_LOCUS5228  intercalatum\n",
       "35    BAP1      gene-Smp_168800.1       mansoni\n",
       "36    BAP1    gene-SINT2_LOCUS956  intercalatum\n",
       "37    BAP1   gene-SMTH1_LOCUS4562      mattheei\n",
       "38    BAP1      gene-MN116_003045       mekongi\n",
       "39    BAP1   gene-SBOV1_LOCUS4028         bovis\n",
       "40    BAP1     gene-KSF78_0005343     japonicum\n",
       "41    BAP1   gene-SSPN1_LOCUS3856     spindalis\n",
       "42   BARD1   gene-SINT2_LOCUS4989  intercalatum\n",
       "43   BARD1      gene-MN116_001474       mekongi\n",
       "44   BARD1     gene-KSF78_0005473     japonicum\n",
       "45   BARD1   gene-SSPN1_LOCUS8341     spindalis\n",
       "46   BARD1        gene-Smp_137090       mansoni\n",
       "47   BARD1   gene-SBOV1_LOCUS9420         bovis\n",
       "48   BARD1  gene-SMTH1_LOCUS10644      mattheei\n",
       "49    CBX4   gene-SSPN1_LOCUS9023     spindalis\n",
       "50    CBX4        gene-Smp_078280       mansoni\n",
       "51    CBX4   gene-SBOV1_LOCUS9010         bovis\n",
       "52    CBX4   gene-SINT2_LOCUS6604  intercalatum\n",
       "53    CBX4     gene-KSF78_0005240     japonicum\n",
       "54    CBX4  gene-SMTH1_LOCUS10167      mattheei\n",
       "55    CBX4      gene-MN116_001223       mekongi"
      ]
     },
     "execution_count": 485,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# выбираем гены для выравнивания\n",
    "\n",
    "import logging\n",
    "\n",
    "family_in_organisms = dict()\n",
    "epigenetic_genes_for_alignment = []\n",
    "\n",
    "for organism in data_config:\n",
    "    current_epigenetic_genes = get_epigenetic_genes(organism)\n",
    "    if current_epigenetic_genes is not None:\n",
    "        logging.info(f'Adding genes for {organism}')\n",
    "        \n",
    "        current_epigenetic_genes['organism'] = organism\n",
    "        epigenetic_genes_for_alignment.append(current_epigenetic_genes)\n",
    "\n",
    "        for _, row in current_epigenetic_genes.iterrows():\n",
    "            if row['family'] not in family_in_organisms:\n",
    "                family_in_organisms[row['family']] = set()\n",
    "            family_in_organisms[row['family']].add(organism)\n",
    "    else:\n",
    "        logging.warning(f'current_epigenetic_genes({organism}) is None')\n",
    "\n",
    "family_in_organisms_array = []\n",
    "for family, organisms in family_in_organisms.items():\n",
    "    family_in_organisms_array.append((family, len(organisms)))\n",
    "\n",
    "family_in_organisms_array = list(sorted(family_in_organisms_array, key=lambda x : -x[1]))\n",
    "family_in_organisms_array = family_in_organisms_array[:8]\n",
    "\n",
    "selected_families = [x[0] for x in family_in_organisms_array]\n",
    "\n",
    "epigenetic_genes_for_alignment = pd.concat(epigenetic_genes_for_alignment)\n",
    "epigenetic_genes_for_alignment = epigenetic_genes_for_alignment[epigenetic_genes_for_alignment['family'].apply(lambda x : x in selected_families)]\n",
    "epigenetic_genes_for_alignment.sort_values('family', inplace=True)\n",
    "epigenetic_genes_for_alignment.drop_duplicates(subset=['family', 'organism'], keep='first', inplace=True)\n",
    "epigenetic_genes_for_alignment.reset_index(inplace=True, drop=True)\n",
    "epigenetic_genes_for_alignment.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "id": "db368604",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family</th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "      <th>record_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SSPN1_LOCUS6014</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336419.1</td>\n",
       "      <td>20174797</td>\n",
       "      <td>20192511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SMTH1_LOCUS6929</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103946.1</td>\n",
       "      <td>19000643</td>\n",
       "      <td>19025157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-KSF78_0008119</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038286.1</td>\n",
       "      <td>12019241</td>\n",
       "      <td>12028457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-Smp_039490.1</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031499.1</td>\n",
       "      <td>6407653</td>\n",
       "      <td>6425032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SBOV1_LOCUS9112</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103961.1</td>\n",
       "      <td>72383764</td>\n",
       "      <td>72392806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-MN116_008246</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068311.1</td>\n",
       "      <td>9544071</td>\n",
       "      <td>9569391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SINT2_LOCUS6163</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103880.1</td>\n",
       "      <td>17878940</td>\n",
       "      <td>17899255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SINT2_LOCUS8012</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>28294149</td>\n",
       "      <td>28315298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SBOV1_LOCUS9059</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103961.1</td>\n",
       "      <td>69774510</td>\n",
       "      <td>69802550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SSPN1_LOCUS9217</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>86989145</td>\n",
       "      <td>86991956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SMTH1_LOCUS10615</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>85876995</td>\n",
       "      <td>85881366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-KSF78_0005440</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>42441718</td>\n",
       "      <td>42464061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-MN116_001510</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>41392895</td>\n",
       "      <td>41398756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-Smp_155070</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031502.1</td>\n",
       "      <td>55549177</td>\n",
       "      <td>55552968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-MN116_002201</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>71954612</td>\n",
       "      <td>71997418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SMTH1_LOCUS7396</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103948.1</td>\n",
       "      <td>12246346</td>\n",
       "      <td>12286004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-Smp_157150</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031502.1</td>\n",
       "      <td>37972738</td>\n",
       "      <td>37998717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SBOV1_LOCUS6579</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103958.1</td>\n",
       "      <td>15163266</td>\n",
       "      <td>15205964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SINT2_LOCUS6554</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103881.1</td>\n",
       "      <td>11733954</td>\n",
       "      <td>11780504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-KSF78_0007878</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038286.1</td>\n",
       "      <td>1806863</td>\n",
       "      <td>1849218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SSPN1_LOCUS6463</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336421.1</td>\n",
       "      <td>16572420</td>\n",
       "      <td>16601429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-MN116_002008</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>62124388</td>\n",
       "      <td>62196869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SMTH1_LOCUS10083</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>65486722</td>\n",
       "      <td>65509237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SINT2_LOCUS8890</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>61543600</td>\n",
       "      <td>61607504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SSPN1_LOCUS8606</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>63284548</td>\n",
       "      <td>63322369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-Smp_162340</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031502.1</td>\n",
       "      <td>45490939</td>\n",
       "      <td>45559573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-KSF78_0004211</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038289.1</td>\n",
       "      <td>78055480</td>\n",
       "      <td>78138330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SBOV1_LOCUS8938</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103961.1</td>\n",
       "      <td>64679616</td>\n",
       "      <td>64745603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SMTH1_LOCUS10333</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>75611087</td>\n",
       "      <td>75644170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-Smp_163110</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031502.1</td>\n",
       "      <td>53300151</td>\n",
       "      <td>53330817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SBOV1_LOCUS9010</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103961.1</td>\n",
       "      <td>67907733</td>\n",
       "      <td>67951675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-MN116_001315</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>34110633</td>\n",
       "      <td>34126820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-KSF78_0005543</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>47227496</td>\n",
       "      <td>47266275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SSPN1_LOCUS9160</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>84866275</td>\n",
       "      <td>84901726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SINT2_LOCUS5228</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>25744885</td>\n",
       "      <td>25784263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-Smp_168800.1</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031497.1</td>\n",
       "      <td>14428597</td>\n",
       "      <td>14430214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SINT2_LOCUS956</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103876.1</td>\n",
       "      <td>35851198</td>\n",
       "      <td>35858909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SMTH1_LOCUS4562</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103945.1</td>\n",
       "      <td>26934207</td>\n",
       "      <td>26937693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-MN116_003045</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068306.1</td>\n",
       "      <td>25384891</td>\n",
       "      <td>25413756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SBOV1_LOCUS4028</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103955.1</td>\n",
       "      <td>25583114</td>\n",
       "      <td>25589180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-KSF78_0005343</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>38289344</td>\n",
       "      <td>38308558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SSPN1_LOCUS3856</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336418.1</td>\n",
       "      <td>21839072</td>\n",
       "      <td>21850802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SINT2_LOCUS4989</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>15489966</td>\n",
       "      <td>15504708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-MN116_001474</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>40142102</td>\n",
       "      <td>40205167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-KSF78_0005473</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>43890028</td>\n",
       "      <td>43902920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SSPN1_LOCUS8341</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>50816278</td>\n",
       "      <td>50861849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-Smp_137090</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031498.1</td>\n",
       "      <td>6736171</td>\n",
       "      <td>6746795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SBOV1_LOCUS9420</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103961.1</td>\n",
       "      <td>85568270</td>\n",
       "      <td>85581490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SMTH1_LOCUS10644</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>86929274</td>\n",
       "      <td>86944028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-SSPN1_LOCUS9023</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>79888213</td>\n",
       "      <td>79917220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-Smp_078280</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031502.1</td>\n",
       "      <td>26206232</td>\n",
       "      <td>26232252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-SBOV1_LOCUS9010</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103961.1</td>\n",
       "      <td>67907733</td>\n",
       "      <td>67951675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-SINT2_LOCUS6604</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103881.1</td>\n",
       "      <td>13847789</td>\n",
       "      <td>13862796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-KSF78_0005240</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>33883412</td>\n",
       "      <td>33885793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-SMTH1_LOCUS10167</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>68689821</td>\n",
       "      <td>68736678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-MN116_001223</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>30297854</td>\n",
       "      <td>30329074</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    family                   gene      organism    record_id     start  \\\n",
       "0    ACINU   gene-SSPN1_LOCUS6014     spindalis   OX336419.1  20174797   \n",
       "1    ACINU   gene-SMTH1_LOCUS6929      mattheei   OX103946.1  19000643   \n",
       "2    ACINU     gene-KSF78_0008119     japonicum   CM038286.1  12019241   \n",
       "3    ACINU      gene-Smp_039490.1       mansoni  NC_031499.1   6407653   \n",
       "4    ACINU   gene-SBOV1_LOCUS9112         bovis   OX103961.1  72383764   \n",
       "5    ACINU      gene-MN116_008246       mekongi   CM068311.1   9544071   \n",
       "6    ACINU   gene-SINT2_LOCUS6163  intercalatum   OX103880.1  17878940   \n",
       "7    AEBP2   gene-SINT2_LOCUS8012  intercalatum   OX103883.1  28294149   \n",
       "8    AEBP2   gene-SBOV1_LOCUS9059         bovis   OX103961.1  69774510   \n",
       "9    AEBP2   gene-SSPN1_LOCUS9217     spindalis   OX336416.1  86989145   \n",
       "10   AEBP2  gene-SMTH1_LOCUS10615      mattheei   OX103943.1  85876995   \n",
       "11   AEBP2     gene-KSF78_0005440     japonicum   CM038283.1  42441718   \n",
       "12   AEBP2      gene-MN116_001510       mekongi   CM068305.1  41392895   \n",
       "13   AEBP2        gene-Smp_155070       mansoni  NC_031502.1  55549177   \n",
       "14  ARID1A      gene-MN116_002201       mekongi   CM068305.1  71954612   \n",
       "15  ARID1A   gene-SMTH1_LOCUS7396      mattheei   OX103948.1  12246346   \n",
       "16  ARID1A        gene-Smp_157150       mansoni  NC_031502.1  37972738   \n",
       "17  ARID1A   gene-SBOV1_LOCUS6579         bovis   OX103958.1  15163266   \n",
       "18  ARID1A   gene-SINT2_LOCUS6554  intercalatum   OX103881.1  11733954   \n",
       "19  ARID1A     gene-KSF78_0007878     japonicum   CM038286.1   1806863   \n",
       "20  ARID1A   gene-SSPN1_LOCUS6463     spindalis   OX336421.1  16572420   \n",
       "21     ATR      gene-MN116_002008       mekongi   CM068305.1  62124388   \n",
       "22     ATR  gene-SMTH1_LOCUS10083      mattheei   OX103943.1  65486722   \n",
       "23     ATR   gene-SINT2_LOCUS8890  intercalatum   OX103883.1  61543600   \n",
       "24     ATR   gene-SSPN1_LOCUS8606     spindalis   OX336416.1  63284548   \n",
       "25     ATR        gene-Smp_162340       mansoni  NC_031502.1  45490939   \n",
       "26     ATR     gene-KSF78_0004211     japonicum   CM038289.1  78055480   \n",
       "27     ATR   gene-SBOV1_LOCUS8938         bovis   OX103961.1  64679616   \n",
       "28    ATRX  gene-SMTH1_LOCUS10333      mattheei   OX103943.1  75611087   \n",
       "29    ATRX        gene-Smp_163110       mansoni  NC_031502.1  53300151   \n",
       "30    ATRX   gene-SBOV1_LOCUS9010         bovis   OX103961.1  67907733   \n",
       "31    ATRX      gene-MN116_001315       mekongi   CM068305.1  34110633   \n",
       "32    ATRX     gene-KSF78_0005543     japonicum   CM038283.1  47227496   \n",
       "33    ATRX   gene-SSPN1_LOCUS9160     spindalis   OX336416.1  84866275   \n",
       "34    ATRX   gene-SINT2_LOCUS5228  intercalatum   OX103879.1  25744885   \n",
       "35    BAP1      gene-Smp_168800.1       mansoni  NC_031497.1  14428597   \n",
       "36    BAP1    gene-SINT2_LOCUS956  intercalatum   OX103876.1  35851198   \n",
       "37    BAP1   gene-SMTH1_LOCUS4562      mattheei   OX103945.1  26934207   \n",
       "38    BAP1      gene-MN116_003045       mekongi   CM068306.1  25384891   \n",
       "39    BAP1   gene-SBOV1_LOCUS4028         bovis   OX103955.1  25583114   \n",
       "40    BAP1     gene-KSF78_0005343     japonicum   CM038283.1  38289344   \n",
       "41    BAP1   gene-SSPN1_LOCUS3856     spindalis   OX336418.1  21839072   \n",
       "42   BARD1   gene-SINT2_LOCUS4989  intercalatum   OX103879.1  15489966   \n",
       "43   BARD1      gene-MN116_001474       mekongi   CM068305.1  40142102   \n",
       "44   BARD1     gene-KSF78_0005473     japonicum   CM038283.1  43890028   \n",
       "45   BARD1   gene-SSPN1_LOCUS8341     spindalis   OX336416.1  50816278   \n",
       "46   BARD1        gene-Smp_137090       mansoni  NC_031498.1   6736171   \n",
       "47   BARD1   gene-SBOV1_LOCUS9420         bovis   OX103961.1  85568270   \n",
       "48   BARD1  gene-SMTH1_LOCUS10644      mattheei   OX103943.1  86929274   \n",
       "49    CBX4   gene-SSPN1_LOCUS9023     spindalis   OX336416.1  79888213   \n",
       "50    CBX4        gene-Smp_078280       mansoni  NC_031502.1  26206232   \n",
       "51    CBX4   gene-SBOV1_LOCUS9010         bovis   OX103961.1  67907733   \n",
       "52    CBX4   gene-SINT2_LOCUS6604  intercalatum   OX103881.1  13847789   \n",
       "53    CBX4     gene-KSF78_0005240     japonicum   CM038283.1  33883412   \n",
       "54    CBX4  gene-SMTH1_LOCUS10167      mattheei   OX103943.1  68689821   \n",
       "55    CBX4      gene-MN116_001223       mekongi   CM068305.1  30297854   \n",
       "\n",
       "         end  \n",
       "0   20192511  \n",
       "1   19025157  \n",
       "2   12028457  \n",
       "3    6425032  \n",
       "4   72392806  \n",
       "5    9569391  \n",
       "6   17899255  \n",
       "7   28315298  \n",
       "8   69802550  \n",
       "9   86991956  \n",
       "10  85881366  \n",
       "11  42464061  \n",
       "12  41398756  \n",
       "13  55552968  \n",
       "14  71997418  \n",
       "15  12286004  \n",
       "16  37998717  \n",
       "17  15205964  \n",
       "18  11780504  \n",
       "19   1849218  \n",
       "20  16601429  \n",
       "21  62196869  \n",
       "22  65509237  \n",
       "23  61607504  \n",
       "24  63322369  \n",
       "25  45559573  \n",
       "26  78138330  \n",
       "27  64745603  \n",
       "28  75644170  \n",
       "29  53330817  \n",
       "30  67951675  \n",
       "31  34126820  \n",
       "32  47266275  \n",
       "33  84901726  \n",
       "34  25784263  \n",
       "35  14430214  \n",
       "36  35858909  \n",
       "37  26937693  \n",
       "38  25413756  \n",
       "39  25589180  \n",
       "40  38308558  \n",
       "41  21850802  \n",
       "42  15504708  \n",
       "43  40205167  \n",
       "44  43902920  \n",
       "45  50861849  \n",
       "46   6746795  \n",
       "47  85581490  \n",
       "48  86944028  \n",
       "49  79917220  \n",
       "50  26232252  \n",
       "51  67951675  \n",
       "52  13862796  \n",
       "53  33885793  \n",
       "54  68736678  \n",
       "55  30329074  "
      ]
     },
     "execution_count": 486,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Добавим в табличку хромосому и координаты каждого\n",
    "\n",
    "import logging\n",
    "\n",
    "gene_organism_to_position = dict()\n",
    "\n",
    "for _, row in epigenetic_genes_for_alignment.iterrows():\n",
    "    gene_organism_to_position[(row['organism'], row['gene'])] = None\n",
    "\n",
    "for organism in data_config:\n",
    "    if organism in gff_files:\n",
    "        with open(gff_files[organism], 'r') as gff_file:\n",
    "            for row in gff_file:\n",
    "                current_row = row.strip()\n",
    "                if len(current_row) == 0 or current_row[0] == '#':\n",
    "                    continue\n",
    "                current_row_splitted = row.split()\n",
    "\n",
    "                if current_row_splitted[2] == 'gene':\n",
    "                    current_gene_id = ''\n",
    "                    if pathlib.Path(gff_files[organism]).suffix == '.gff':\n",
    "                        current_gene_id = current_row_splitted[8].split('ID=')[1].split(';')[0]\n",
    "                    elif pathlib.Path(gff_files[organism]).suffix == '.gff3':\n",
    "                        current_gene_id = current_row_splitted[8].split('ID=')[1].split(';')[0]\n",
    "                        if current_gene_id == 'MS3_00007274':\n",
    "                            logging.error(f'{current_gene_id}, {current_row_splitted[3]}, {current_row_splitted[4]}')\n",
    "                    if (organism, current_gene_id) in gene_organism_to_position:\n",
    "                        gene_organism_to_position[(organism, current_gene_id)] = (current_row_splitted[0], int(current_row_splitted[3]), int(current_row_splitted[4]))\n",
    "    else:\n",
    "        logging.warning(f'No such gff file for {organism}')\n",
    "\n",
    "def get_position(row):\n",
    "    res = gene_organism_to_position[(row['organism'], row['gene'])]\n",
    "    return res\n",
    "\n",
    "epigenetic_genes_for_alignment[['record_id', 'start', 'end']] = \\\n",
    "    epigenetic_genes_for_alignment.apply(get_position, axis=1, result_type='expand')\n",
    "\n",
    "epigenetic_genes_for_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 487,
   "id": "6555a129",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family</th>\n",
       "      <th>gene</th>\n",
       "      <th>organism</th>\n",
       "      <th>record_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SSPN1_LOCUS6014</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336419.1</td>\n",
       "      <td>20174797</td>\n",
       "      <td>20192511</td>\n",
       "      <td>atgtctactccggagggtgttctgttgccagatggccggagaattg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SMTH1_LOCUS6929</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103946.1</td>\n",
       "      <td>19000643</td>\n",
       "      <td>19025157</td>\n",
       "      <td>tcaacgtcgtgatcgataacgatttgatgaggaacgatctctacgt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-KSF78_0008119</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038286.1</td>\n",
       "      <td>12019241</td>\n",
       "      <td>12028457</td>\n",
       "      <td>gcatttgaatcaagtcaccgcggatagtttaagaatagcttattta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-Smp_039490.1</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031499.1</td>\n",
       "      <td>6407653</td>\n",
       "      <td>6425032</td>\n",
       "      <td>aacggacaatgtcttcataaaatgacaaaacaaattttttttcgga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SBOV1_LOCUS9112</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103961.1</td>\n",
       "      <td>72383764</td>\n",
       "      <td>72392806</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-MN116_008246</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068311.1</td>\n",
       "      <td>9544071</td>\n",
       "      <td>9569391</td>\n",
       "      <td>atgtctactccagaggacgttattttacctgatggtcgtagaattg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>gene-SINT2_LOCUS6163</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103880.1</td>\n",
       "      <td>17878940</td>\n",
       "      <td>17899255</td>\n",
       "      <td>ttaacgtcgtgatcgataacgatttgatgaggaacgatctctacgt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SINT2_LOCUS8012</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>28294149</td>\n",
       "      <td>28315298</td>\n",
       "      <td>atgtcttctgggatagaggtgattgagtctggtttccatgttactc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SBOV1_LOCUS9059</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103961.1</td>\n",
       "      <td>69774510</td>\n",
       "      <td>69802550</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SSPN1_LOCUS9217</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>86989145</td>\n",
       "      <td>86991956</td>\n",
       "      <td>atgagtatattattcaatccatcattttttgatatacacttctttt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-SMTH1_LOCUS10615</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>85876995</td>\n",
       "      <td>85881366</td>\n",
       "      <td>gttcaccatttctagtcatgtttgggagatgaaaacctgtttatct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-KSF78_0005440</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>42441718</td>\n",
       "      <td>42464061</td>\n",
       "      <td>aatttatttggctgattcatcggattgtgtccttctattacgatct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-MN116_001510</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>41392895</td>\n",
       "      <td>41398756</td>\n",
       "      <td>atgaatcgcacaagttctccgttggatgaagcatatgggacataca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AEBP2</td>\n",
       "      <td>gene-Smp_155070</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031502.1</td>\n",
       "      <td>55549177</td>\n",
       "      <td>55552968</td>\n",
       "      <td>atgtcttattcatttaattactcaacgccaaaattctttaatgaag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-MN116_002201</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>71954612</td>\n",
       "      <td>71997418</td>\n",
       "      <td>atgaatatggtaactttcttttgtcttggtaaagttgttcgagaaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SMTH1_LOCUS7396</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103948.1</td>\n",
       "      <td>12246346</td>\n",
       "      <td>12286004</td>\n",
       "      <td>ctaatgcatgagcttttcttctacaaccttccatgtatccttttca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-Smp_157150</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031502.1</td>\n",
       "      <td>37972738</td>\n",
       "      <td>37998717</td>\n",
       "      <td>atgagttcccttccattatgcatgaagccatatgagcagtttatgc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SBOV1_LOCUS6579</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103958.1</td>\n",
       "      <td>15163266</td>\n",
       "      <td>15205964</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SINT2_LOCUS6554</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103881.1</td>\n",
       "      <td>11733954</td>\n",
       "      <td>11780504</td>\n",
       "      <td>gagagagagagagagagagagattaactcttgctacatatgaacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-KSF78_0007878</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038286.1</td>\n",
       "      <td>1806863</td>\n",
       "      <td>1849218</td>\n",
       "      <td>cgaaacaacgattcatttattaaccctgtaaaatatttatatcctc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>gene-SSPN1_LOCUS6463</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336421.1</td>\n",
       "      <td>16572420</td>\n",
       "      <td>16601429</td>\n",
       "      <td>agaagaagaagaagaagaagaagaagaagaagaagaagaatgcatg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-MN116_002008</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>62124388</td>\n",
       "      <td>62196869</td>\n",
       "      <td>ttacaagaacggcatccatcctttgtacatttgacagagtcgatca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SMTH1_LOCUS10083</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>65486722</td>\n",
       "      <td>65509237</td>\n",
       "      <td>ttataagaatggcatccatcctttatacatttgacagagtcgatca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SINT2_LOCUS8890</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103883.1</td>\n",
       "      <td>61543600</td>\n",
       "      <td>61607504</td>\n",
       "      <td>ttataagaatggcatccatcctttatacatttgacagagtcggtca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SSPN1_LOCUS8606</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>63284548</td>\n",
       "      <td>63322369</td>\n",
       "      <td>attatctagtaattacttttctagtgtttccaaagcaacattatta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-Smp_162340</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031502.1</td>\n",
       "      <td>45490939</td>\n",
       "      <td>45559573</td>\n",
       "      <td>atgaaaaacagcgaaaggttagttgctaagatgtgtttcgtgctct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-KSF78_0004211</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038289.1</td>\n",
       "      <td>78055480</td>\n",
       "      <td>78138330</td>\n",
       "      <td>ccgattagcgtttcttcaagcatgttgttttttacgggatggggtt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ATR</td>\n",
       "      <td>gene-SBOV1_LOCUS8938</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103961.1</td>\n",
       "      <td>64679616</td>\n",
       "      <td>64745603</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SMTH1_LOCUS10333</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>75611087</td>\n",
       "      <td>75644170</td>\n",
       "      <td>tcatattgctaattcttcgtcatatcgtaaacgcaagaaatcttgt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-Smp_163110</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031502.1</td>\n",
       "      <td>53300151</td>\n",
       "      <td>53330817</td>\n",
       "      <td>gtaacgctacggggtagtttttggttcgaggaatggccacggcatt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SBOV1_LOCUS9010</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103961.1</td>\n",
       "      <td>67907733</td>\n",
       "      <td>67951675</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-MN116_001315</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>34110633</td>\n",
       "      <td>34126820</td>\n",
       "      <td>ctagaagtcttcatcatcagggtcatcatcggctacatcattgtca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-KSF78_0005543</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>47227496</td>\n",
       "      <td>47266275</td>\n",
       "      <td>ttaaaattgcggaaaatgaaaatttggaaaactttgtgaatgtaaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SSPN1_LOCUS9160</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>84866275</td>\n",
       "      <td>84901726</td>\n",
       "      <td>ttaaagatataggctcgcacagaaaatgatatctcttttgattagg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>gene-SINT2_LOCUS5228</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>25744885</td>\n",
       "      <td>25784263</td>\n",
       "      <td>cagtacacttcgctgtaatatctggtaatggtcatttcagtaaaga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-Smp_168800.1</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031497.1</td>\n",
       "      <td>14428597</td>\n",
       "      <td>14430214</td>\n",
       "      <td>ttacgataggcataacttaccgagttcatatagagatccgtaatgc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SINT2_LOCUS956</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103876.1</td>\n",
       "      <td>35851198</td>\n",
       "      <td>35858909</td>\n",
       "      <td>cgctagatatattcatgtttacacttggctcatctcatcaatttcc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SMTH1_LOCUS4562</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103945.1</td>\n",
       "      <td>26934207</td>\n",
       "      <td>26937693</td>\n",
       "      <td>tcagttattatcagaacataatcccatcaaactgaattctacactt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-MN116_003045</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068306.1</td>\n",
       "      <td>25384891</td>\n",
       "      <td>25413756</td>\n",
       "      <td>tcaaagacgccgtttccgaaatcgttgagaattggatgaagtagac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SBOV1_LOCUS4028</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103955.1</td>\n",
       "      <td>25583114</td>\n",
       "      <td>25589180</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-KSF78_0005343</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>38289344</td>\n",
       "      <td>38308558</td>\n",
       "      <td>ctcagtttatttgaaattatcaatgttaaccttggtaatgatatat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>gene-SSPN1_LOCUS3856</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336418.1</td>\n",
       "      <td>21839072</td>\n",
       "      <td>21850802</td>\n",
       "      <td>agtttatataggctagcctcctgttgtcgaaaattttggatattgc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SINT2_LOCUS4989</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103879.1</td>\n",
       "      <td>15489966</td>\n",
       "      <td>15504708</td>\n",
       "      <td>taaagatctcaaaaattatattgaaattgactgaagcaaattaaaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-MN116_001474</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>40142102</td>\n",
       "      <td>40205167</td>\n",
       "      <td>ttacagtattgtgcatttccgacaacattcttgacgaaataaatta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-KSF78_0005473</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>43890028</td>\n",
       "      <td>43902920</td>\n",
       "      <td>cgttgtaactgacagactgatattgccttgacttagcacataatac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SSPN1_LOCUS8341</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>50816278</td>\n",
       "      <td>50861849</td>\n",
       "      <td>agattatttaccagattgcaataaattatttgcactatcctatcgt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-Smp_137090</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031498.1</td>\n",
       "      <td>6736171</td>\n",
       "      <td>6746795</td>\n",
       "      <td>ttaacagttgggaaaaggaagtagacgatattctgcagcacaatcc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SBOV1_LOCUS9420</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103961.1</td>\n",
       "      <td>85568270</td>\n",
       "      <td>85581490</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>gene-SMTH1_LOCUS10644</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>86929274</td>\n",
       "      <td>86944028</td>\n",
       "      <td>tctctctctctatgcatacttacatttaagaaatgaatcacaacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-SSPN1_LOCUS9023</td>\n",
       "      <td>spindalis</td>\n",
       "      <td>OX336416.1</td>\n",
       "      <td>79888213</td>\n",
       "      <td>79917220</td>\n",
       "      <td>ttcacatgtacggtagaaatttattttttttacaggagtcttcgat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-Smp_078280</td>\n",
       "      <td>mansoni</td>\n",
       "      <td>NC_031502.1</td>\n",
       "      <td>26206232</td>\n",
       "      <td>26232252</td>\n",
       "      <td>ttacacaaaatcatccatttcatttaaaatgaactccatagctgta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-SBOV1_LOCUS9010</td>\n",
       "      <td>bovis</td>\n",
       "      <td>OX103961.1</td>\n",
       "      <td>67907733</td>\n",
       "      <td>67951675</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-SINT2_LOCUS6604</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>OX103881.1</td>\n",
       "      <td>13847789</td>\n",
       "      <td>13862796</td>\n",
       "      <td>cccgcgaagaaaaaaggatttaacgcactaatttttgaatggttga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-KSF78_0005240</td>\n",
       "      <td>japonicum</td>\n",
       "      <td>CM038283.1</td>\n",
       "      <td>33883412</td>\n",
       "      <td>33885793</td>\n",
       "      <td>tctattttatctttcatatatttgtgatctttgtacagaaaattat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-SMTH1_LOCUS10167</td>\n",
       "      <td>mattheei</td>\n",
       "      <td>OX103943.1</td>\n",
       "      <td>68689821</td>\n",
       "      <td>68736678</td>\n",
       "      <td>atggaacatcagtaccgatcattcggtgtcgagtcatctagcttcc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>gene-MN116_001223</td>\n",
       "      <td>mekongi</td>\n",
       "      <td>CM068305.1</td>\n",
       "      <td>30297854</td>\n",
       "      <td>30329074</td>\n",
       "      <td>atggaatcggctcctgttaatactgacaaaccgtgtgacatccttg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    family                   gene      organism    record_id     start  \\\n",
       "0    ACINU   gene-SSPN1_LOCUS6014     spindalis   OX336419.1  20174797   \n",
       "1    ACINU   gene-SMTH1_LOCUS6929      mattheei   OX103946.1  19000643   \n",
       "2    ACINU     gene-KSF78_0008119     japonicum   CM038286.1  12019241   \n",
       "3    ACINU      gene-Smp_039490.1       mansoni  NC_031499.1   6407653   \n",
       "4    ACINU   gene-SBOV1_LOCUS9112         bovis   OX103961.1  72383764   \n",
       "5    ACINU      gene-MN116_008246       mekongi   CM068311.1   9544071   \n",
       "6    ACINU   gene-SINT2_LOCUS6163  intercalatum   OX103880.1  17878940   \n",
       "7    AEBP2   gene-SINT2_LOCUS8012  intercalatum   OX103883.1  28294149   \n",
       "8    AEBP2   gene-SBOV1_LOCUS9059         bovis   OX103961.1  69774510   \n",
       "9    AEBP2   gene-SSPN1_LOCUS9217     spindalis   OX336416.1  86989145   \n",
       "10   AEBP2  gene-SMTH1_LOCUS10615      mattheei   OX103943.1  85876995   \n",
       "11   AEBP2     gene-KSF78_0005440     japonicum   CM038283.1  42441718   \n",
       "12   AEBP2      gene-MN116_001510       mekongi   CM068305.1  41392895   \n",
       "13   AEBP2        gene-Smp_155070       mansoni  NC_031502.1  55549177   \n",
       "14  ARID1A      gene-MN116_002201       mekongi   CM068305.1  71954612   \n",
       "15  ARID1A   gene-SMTH1_LOCUS7396      mattheei   OX103948.1  12246346   \n",
       "16  ARID1A        gene-Smp_157150       mansoni  NC_031502.1  37972738   \n",
       "17  ARID1A   gene-SBOV1_LOCUS6579         bovis   OX103958.1  15163266   \n",
       "18  ARID1A   gene-SINT2_LOCUS6554  intercalatum   OX103881.1  11733954   \n",
       "19  ARID1A     gene-KSF78_0007878     japonicum   CM038286.1   1806863   \n",
       "20  ARID1A   gene-SSPN1_LOCUS6463     spindalis   OX336421.1  16572420   \n",
       "21     ATR      gene-MN116_002008       mekongi   CM068305.1  62124388   \n",
       "22     ATR  gene-SMTH1_LOCUS10083      mattheei   OX103943.1  65486722   \n",
       "23     ATR   gene-SINT2_LOCUS8890  intercalatum   OX103883.1  61543600   \n",
       "24     ATR   gene-SSPN1_LOCUS8606     spindalis   OX336416.1  63284548   \n",
       "25     ATR        gene-Smp_162340       mansoni  NC_031502.1  45490939   \n",
       "26     ATR     gene-KSF78_0004211     japonicum   CM038289.1  78055480   \n",
       "27     ATR   gene-SBOV1_LOCUS8938         bovis   OX103961.1  64679616   \n",
       "28    ATRX  gene-SMTH1_LOCUS10333      mattheei   OX103943.1  75611087   \n",
       "29    ATRX        gene-Smp_163110       mansoni  NC_031502.1  53300151   \n",
       "30    ATRX   gene-SBOV1_LOCUS9010         bovis   OX103961.1  67907733   \n",
       "31    ATRX      gene-MN116_001315       mekongi   CM068305.1  34110633   \n",
       "32    ATRX     gene-KSF78_0005543     japonicum   CM038283.1  47227496   \n",
       "33    ATRX   gene-SSPN1_LOCUS9160     spindalis   OX336416.1  84866275   \n",
       "34    ATRX   gene-SINT2_LOCUS5228  intercalatum   OX103879.1  25744885   \n",
       "35    BAP1      gene-Smp_168800.1       mansoni  NC_031497.1  14428597   \n",
       "36    BAP1    gene-SINT2_LOCUS956  intercalatum   OX103876.1  35851198   \n",
       "37    BAP1   gene-SMTH1_LOCUS4562      mattheei   OX103945.1  26934207   \n",
       "38    BAP1      gene-MN116_003045       mekongi   CM068306.1  25384891   \n",
       "39    BAP1   gene-SBOV1_LOCUS4028         bovis   OX103955.1  25583114   \n",
       "40    BAP1     gene-KSF78_0005343     japonicum   CM038283.1  38289344   \n",
       "41    BAP1   gene-SSPN1_LOCUS3856     spindalis   OX336418.1  21839072   \n",
       "42   BARD1   gene-SINT2_LOCUS4989  intercalatum   OX103879.1  15489966   \n",
       "43   BARD1      gene-MN116_001474       mekongi   CM068305.1  40142102   \n",
       "44   BARD1     gene-KSF78_0005473     japonicum   CM038283.1  43890028   \n",
       "45   BARD1   gene-SSPN1_LOCUS8341     spindalis   OX336416.1  50816278   \n",
       "46   BARD1        gene-Smp_137090       mansoni  NC_031498.1   6736171   \n",
       "47   BARD1   gene-SBOV1_LOCUS9420         bovis   OX103961.1  85568270   \n",
       "48   BARD1  gene-SMTH1_LOCUS10644      mattheei   OX103943.1  86929274   \n",
       "49    CBX4   gene-SSPN1_LOCUS9023     spindalis   OX336416.1  79888213   \n",
       "50    CBX4        gene-Smp_078280       mansoni  NC_031502.1  26206232   \n",
       "51    CBX4   gene-SBOV1_LOCUS9010         bovis   OX103961.1  67907733   \n",
       "52    CBX4   gene-SINT2_LOCUS6604  intercalatum   OX103881.1  13847789   \n",
       "53    CBX4     gene-KSF78_0005240     japonicum   CM038283.1  33883412   \n",
       "54    CBX4  gene-SMTH1_LOCUS10167      mattheei   OX103943.1  68689821   \n",
       "55    CBX4      gene-MN116_001223       mekongi   CM068305.1  30297854   \n",
       "\n",
       "         end                                               data  \n",
       "0   20192511  atgtctactccggagggtgttctgttgccagatggccggagaattg...  \n",
       "1   19025157  tcaacgtcgtgatcgataacgatttgatgaggaacgatctctacgt...  \n",
       "2   12028457  gcatttgaatcaagtcaccgcggatagtttaagaatagcttattta...  \n",
       "3    6425032  aacggacaatgtcttcataaaatgacaaaacaaattttttttcgga...  \n",
       "4   72392806                                               none  \n",
       "5    9569391  atgtctactccagaggacgttattttacctgatggtcgtagaattg...  \n",
       "6   17899255  ttaacgtcgtgatcgataacgatttgatgaggaacgatctctacgt...  \n",
       "7   28315298  atgtcttctgggatagaggtgattgagtctggtttccatgttactc...  \n",
       "8   69802550                                               none  \n",
       "9   86991956  atgagtatattattcaatccatcattttttgatatacacttctttt...  \n",
       "10  85881366  gttcaccatttctagtcatgtttgggagatgaaaacctgtttatct...  \n",
       "11  42464061  aatttatttggctgattcatcggattgtgtccttctattacgatct...  \n",
       "12  41398756  atgaatcgcacaagttctccgttggatgaagcatatgggacataca...  \n",
       "13  55552968  atgtcttattcatttaattactcaacgccaaaattctttaatgaag...  \n",
       "14  71997418  atgaatatggtaactttcttttgtcttggtaaagttgttcgagaaa...  \n",
       "15  12286004  ctaatgcatgagcttttcttctacaaccttccatgtatccttttca...  \n",
       "16  37998717  atgagttcccttccattatgcatgaagccatatgagcagtttatgc...  \n",
       "17  15205964                                               none  \n",
       "18  11780504  gagagagagagagagagagagattaactcttgctacatatgaacat...  \n",
       "19   1849218  cgaaacaacgattcatttattaaccctgtaaaatatttatatcctc...  \n",
       "20  16601429  agaagaagaagaagaagaagaagaagaagaagaagaagaatgcatg...  \n",
       "21  62196869  ttacaagaacggcatccatcctttgtacatttgacagagtcgatca...  \n",
       "22  65509237  ttataagaatggcatccatcctttatacatttgacagagtcgatca...  \n",
       "23  61607504  ttataagaatggcatccatcctttatacatttgacagagtcggtca...  \n",
       "24  63322369  attatctagtaattacttttctagtgtttccaaagcaacattatta...  \n",
       "25  45559573  atgaaaaacagcgaaaggttagttgctaagatgtgtttcgtgctct...  \n",
       "26  78138330  ccgattagcgtttcttcaagcatgttgttttttacgggatggggtt...  \n",
       "27  64745603                                               none  \n",
       "28  75644170  tcatattgctaattcttcgtcatatcgtaaacgcaagaaatcttgt...  \n",
       "29  53330817  gtaacgctacggggtagtttttggttcgaggaatggccacggcatt...  \n",
       "30  67951675                                               none  \n",
       "31  34126820  ctagaagtcttcatcatcagggtcatcatcggctacatcattgtca...  \n",
       "32  47266275  ttaaaattgcggaaaatgaaaatttggaaaactttgtgaatgtaaa...  \n",
       "33  84901726  ttaaagatataggctcgcacagaaaatgatatctcttttgattagg...  \n",
       "34  25784263  cagtacacttcgctgtaatatctggtaatggtcatttcagtaaaga...  \n",
       "35  14430214  ttacgataggcataacttaccgagttcatatagagatccgtaatgc...  \n",
       "36  35858909  cgctagatatattcatgtttacacttggctcatctcatcaatttcc...  \n",
       "37  26937693  tcagttattatcagaacataatcccatcaaactgaattctacactt...  \n",
       "38  25413756  tcaaagacgccgtttccgaaatcgttgagaattggatgaagtagac...  \n",
       "39  25589180                                               none  \n",
       "40  38308558  ctcagtttatttgaaattatcaatgttaaccttggtaatgatatat...  \n",
       "41  21850802  agtttatataggctagcctcctgttgtcgaaaattttggatattgc...  \n",
       "42  15504708  taaagatctcaaaaattatattgaaattgactgaagcaaattaaaa...  \n",
       "43  40205167  ttacagtattgtgcatttccgacaacattcttgacgaaataaatta...  \n",
       "44  43902920  cgttgtaactgacagactgatattgccttgacttagcacataatac...  \n",
       "45  50861849  agattatttaccagattgcaataaattatttgcactatcctatcgt...  \n",
       "46   6746795  ttaacagttgggaaaaggaagtagacgatattctgcagcacaatcc...  \n",
       "47  85581490                                               none  \n",
       "48  86944028  tctctctctctatgcatacttacatttaagaaatgaatcacaacat...  \n",
       "49  79917220  ttcacatgtacggtagaaatttattttttttacaggagtcttcgat...  \n",
       "50  26232252  ttacacaaaatcatccatttcatttaaaatgaactccatagctgta...  \n",
       "51  67951675                                               none  \n",
       "52  13862796  cccgcgaagaaaaaaggatttaacgcactaatttttgaatggttga...  \n",
       "53  33885793  tctattttatctttcatatatttgtgatctttgtacagaaaattat...  \n",
       "54  68736678  atggaacatcagtaccgatcattcggtgtcgagtcatctagcttcc...  \n",
       "55  30329074  atggaatcggctcctgttaatactgacaaaccgtgtgacatccttg...  "
      ]
     },
     "execution_count": 487,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Добавим в табличку последовательность гена\n",
    "\n",
    "import pandas as pd\n",
    "import logging\n",
    "from Bio import SeqIO\n",
    "\n",
    "record_to_position = dict()\n",
    "\n",
    "for _, row in epigenetic_genes_for_alignment.iterrows():\n",
    "    if row['record_id'] not in record_to_position:\n",
    "        record_to_position[row['record_id']] = dict()\n",
    "    record_to_position[row['record_id']][(row['start'], row['end'])] = None\n",
    "\n",
    "record_to_position\n",
    "\n",
    "for organism in data_config:\n",
    "    if organism == 'bovis':\n",
    "        continue\n",
    "    if organism in fna_files:\n",
    "        records = list(SeqIO.parse(fna_files[organism], \"fasta\"))\n",
    "        for record in records:\n",
    "            if record.id in record_to_position:\n",
    "                for requested_range in record_to_position[record.id]:\n",
    "                    record_to_position[record.id][requested_range] = record.seq[requested_range[0]-1:requested_range[1]]\n",
    "\n",
    "    else:\n",
    "        logging.warning(f'No such fna file for {organism}')\n",
    "\n",
    "def get_sequence(row):\n",
    "    return [record_to_position[row['record_id']][(row['start'], row['end'])]]\n",
    "        \n",
    "epigenetic_genes_for_alignment[['data']] = \\\n",
    "    epigenetic_genes_for_alignment.apply(get_sequence, axis=1, result_type='expand')\n",
    "\n",
    "epigenetic_genes_for_alignment['data'] = epigenetic_genes_for_alignment['data'].apply(lambda s : str(s).lower())\n",
    "epigenetic_genes_for_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 387,
   "id": "0609546e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Running allignment for data/epigenetic_genes_alignment_ACINU.meg\n",
      "INFO:root:Running allignment for data/epigenetic_genes_alignment_AEBP2.meg\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Running allignment for data/epigenetic_genes_alignment_ARID1A.meg\n",
      "INFO:root:Running allignment for data/epigenetic_genes_alignment_ATR.meg\n",
      "INFO:root:Running allignment for data/epigenetic_genes_alignment_ATRX.meg\n",
      "INFO:root:Running allignment for data/epigenetic_genes_alignment_BAP1.meg\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/18/2025 06:33:58 PM\n",
      "Alignment settings:\n",
      "                          : Pairwise Alignment                  : ====================\n",
      "                          : DNAPWGapOpeningPenalty              : 15.00\n",
      "                          : DNAPWGapExtensionPenalty            : 6.66\n",
      "                          : Multiple Alignment                  : ====================\n",
      "                          : DNAMAGapOpeningPenalty              : 15.00\n",
      "                          : DNAMAGapExtensionPenalty            : 6.66\n",
      "                          : Global Options                      : ====================\n",
      "                          : DNA Weight Matrix                   : IUB\n",
      "                          : TransitionWeightNEdit               : 0.50\n",
      "                          : UseNegativeMatrix                   : ON\n",
      "                          : DelayDivergentCutoff                : 30\n",
      "                          : KeepPredefinedGaps                  : True\n",
      "                          : datatype                            : snNucleotide\n",
      "                          : containsCodingNuc                   : False\n",
      "                          : missingBaseSymbol                   : ?\n",
      "                          : identicalBasMEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/18/2025 06:33:58 PM\n",
      "Alignment settings:\n",
      "                          : Pairwise Alignment                  : ====================\n",
      "                          : DNAPWGapOpeningPenalty              : 15.00\n",
      "                          : DNAPWGapExtensionPenalty            : 6.66\n",
      "                          : Multiple Alignment                  : ====================\n",
      "                          : DNAMAGapOpeningPenalty              : 15.00\n",
      "                          : DNAMAGapExtensionPenalty            : 6.66\n",
      "                          : Global Options                      : ====================\n",
      "                          : DNA Weight Matrix                   : IUB\n",
      "                          : TransitionWeightNEdit               : 0.50\n",
      "                          : UseNegativeMatrix                   : ON\n",
      "                          : DelayDivergentCutoff                : 30\n",
      "                          : KeepPredefinedGaps                  : True\n",
      "                          : datatype                            : snNucleotide\n",
      "                          : containsCodingNuc                   : False\n",
      "                          : missingBaseSymbol                   : ?\n",
      "                          : identicalBasMEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/18/2025 06:33:58 PM\n",
      "Alignment settings:\n",
      "                          : Pairwise Alignment                  : ====================\n",
      "                          : DNAPWGapOpeningPenalty              : 15.00\n",
      "                          : DNAPWGapExtensionPenalty            : 6.66\n",
      "                          : Multiple Alignment                  : ====================\n",
      "                          : DNAMAGapOpeningPenalty              : 15.00\n",
      "                          : DNAMAGapExtensionPenalty            : 6.66\n",
      "                          : Global Options                      : ====================\n",
      "                          : DNA Weight Matrix                   : IUB\n",
      "                          : TransitionWeightNEdit               : 0.50\n",
      "                          : UseNegativeMatrix                   : ON\n",
      "                          : DelayDivergentCutoff                : 30\n",
      "                          : KeepPredefinedGaps                  : True\n",
      "                          : datatype                            : snNucleotide\n",
      "                          : containsCodingNuc                   : False\n",
      "                          : missingBaseSymbol                   : ?\n",
      "                          : identicalBasMEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/18/2025 06:33:58 PM\n",
      "Alignment settings:\n",
      "                          : Pairwise Alignment                  : ====================\n",
      "                          : DNAPWGapOpeningPenalty              : 15.00\n",
      "                          : DNAPWGapExtensionPenalty            : 6.66\n",
      "                          : Multiple Alignment                  : ====================\n",
      "                          : DNAMAGapOpeningPenalty              : 15.00\n",
      "                          : DNAMAGapExtensionPenalty            : 6.66\n",
      "                          : Global Options                      : ====================\n",
      "                          : DNA Weight Matrix                   : IUB\n",
      "                          : TransitionWeightNEdit               : 0.50\n",
      "                          : UseNegativeMatrix                   : ON\n",
      "                          : DelayDivergentCutoff                : 30\n",
      "                          : KeepPredefinedGaps                  : True\n",
      "                          : datatype                            : snNucleotide\n",
      "                          : containsCodingNuc                   : False\n",
      "                          : missingBaseSymbol                   : ?\n",
      "                          : identicalBasMEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/18/2025 06:33:58 PM\n",
      "Alignment settings:\n",
      "                          : Pairwise Alignment                  : ====================\n",
      "                          : DNAPWGapOpeningPenalty              : 15.00\n",
      "                          : DNAPWGapExtensionPenalty            : 6.66\n",
      "                          : Multiple Alignment                  : ====================\n",
      "                          : DNAMAGapOpeningPenalty              : 15.00\n",
      "                          : DNAMAGapExtensionPenalty            : 6.66\n",
      "                          : Global Options                      : ====================\n",
      "                          : DNA Weight Matrix                   : IUB\n",
      "                          : TransitionWeightNEdit               : 0.50\n",
      "                          : UseNegativeMatrix                   : ON\n",
      "                          : DelayDivergentCutoff                : 30\n",
      "                          : KeepPredefinedGaps                  : True\n",
      "                          : datatype                            : snNucleotide\n",
      "                          : containsCodingNuc                   : False\n",
      "                          : missingBaseSymbol                   : ?\n",
      "                          : identicalBasMEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/18/2025 06:33:58 PM\n",
      "Alignment settings:\n",
      "                          : Pairwise Alignment                  : ====================\n",
      "                          : DNAPWGapOpeningPenalty              : 15.00\n",
      "                          : DNAPWGapExtensionPenalty            : 6.66\n",
      "                          : Multiple Alignment                  : ====================\n",
      "                          : DNAMAGapOpeningPenalty              : 15.00\n",
      "                          : DNAMAGapExtensionPenalty            : 6.66\n",
      "                          : Global Options                      : ====================\n",
      "                          : DNA Weight Matrix                   : IUB\n",
      "                          : TransitionWeightNEdit               : 0.50\n",
      "                          : UseNegativeMatrix                   : ON\n",
      "                          : DelayDivergentCutoff                : 30\n",
      "                          : KeepPredefinedGaps                  : True\n",
      "                          : datatype                            : snNucleotide\n",
      "                          : containsCodingNuc                   : False\n",
      "                          : missingBaseSymbol                   : ?\n",
      "                          : identicalBas"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allignment for data/epigenetic_genes_for_alignment_AEBP2.fasta done\n",
      "INFO:root:Running allignment for data/epigenetic_genes_alignment_BARD1.meg\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing pairwise alignment\n",
      "\n",
      "Executing multiple sequence alignment\n",
      ".\n",
      "eSymbol                 : .\n",
      "                          : gapSymbol                           : -\n",
      " \n",
      ".\n",
      "\n",
      "       100% Analysis Complete                                                               \n",
      " \n",
      "MEGA has completed the requested action\n",
      "\n",
      "Terminating the megacc process with successful exit status\n",
      "\n",
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/18/2025 06:35:57 PM\n",
      "Alignment settings:\n",
      "                          : Pairwise Alignment                  : ====================\n",
      "                          : DNAPWGapOpeningPenalty              : 15.00\n",
      "                          : DNAPWGapExtensionPenalty            : 6.66\n",
      "                          : Multiple Alignment                  : ====================\n",
      "                          : DNAMAGapOpeningPenalty              : 15.00\n",
      "                          : DNAMAGapExtensionPenalty            : 6.66\n",
      "                          : Global Options                      : ====================\n",
      "                          : DNA Weight Matrix                   : IUB\n",
      "                          : TransitionWeightNEdit               : 0.50\n",
      "                          : UseNegativeMatrix                   : ON\n",
      "                          : DelayDivergentCutoff                : 30\n",
      "                          : KeepPredefinedGaps                  : True\n",
      "                          : datatype                            : snNucleotide\n",
      "                          : containsCodingNuc                   : False\n",
      "                          : missingBaseSymbol                   : ?\n",
      "                          : identicalBas"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allignment for data/epigenetic_genes_for_alignment_BAP1.fasta done\n",
      "INFO:root:Running allignment for data/epigenetic_genes_alignment_CBX4.meg\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing pairwise alignment\n",
      "\n",
      "Executing multiple sequence alignment\n",
      ".\n",
      "eSymbol                 : .\n",
      "                          : gapSymbol                           : -\n",
      " \n",
      ".\n",
      "\n",
      "       100% Analysis Complete                                                               \n",
      " \n",
      "MEGA has completed the requested action\n",
      "\n",
      "Terminating the megacc process with successful exit status\n",
      "\n",
      "MEGA-CC 11.0.13 Molecular Evolutionary Genetics Analysis\n",
      "Build#: 11220624-x86_64\n",
      "06/18/2025 06:36:19 PM\n",
      "Alignment settings:\n",
      "                          : Pairwise Alignment                  : ====================\n",
      "                          : DNAPWGapOpeningPenalty              : 15.00\n",
      "                          : DNAPWGapExtensionPenalty            : 6.66\n",
      "                          : Multiple Alignment                  : ====================\n",
      "                          : DNAMAGapOpeningPenalty              : 15.00\n",
      "                          : DNAMAGapExtensionPenalty            : 6.66\n",
      "                          : Global Options                      : ====================\n",
      "                          : DNA Weight Matrix                   : IUB\n",
      "                          : TransitionWeightNEdit               : 0.50\n",
      "                          : UseNegativeMatrix                   : ON\n",
      "                          : DelayDivergentCutoff                : 30\n",
      "                          : KeepPredefinedGaps                  : True\n",
      "                          : datatype                            : snNucleotide\n",
      "                          : containsCodingNuc                   : False\n",
      "                          : missingBaseSymbol                   : ?\n",
      "                          : identicalBas"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allignment for data/epigenetic_genes_for_alignment_ACINU.fasta done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing pairwise alignment\n",
      "\n",
      "Executing multiple sequence alignment\n",
      ".\n",
      "eSymbol                 : .\n",
      "                          : gapSymbol                           : -\n",
      " \n",
      ".\n",
      "\n",
      "       100% Analysis Complete                                                               \n",
      " \n",
      "MEGA has completed the requested action\n",
      "\n",
      "Terminating the megacc process with successful exit status\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allignment for data/epigenetic_genes_for_alignment_BARD1.fasta done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing pairwise alignment\n",
      "\n",
      "Executing multiple sequence alignment\n",
      ".\n",
      "eSymbol                 : .\n",
      "                          : gapSymbol                           : -\n",
      " \n",
      ".\n",
      "\n",
      "       100% Analysis Complete                                                               \n",
      " \n",
      "MEGA has completed the requested action\n",
      "\n",
      "Terminating the megacc process with successful exit status\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allignment for data/epigenetic_genes_for_alignment_CBX4.fasta done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing pairwise alignment\n",
      "\n",
      "Executing multiple sequence alignment\n",
      ".\n",
      "eSymbol                 : .\n",
      "                          : gapSymbol                           : -\n",
      " \n",
      ".\n",
      "\n",
      "       100% Analysis Complete                                                               \n",
      " \n",
      "MEGA has completed the requested action\n",
      "\n",
      "Terminating the megacc process with successful exit status\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allignment for data/epigenetic_genes_for_alignment_ATRX.fasta done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing pairwise alignment\n",
      "\n",
      "Executing multiple sequence alignment\n",
      ".\n",
      "eSymbol                 : .\n",
      "                          : gapSymbol                           : -\n",
      " \n",
      ".\n",
      "\n",
      "       100% Analysis Complete                                                               \n",
      " \n",
      "MEGA has completed the requested action\n",
      "\n",
      "Terminating the megacc process with successful exit status\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allignment for data/epigenetic_genes_for_alignment_ARID1A.fasta done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing pairwise alignment\n",
      "\n",
      "Executing multiple sequence alignment\n",
      ".\n",
      "eSymbol                 : .\n",
      "                          : gapSymbol                           : -\n",
      " \n",
      ".\n",
      "\n",
      "       100% Analysis Complete                                                               \n",
      " \n",
      "MEGA has completed the requested action\n",
      "\n",
      "Terminating the megacc process with successful exit status\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allignment for data/epigenetic_genes_for_alignment_ATR.fasta done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing pairwise alignment\n",
      "\n",
      "Executing multiple sequence alignment\n",
      ".\n",
      "eSymbol                 : .\n",
      "                          : gapSymbol                           : -\n",
      " \n",
      ".\n",
      "\n",
      "       100% Analysis Complete                                                               \n",
      " \n",
      "MEGA has completed the requested action\n",
      "\n",
      "Terminating the megacc process with successful exit status\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import logging\n",
    "import subprocess\n",
    "import concurrent.futures\n",
    "import glob\n",
    "\n",
    "# настройки меги для выравнивания\n",
    "with open(\"data/muscle_align.mao\", \"w\") as f:\n",
    "    f.write(\"\"\"; Please do not edit this file! If this file is modified, results are unpredictable.\n",
    "; Instead of modifying this file, simply create a new MEGA Analysis Options file by using the MEGA Prototyper.\n",
    "[ DataSettings ]\n",
    "datatype                 = snNucleotide            \n",
    "containsCodingNuc        = False                   \n",
    "missingBaseSymbol        = ?                       \n",
    "identicalBaseSymbol      = .                       \n",
    "gapSymbol                = -                       \n",
    "[ ProcessTypes ]\n",
    "ppAlign                  = true                    \n",
    "ppClustalW               = true                    \n",
    "[ AnalysisSettings ]\n",
    "Pairwise Alignment       = ====================    \n",
    "DNAPWGapOpeningPenalty   = 15.00                   \n",
    "DNAPWGapExtensionPenalty = 6.66                    \n",
    "Multiple Alignment       = ====================    \n",
    "DNAMAGapOpeningPenalty   = 15.00                   \n",
    "DNAMAGapExtensionPenalty = 6.66                    \n",
    "Global Options           = ====================    \n",
    "DNA Weight Matrix        = IUB                     \n",
    "TransitionWeightNEdit    = 0.50                    \n",
    "UseNegativeMatrix        = ON                      \n",
    "DelayDivergentCutoff     = 30                      \n",
    "KeepPredefinedGaps       = True\"\"\")\n",
    "\n",
    "epigenetec_alignment_results = []\n",
    "\n",
    "\n",
    "\n",
    "def run_alignment(file):\n",
    "    logging.info(f'Running allignment for {file[1]}')\n",
    "    os.system(f'rm -f {file[1]}')\n",
    "    subprocess.run([f'./local_programs/mega/megacc', '-a', 'data/muscle_align.mao', '-d', file[0], '-o', file[1]])\n",
    "\n",
    "    logging.info(f'Allignment for {file[0]} done')\n",
    "\n",
    "files = []\n",
    "\n",
    "for family, subdf in epigenetic_genes_for_alignment.groupby('family'):\n",
    "    # запишем последовательности в fasta файл\n",
    "    current_fasta_file = f'data/epigenetic_genes_for_alignment_{family}.fasta'\n",
    "    current_meg_file = f'data/epigenetic_genes_alignment_{family}.meg'\n",
    "    files.append((current_fasta_file, current_meg_file))\n",
    "    with open(current_fasta_file, 'w') as fasta_file:\n",
    "        for _, row in subdf.iterrows():\n",
    "            seq_id, seq = f'{row['organism']}:{row['gene']}', row['data']\n",
    "            fasta_file.write(f\">{seq_id}\\n{seq}\\n\")\n",
    "    # запуск меги\n",
    "    \n",
    "    epigenetec_alignment_results.append((family, current_meg_file))\n",
    "\n",
    "# у меня всего 6 ядер(((\n",
    "# подарите amd threadripper\n",
    "with concurrent.futures.ThreadPoolExecutor(max_workers=6) as executor:\n",
    "    executor.map(run_alignment, files)\n",
    "    executor.shutdown(wait=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 388,
   "id": "df453ec4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Running tree building for ACINU\n",
      "INFO:root:Tree building for ACINU done\n",
      "INFO:root:Running tree building for AEBP2\n",
      "INFO:root:Tree building for AEBP2 done\n",
      "INFO:root:Running tree building for ARID1A\n",
      "INFO:root:Tree building for ARID1A done\n",
      "INFO:root:Running tree building for ATR\n",
      "INFO:root:Tree building for ATR done\n",
      "INFO:root:Running tree building for ATRX\n",
      "INFO:root:Tree building for ATRX done\n",
      "INFO:root:Running tree building for BAP1\n",
      "INFO:root:Tree building for BAP1 done\n",
      "INFO:root:Running tree building for BARD1\n",
      "INFO:root:Tree building for BARD1 done\n",
      "INFO:root:Running tree building for CBX4\n",
      "INFO:root:Tree building for CBX4 done\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# настройки меги для построения дерева\n",
    "with open(\"data/nj_tree.mao\", \"w\") as f:\n",
    "    f.write(\"\"\"[ DataSettings ]\n",
    "datatype                             = snNucleotide                        \n",
    "containsCodingNuc                    = True                                \n",
    "MissingBaseSymbol                    = ?                                   \n",
    "IdenticalBaseSymbol                  = .                                   \n",
    "GapSymbol                            = -                                   \n",
    "[ ProcessTypes ]\n",
    "ppInfer                              = true                                \n",
    "ppML                                 = true                                \n",
    "[ AnalysisSettings ]\n",
    "Analysis                             = Phylogeny Reconstruction            \n",
    "Statistical Method                   = Maximum Likelihood                  \n",
    "Phylogeny Test                       = ====================                \n",
    "Test of Phylogeny                    = Bootstrap method                    \n",
    "No. of Bootstrap Replications        = 500                                 \n",
    "Substitution Model                   = ====================                \n",
    "Substitutions Type                   = Nucleotide                          \n",
    "Model/Method                         = Kimura 2-parameter model            \n",
    "Rates and Patterns                   = ====================                \n",
    "Rates among Sites                    = Uniform Rates                       \n",
    "No of Discrete Gamma Categories      = Not Applicable                      \n",
    "Data Subset to Use                   = ====================                \n",
    "Gaps/Missing Data Treatment          = Complete deletion                   \n",
    "Site Coverage Cutoff (%)             = Not Applicable                      \n",
    "Select Codon Positions               = 1st, 2nd, 3rd, Non-Coding           \n",
    "Tree Inference Options               = ====================                \n",
    "ML Heuristic Method                  = Nearest-Neighbor-Interchange (NNI)  \n",
    "Initial Tree for ML                  = Make initial tree automatically (Neighbor Joining)\n",
    "Branch Swap Filter                   = Strong                              \n",
    "System Resource Usage                = ====================                \n",
    "Number of Threads                    = 1                                   \n",
    "Genetic Code Table                   = Not Applicable                      \n",
    "Genetic Code                         = Not Applicable                      \n",
    "Has Time Limit                       = False                               \n",
    "Maximum Execution Time               = -1  \"\"\")\n",
    "\n",
    "epigenetec_trees_results = []\n",
    "\n",
    "# запуск меги\n",
    "for family, current_alignment_file in epigenetec_alignment_results:\n",
    "    current_nwk_file = f'data/epigenetic_genes_tree_{family}.nwk'\n",
    "    logging.info(f'Running tree building for {family}')\n",
    "    os.system(f'rm -f {current_nwk_file}')\n",
    "    os.system(f'./local_programs/mega/megacc -a data/nj_tree.mao -d {current_alignment_file} -o {current_nwk_file} > /dev/null 2>&1')\n",
    "    epigenetec_trees_results.append((family, current_nwk_file))\n",
    "    logging.info(f'Tree building for {family} done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 390,
   "id": "1b82213b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ACINU:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" style=\"fill-opacity:1; color-rendering:auto; color-interpolation:auto; text-rendering:auto; stroke:black; stroke-linecap:square; stroke-miterlimit:10; shape-rendering:auto; stroke-opacity:1; fill:black; stroke-dasharray:none; font-weight:normal; stroke-width:1; font-family:'Dialog'; font-style:normal; stroke-linejoin:miter; font-size:12; stroke-dashoffset:0; image-rendering:auto;\" width=\"800.0\" viewBox=\"0 0 800.0 600.0\" height=\"600.0\"><!--Generated by the Batik Graphics2D SVG Generator--><defs id=\"genericDefs\"/><g><g transform=\"translate(6,6)\" style=\"text-rendering:optimizeLegibility; font-size:10; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><line y2=\"570\" style=\"fill:none;\" x1=\"275.3559\" x2=\"366.9441\" y1=\"570\"/><text x=\"314.2\" xml:space=\"preserve\" y=\"583\" style=\"stroke:none;\">0.2</text><path style=\"fill:none;\" d=\"M642.3 114 L220.7926 114 L220.7926 168.5\"/><path style=\"fill:none;\" d=\"M281.4373 332 L272.6824 332 L272.6824 386.5\"/><path style=\"fill:none;\" d=\"M91.763 468.25 L6.3594 468.25 L6.3594 213.9167\"/><path style=\"fill:none;\" d=\"M0 213.9167 L6.3594 213.9167\"/><path style=\"fill:none;\" d=\"M272.6824 386.5 L91.763 386.5 L91.763 468.25\"/><path style=\"fill:none;\" d=\"M220.7926 168.5 L6.3594 168.5 L6.3594 213.9167\"/><path style=\"fill:none;\" d=\"M278.6319 441 L272.6824 441 L272.6824 386.5\"/><path style=\"fill:none;\" d=\"M520.2275 5 L6.3594 5 L6.3594 213.9167\"/><path style=\"fill:none;\" d=\"M405.6898 223 L220.7926 223 L220.7926 168.5\"/><path style=\"fill:none;\" d=\"M441.3502 550 L91.763 550 L91.763 468.25\"/><text x=\"0\" y=\"6\" transform=\"matrix(1,0,0,1,652.3,109)\" style=\"font-size:8; stroke:none;\" xml:space=\"preserve\">spindalis:gene-SSPN1 LOCUS6014</text></g><g transform=\"matrix(1,0,0,1,297.4373,333)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mattheei:gene-SMTH1 LOCUS6929</text></g><g transform=\"matrix(1,0,0,1,294.6319,442)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS6163</text></g><g transform=\"matrix(1,0,0,1,536.2276,6)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mansoni:gene-Smp 062720.1</text></g><g transform=\"matrix(1,0,0,1,421.6899,224)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mekongi:gene-MN116 000671</text></g><g transform=\"matrix(1,0,0,1,457.3502,551)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">japonicum:gene-KSF78 0008119</text></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'AEBP2:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" style=\"fill-opacity:1; color-rendering:auto; color-interpolation:auto; text-rendering:auto; stroke:black; stroke-linecap:square; stroke-miterlimit:10; shape-rendering:auto; stroke-opacity:1; fill:black; stroke-dasharray:none; font-weight:normal; stroke-width:1; font-family:'Dialog'; font-style:normal; stroke-linejoin:miter; font-size:12; stroke-dashoffset:0; image-rendering:auto;\" width=\"800.0\" viewBox=\"0 0 800.0 600.0\" height=\"600.0\"><!--Generated by the Batik Graphics2D SVG Generator--><defs id=\"genericDefs\"/><g><g transform=\"translate(6,6)\" style=\"text-rendering:optimizeLegibility; font-size:10; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><line y2=\"570\" style=\"fill:none;\" x1=\"288.7318\" x2=\"352.8345\" y1=\"570\"/><text x=\"313.8332\" xml:space=\"preserve\" y=\"583\" style=\"stroke:none;\">0.1</text><path style=\"fill:none;\" d=\"M496.7963 114 L203.9476 114 L203.9476 168.5\"/><path style=\"fill:none;\" d=\"M90.0076 86.75 L6.3521 86.75 L6.3521 341.0833\"/><path style=\"fill:none;\" d=\"M580.4449 550 L6.3521 550 L6.3521 341.0833\"/><path style=\"fill:none;\" d=\"M0 341.0833 L6.3521 341.0833\"/><path style=\"fill:none;\" d=\"M613.1941 441 L86.9694 441 L86.9694 386.5\"/><path style=\"fill:none;\" d=\"M203.9476 168.5 L90.0076 168.5 L90.0076 86.75\"/><path style=\"fill:none;\" d=\"M641.5663 332 L86.9694 332 L86.9694 386.5\"/><path style=\"fill:none;\" d=\"M602.8011 5 L90.0076 5 L90.0076 86.75\"/><path style=\"fill:none;\" d=\"M248.6321 223 L203.9476 223 L203.9476 168.5\"/><path style=\"fill:none;\" d=\"M86.9694 386.5 L6.3521 386.5 L6.3521 341.0833\"/><text x=\"0\" y=\"6\" transform=\"matrix(1,0,0,1,506.7962,109)\" style=\"font-size:8; stroke:none;\" xml:space=\"preserve\">mansoni:gene-Smp 155070</text></g><g transform=\"matrix(1,0,0,1,596.4449,551)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mekongi:gene-MN116 001510</text></g><g transform=\"matrix(1,0,0,1,629.1941,442)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">japonicum:gene-KSF78 0005440</text></g><g transform=\"matrix(1,0,0,1,657.5663,333)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mattheei:gene-SMTH1 LOCUS10615</text></g><g transform=\"matrix(1,0,0,1,618.8011,6)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS8012</text></g><g transform=\"matrix(1,0,0,1,264.6321,224)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">spindalis:gene-SSPN1 LOCUS9217</text></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ARID1A:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" style=\"fill-opacity:1; color-rendering:auto; color-interpolation:auto; text-rendering:auto; stroke:black; stroke-linecap:square; stroke-miterlimit:10; shape-rendering:auto; stroke-opacity:1; fill:black; stroke-dasharray:none; font-weight:normal; stroke-width:1; font-family:'Dialog'; font-style:normal; stroke-linejoin:miter; font-size:12; stroke-dashoffset:0; image-rendering:auto;\" width=\"800.0\" viewBox=\"0 0 800.0 600.0\" height=\"600.0\"><!--Generated by the Batik Graphics2D SVG Generator--><defs id=\"genericDefs\"/><g><g transform=\"translate(6,6)\" style=\"text-rendering:optimizeLegibility; font-size:10; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><line y2=\"570\" style=\"fill:none;\" x1=\"288.6413\" x2=\"353.0116\" y1=\"570\"/><text x=\"313.8764\" xml:space=\"preserve\" y=\"583\" style=\"stroke:none;\">0.1</text><path style=\"fill:none;\" d=\"M522.5929 114 L75.46 114 L75.46 59.5\"/><path style=\"fill:none;\" d=\"M544.1515 332 L531.7254 332 L531.7254 277.5\"/><path style=\"fill:none;\" d=\"M641.6529 550 L6.353 550 L6.353 322.9167\"/><path style=\"fill:none;\" d=\"M0 322.9167 L6.353 322.9167\"/><path style=\"fill:none;\" d=\"M297.6872 441 L76.6418 441 L76.6418 359.25\"/><path style=\"fill:none;\" d=\"M547.0037 223 L531.7254 223 L531.7254 277.5\"/><path style=\"fill:none;\" d=\"M531.7254 277.5 L76.6418 277.5 L76.6418 359.25\"/><path style=\"fill:none;\" d=\"M468.2149 5 L75.46 5 L75.46 59.5\"/><path style=\"fill:none;\" d=\"M75.46 59.5 L6.353 59.5 L6.353 322.9167\"/><path style=\"fill:none;\" d=\"M76.6418 359.25 L6.353 359.25 L6.353 322.9167\"/><text x=\"0\" y=\"6\" transform=\"matrix(1,0,0,1,532.5929,109)\" style=\"font-size:8; stroke:none;\" xml:space=\"preserve\">mekongi:gene-MN116 007544</text></g><g transform=\"matrix(1,0,0,1,560.1515,333)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mattheei:gene-SMTH1 LOCUS7396</text></g><g transform=\"matrix(1,0,0,1,657.6529,551)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">spindalis:gene-SSPN1 LOCUS8855</text></g><g transform=\"matrix(1,0,0,1,313.6872,442)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">japonicum:gene-KSF78 0007878</text></g><g transform=\"matrix(1,0,0,1,563.0037,224)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS6554</text></g><g transform=\"matrix(1,0,0,1,484.2149,6)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mansoni:gene-Smp 157150</text></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ATR:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" style=\"fill-opacity:1; color-rendering:auto; color-interpolation:auto; text-rendering:auto; stroke:black; stroke-linecap:square; stroke-miterlimit:10; shape-rendering:auto; stroke-opacity:1; fill:black; stroke-dasharray:none; font-weight:normal; stroke-width:1; font-family:'Dialog'; font-style:normal; stroke-linejoin:miter; font-size:12; stroke-dashoffset:0; image-rendering:auto;\" width=\"800.0\" viewBox=\"0 0 800.0 600.0\" height=\"600.0\"><!--Generated by the Batik Graphics2D SVG Generator--><defs id=\"genericDefs\"/><g><g transform=\"translate(6,6)\" style=\"text-rendering:optimizeLegibility; font-size:10; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><line y2=\"570\" style=\"fill:none;\" x1=\"286.6904\" x2=\"354.8433\" y1=\"570\"/><text x=\"313.8169\" xml:space=\"preserve\" y=\"583\" style=\"stroke:none;\">0.1</text><path style=\"fill:none;\" d=\"M542.7433 114 L113.9758 114 L113.9758 236.625\"/><path style=\"fill:none;\" d=\"M641.5338 550 L6.3518 550 L6.3518 263.875\"/><path style=\"fill:none;\" d=\"M0 263.875 L6.3518 263.875\"/><path style=\"fill:none;\" d=\"M262.1738 359.25 L113.9758 359.25 L113.9758 236.625\"/><path style=\"fill:none;\" d=\"M622.5102 332 L592.3981 332 L592.3981 277.5\"/><path style=\"fill:none;\" d=\"M592.3981 277.5 L262.1738 277.5 L262.1738 359.25\"/><path style=\"fill:none;\" d=\"M567.7388 441 L262.1738 441 L262.1738 359.25\"/><path style=\"fill:none;\" d=\"M541.4089 5 L6.3518 5 L6.3518 263.875\"/><path style=\"fill:none;\" d=\"M619.7191 223 L592.3981 223 L592.3981 277.5\"/><path style=\"fill:none;\" d=\"M113.9758 236.625 L6.3518 236.625 L6.3518 263.875\"/><text x=\"0\" y=\"6\" transform=\"matrix(1,0,0,1,552.7433,109)\" style=\"font-size:8; stroke:none;\" xml:space=\"preserve\">spindalis:gene-SSPN1 LOCUS8606</text></g><g transform=\"matrix(1,0,0,1,657.5338,551)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mansoni:gene-Smp 162340</text></g><g transform=\"matrix(1,0,0,1,638.5102,333)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS8890</text></g><g transform=\"matrix(1,0,0,1,583.7388,442)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mekongi:gene-MN116 002008</text></g><g transform=\"matrix(1,0,0,1,557.4089,6)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">japonicum:gene-KSF78 0005441</text></g><g transform=\"matrix(1,0,0,1,635.7191,224)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mattheei:gene-SMTH1 LOCUS10084</text></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ATRX:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" style=\"fill-opacity:1; color-rendering:auto; color-interpolation:auto; text-rendering:auto; stroke:black; stroke-linecap:square; stroke-miterlimit:10; shape-rendering:auto; stroke-opacity:1; fill:black; stroke-dasharray:none; font-weight:normal; stroke-width:1; font-family:'Dialog'; font-style:normal; stroke-linejoin:miter; font-size:12; stroke-dashoffset:0; image-rendering:auto;\" width=\"800.0\" viewBox=\"0 0 800.0 600.0\" height=\"600.0\"><!--Generated by the Batik Graphics2D SVG Generator--><defs id=\"genericDefs\"/><g><g transform=\"translate(6,6)\" style=\"text-rendering:optimizeLegibility; font-size:10; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><line y2=\"570\" style=\"fill:none;\" x1=\"273.3199\" x2=\"368.4901\" y1=\"570\"/><text x=\"313.955\" xml:space=\"preserve\" y=\"583\" style=\"stroke:none;\">0.2</text><path style=\"fill:none;\" d=\"M403.3377 114 L230.1084 114 L230.1084 168.5\"/><path style=\"fill:none;\" d=\"M93.8389 86.75 L47.295 86.75 L47.295 209.375\"/><path style=\"fill:none;\" d=\"M476.1759 550 L6.3546 550 L6.3546 400.125\"/><path style=\"fill:none;\" d=\"M0 400.125 L6.3546 400.125\"/><path style=\"fill:none;\" d=\"M47.295 209.375 L6.3546 209.375 L6.3546 400.125\"/><path style=\"fill:none;\" d=\"M230.1084 168.5 L93.8389 168.5 L93.8389 86.75\"/><path style=\"fill:none;\" d=\"M497.0913 332 L47.295 332 L47.295 209.375\"/><path style=\"fill:none;\" d=\"M292.6081 5 L93.8389 5 L93.8389 86.75\"/><path style=\"fill:none;\" d=\"M641.81 223 L230.1084 223 L230.1084 168.5\"/><path style=\"fill:none;\" d=\"M352.8861 441 L6.3546 441 L6.3546 400.125\"/><text x=\"0\" y=\"6\" transform=\"matrix(1,0,0,1,413.3377,109)\" style=\"font-size:8; stroke:none;\" xml:space=\"preserve\">mattheei:gene-SMTH1 LOCUS10333</text></g><g transform=\"matrix(1,0,0,1,492.1759,551)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mekongi:gene-MN116 001315</text></g><g transform=\"matrix(1,0,0,1,513.0914,333)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS5228</text></g><g transform=\"matrix(1,0,0,1,308.6081,6)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">spindalis:gene-SSPN1 LOCUS9160</text></g><g transform=\"matrix(1,0,0,1,657.81,224)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mansoni:gene-Smp 163110</text></g><g transform=\"matrix(1,0,0,1,368.8861,442)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">japonicum:gene-KSF78 0005543</text></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'BAP1:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" style=\"fill-opacity:1; color-rendering:auto; color-interpolation:auto; text-rendering:auto; stroke:black; stroke-linecap:square; stroke-miterlimit:10; shape-rendering:auto; stroke-opacity:1; fill:black; stroke-dasharray:none; font-weight:normal; stroke-width:1; font-family:'Dialog'; font-style:normal; stroke-linejoin:miter; font-size:12; stroke-dashoffset:0; image-rendering:auto;\" width=\"800.0\" viewBox=\"0 0 800.0 600.0\" height=\"600.0\"><!--Generated by the Batik Graphics2D SVG Generator--><defs id=\"genericDefs\"/><g><g transform=\"translate(6,6)\" style=\"text-rendering:optimizeLegibility; font-size:10; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><line y2=\"570\" style=\"fill:none;\" x1=\"290.4844\" x2=\"355.4941\" y1=\"570\"/><text x=\"316.0393\" xml:space=\"preserve\" y=\"583\" style=\"stroke:none;\">0.1</text><path style=\"fill:none;\" d=\"M186.4399 114 L165.0082 114 L165.0082 59.5\"/><path style=\"fill:none;\" d=\"M533.6126 332 L211.5626 332 L211.5626 386.5\"/><path style=\"fill:none;\" d=\"M519.6104 550 L6.3958 550 L6.3958 304.75\"/><path style=\"fill:none;\" d=\"M0 304.75 L6.3958 304.75\"/><path style=\"fill:none;\" d=\"M211.5626 386.5 L101.528 386.5 L101.528 304.75\"/><path style=\"fill:none;\" d=\"M645.9785 223 L101.528 223 L101.528 304.75\"/><path style=\"fill:none;\" d=\"M629.8897 441 L211.5626 441 L211.5626 386.5\"/><path style=\"fill:none;\" d=\"M183.7094 5 L165.0082 5 L165.0082 59.5\"/><path style=\"fill:none;\" d=\"M165.0082 59.5 L6.3958 59.5 L6.3958 304.75\"/><path style=\"fill:none;\" d=\"M101.528 304.75 L6.3958 304.75 L6.3958 304.75\"/><text x=\"0\" y=\"6\" transform=\"matrix(1,0,0,1,196.4399,109)\" style=\"font-size:8; stroke:none;\" xml:space=\"preserve\">mansoni:gene-Smp 168800.1</text></g><g transform=\"matrix(1,0,0,1,549.6126,333)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">spindalis:gene-SSPN1 LOCUS3856</text></g><g transform=\"matrix(1,0,0,1,535.6104,551)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">japonicum:gene-KSF78 0005343</text></g><g transform=\"matrix(1,0,0,1,661.9785,224)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mekongi:gene-MN116 003045</text></g><g transform=\"matrix(1,0,0,1,645.8897,442)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS956</text></g><g transform=\"matrix(1,0,0,1,199.7094,6)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mattheei:gene-SMTH1 LOCUS4562</text></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'BARD1:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" style=\"fill-opacity:1; color-rendering:auto; color-interpolation:auto; text-rendering:auto; stroke:black; stroke-linecap:square; stroke-miterlimit:10; shape-rendering:auto; stroke-opacity:1; fill:black; stroke-dasharray:none; font-weight:normal; stroke-width:1; font-family:'Dialog'; font-style:normal; stroke-linejoin:miter; font-size:12; stroke-dashoffset:0; image-rendering:auto;\" width=\"800.0\" viewBox=\"0 0 800.0 600.0\" height=\"600.0\"><!--Generated by the Batik Graphics2D SVG Generator--><defs id=\"genericDefs\"/><g><g transform=\"translate(6,6)\" style=\"text-rendering:optimizeLegibility; font-size:10; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><line y2=\"570\" style=\"fill:none;\" x1=\"272.8924\" x2=\"368.6708\" y1=\"570\"/><text x=\"313.8316\" xml:space=\"preserve\" y=\"583\" style=\"stroke:none;\">0.2</text><path style=\"fill:none;\" d=\"M641.5632 114 L599.3885 114 L599.3885 168.5\"/><path style=\"fill:none;\" d=\"M233.3173 86.75 L109.3305 86.75 L109.3305 209.375\"/><path style=\"fill:none;\" d=\"M553.2234 550 L6.3521 550 L6.3521 400.125\"/><path style=\"fill:none;\" d=\"M0 400.125 L6.3521 400.125\"/><path style=\"fill:none;\" d=\"M109.3305 209.375 L6.3521 209.375 L6.3521 400.125\"/><path style=\"fill:none;\" d=\"M599.3885 168.5 L233.3173 168.5 L233.3173 86.75\"/><path style=\"fill:none;\" d=\"M556.0739 332 L109.3305 332 L109.3305 209.375\"/><path style=\"fill:none;\" d=\"M378.9092 5 L233.3173 5 L233.3173 86.75\"/><path style=\"fill:none;\" d=\"M614.0931 223 L599.3885 223 L599.3885 168.5\"/><path style=\"fill:none;\" d=\"M247.2247 441 L6.3521 441 L6.3521 400.125\"/><text x=\"0\" y=\"6\" transform=\"matrix(1,0,0,1,651.5632,109)\" style=\"font-size:8; stroke:none;\" xml:space=\"preserve\">mansoni:gene-Smp 137090</text></g><g transform=\"matrix(1,0,0,1,569.2234,551)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">japonicum:gene-KSF78 0006244</text></g><g transform=\"matrix(1,0,0,1,572.0738,333)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mattheei:gene-SMTH1 LOCUS10644</text></g><g transform=\"matrix(1,0,0,1,394.9092,6)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">spindalis:gene-SSPN1 LOCUS9194</text></g><g transform=\"matrix(1,0,0,1,630.0931,224)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS4989</text></g><g transform=\"matrix(1,0,0,1,263.2247,442)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mekongi:gene-MN116 004107</text></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CBX4:'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" style=\"fill-opacity:1; color-rendering:auto; color-interpolation:auto; text-rendering:auto; stroke:black; stroke-linecap:square; stroke-miterlimit:10; shape-rendering:auto; stroke-opacity:1; fill:black; stroke-dasharray:none; font-weight:normal; stroke-width:1; font-family:'Dialog'; font-style:normal; stroke-linejoin:miter; font-size:12; stroke-dashoffset:0; image-rendering:auto;\" width=\"800.0\" viewBox=\"0 0 800.0 600.0\" height=\"600.0\"><!--Generated by the Batik Graphics2D SVG Generator--><defs id=\"genericDefs\"/><g><g transform=\"translate(6,6)\" style=\"text-rendering:optimizeLegibility; font-size:10; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><line y2=\"570\" style=\"fill:none;\" x1=\"279.141\" x2=\"363.3526\" y1=\"570\"/><text x=\"314.2968\" xml:space=\"preserve\" y=\"583\" style=\"stroke:none;\">0.2</text><path style=\"fill:none;\" d=\"M411.1345 114 L191.9751 114 L191.9751 236.625\"/><path style=\"fill:none;\" d=\"M451.9069 550 L6.3613 550 L6.3613 263.875\"/><path style=\"fill:none;\" d=\"M0 263.875 L6.3613 263.875\"/><path style=\"fill:none;\" d=\"M362.7004 359.25 L191.9751 359.25 L191.9751 236.625\"/><path style=\"fill:none;\" d=\"M584.2135 332 L562.2617 332 L562.2617 277.5\"/><path style=\"fill:none;\" d=\"M562.2617 277.5 L362.7004 277.5 L362.7004 359.25\"/><path style=\"fill:none;\" d=\"M642.4937 441 L362.7004 441 L362.7004 359.25\"/><path style=\"fill:none;\" d=\"M226.5649 5 L6.3613 5 L6.3613 263.875\"/><path style=\"fill:none;\" d=\"M570.4294 223 L562.2617 223 L562.2617 277.5\"/><path style=\"fill:none;\" d=\"M191.9751 236.625 L6.3613 236.625 L6.3613 263.875\"/><text x=\"0\" y=\"6\" transform=\"matrix(1,0,0,1,421.1345,109)\" style=\"font-size:8; stroke:none;\" xml:space=\"preserve\">japonicum:gene-KSF78 0005240</text></g><g transform=\"matrix(1,0,0,1,467.9068,551)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mattheei:gene-SMTH1 LOCUS10167</text></g><g transform=\"matrix(1,0,0,1,600.2135,333)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">spindalis:gene-SSPN1 LOCUS9023</text></g><g transform=\"matrix(1,0,0,1,658.4937,442)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mekongi:gene-MN116 001223</text></g><g transform=\"matrix(1,0,0,1,242.5649,6)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">intercalatum:gene-SINT2 LOCUS6604</text></g><g transform=\"matrix(1,0,0,1,586.4294,224)\" style=\"text-rendering:optimizeLegibility; font-size:8; font-family:sans-serif; stroke-linecap:round; stroke-linejoin:round;\"><text x=\"0\" xml:space=\"preserve\" y=\"6\" style=\"stroke:none;\">mansoni:gene-Smp 078280</text></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# картинки деревьев\n",
    "from IPython.display import SVG, display\n",
    "import os\n",
    "\n",
    "for family, epigenetec_trees_result in epigenetec_trees_results:\n",
    "    current_image_file = f'data/epigenetic_genes_tree_{family}.svg'\n",
    "    os.system(f'figtree -graphic SVG {epigenetec_trees_result} {current_image_file} > /dev/null 2>&1')\n",
    "    display(f'{family}:')\n",
    "    display(SVG(filename=current_image_file))\n",
    "\n",
    "# более красивые картинки в отчете (тоже figtree, только через gui с настройками)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "392245fa",
   "metadata": {},
   "source": [
    "### 2 и 3: Ищем ортологи"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 489,
   "id": "e815320d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:local_programs/OrthoFinder_source.tar.gz alreasy exists\n",
      "INFO:root:local_programs/OrthoFinder_source alreasy exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OrthoFinder version 3.0.1b1 Copyright (C) 2014 David Emms\n",
      "\n",
      "SIMPLE USAGE:\n",
      "Run full OrthoFinder analysis on FASTA format proteomes in <dir>\n",
      "  orthofinder [options] -f <dir>\n",
      "\n",
      "To assign species from <dir1> to existing OrthoFinder orthogroups in <dir2>\n",
      "  orthofinder [options] --assign <dir1> --core <dir2>\n",
      "\n",
      "OPTIONS:\n",
      " -t <int>        Number of parallel sequence search threads [Default = 12]\n",
      " -a <int>        Number of parallel analysis threads\n",
      " -d              Input is DNA sequences\n",
      " -M <txt>        Method for gene tree inference. Options 'dendroblast' & 'msa'\n",
      "                 [Default = msa]\n",
      " -S <txt>        Sequence search program [Default = diamond]\n",
      "                 Options: blast, diamond, diamond_ultra_sens, blast_gz, mmseqs, blast_nucl\n",
      " -A <txt>        MSA program, requires '-M msa' [Default = mafft]\n",
      "                 Options: mafft, muscle, mafft_memsave\n",
      " -T <txt>        Tree inference method, requires '-M msa' [Default = fasttree]\n",
      "                 Options: fasttree, fasttree_fastest, raxml, raxml-ng, iqtree\n",
      " -s <file>       User-specified rooted species tree\n",
      " -I <int>        MCL inflation parameter [Default = 1.2]\n",
      " --save-space    Only create one compressed orthologs file per species\n",
      " -x <file>       Info for outputting results in OrthoXML format\n",
      " -p <dir>        Write the temporary pickle files to <dir>\n",
      " -1              Only perform one-way sequence search\n",
      " -X              Don't add species names to sequence IDs\n",
      " -y              Split paralogous clades below root of a HOG into separate HOGs\n",
      " -z              Don't trim MSAs (columns>=90% gap, min. alignment length 500)\n",
      " -n <txt>        Name to append to the results directory\n",
      " -o <txt>        Non-default results directory\n",
      " -h              Print this help text\n",
      "\n",
      "WORKFLOW STOPPING OPTIONS:\n",
      " -op             Stop after preparing input files for BLAST\n",
      " -og             Stop after inferring orthogroups\n",
      " -os             Stop after writing sequence files for orthogroups\n",
      "                 (requires '-M msa')\n",
      " -oa             Stop after inferring alignments for orthogroups\n",
      "                 (requires '-M msa')\n",
      " -ot             Stop after inferring gene trees for orthogroups \n",
      "\n",
      "WORKFLOW RESTART COMMANDS:\n",
      " -b  <dir>         Start OrthoFinder from pre-computed BLAST results in <dir>\n",
      " -fg <dir>         Start OrthoFinder from pre-computed orthogroups in <dir>\n",
      " -ft <dir>         Start OrthoFinder from pre-computed gene trees in <dir>\n",
      "\n",
      "LICENSE:\n",
      " Distributed under the GNU General Public License (GPLv3). See License.md\n",
      "\n",
      "CITATION:\n",
      " When publishing work that uses OrthoFinder please cite:\n",
      " Emms D.M. & Kelly S. (2019), Genome Biology 20:238\n",
      "\n",
      " If you use the species tree in your work then please also cite:\n",
      " Emms D.M. & Kelly S. (2017), MBE 34(12): 3267-3278\n",
      " Emms D.M. & Kelly S. (2018), bioRxiv https://doi.org/10.1101/267914\n"
     ]
    }
   ],
   "source": [
    "# установка orthofinder\n",
    "\n",
    "import logging\n",
    "import pathlib\n",
    "\n",
    "if pathlib.Path('local_programs/OrthoFinder_source.tar.gz').is_file():\n",
    "    logging.info('local_programs/OrthoFinder_source.tar.gz alreasy exists')\n",
    "else:\n",
    "    logging.info('Downloading OrthoFinder_source.tar.gz')\n",
    "    ! cd local_programs/ && wget https://github.com/davidemms/OrthoFinder/releases/download/3.0.1b1/OrthoFinder_source.tar.gz\n",
    "\n",
    "if pathlib.Path('local_programs/OrthoFinder_source').is_dir():\n",
    "    logging.info('local_programs/OrthoFinder_source alreasy exists')\n",
    "else:\n",
    "    logging.info('Extracting OrthoFinder_source.tar.gz')\n",
    "    ! cd local_programs/ && tar -xzf OrthoFinder_source.tar.gz\n",
    "\n",
    "! python local_programs/OrthoFinder_source/orthofinder.py -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 490,
   "id": "2900bd95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bovis.faa\t intercalatum.faa  mansoni.faa\t mekongi.faa\n",
      "haematobium.faa  japonicum.faa\t   mattheei.faa  spindalis.faa\n"
     ]
    }
   ],
   "source": [
    "# готовим файлы для поиска ортологов (нужно положить их в отдельную папку)\n",
    "\n",
    "import os\n",
    "\n",
    "! mkdir -p data/proteom\n",
    "for organism, proteom_file in proteom_files.items():\n",
    "    os.system(f'cp {proteom_file} data/proteom/{organism}.faa')\n",
    "\n",
    "! ls data/proteom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 491,
   "id": "7c6f6014",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OrthoFinder version 3.0.1b1 Copyright (C) 2014 David Emms\n",
      "\n",
      "2025-06-18 23:00:43 : Starting OrthoFinder 3.0.1b1\n",
      "6 thread(s) for highly parallel tasks (BLAST searches etc.)\n",
      "6 thread(s) for OrthoFinder algorithm\n",
      "\n",
      "Results directory: /home/andrejj/code/hse/bio/project_common/bioinf-project-common/data/proteom/OrthoFinder/Results_Jun18/\n",
      "\n",
      "Checking required programs are installed\n",
      "----------------------------------------\n",
      "Running with the recommended MSA tree inference by default. To revert to legacy method use '-M dendroblast'.\n",
      "\n",
      "Test can run \"mcl -h\" - ok\n",
      "Test can run \"mafft\" - failed\n",
      "\n",
      "/bin/sh: line 1: mafft: command not found\n",
      "\n",
      "Trying OrthoFinder packaged version of MAFFT\n",
      "\n",
      "Test can run \"mafft\" - ok\n",
      "Test can run \"fasttree\" - ok\n",
      "\n",
      "Dividing up work for BLAST for parallel processing\n",
      "--------------------------------------------------\n",
      "2025-06-18 23:00:44 : Creating diamond database 1 of 8\n",
      "2025-06-18 23:00:44 : Creating diamond database 2 of 8\n",
      "2025-06-18 23:00:44 : Creating diamond database 3 of 8\n",
      "2025-06-18 23:00:44 : Creating diamond database 4 of 8\n",
      "2025-06-18 23:00:44 : Creating diamond database 5 of 8\n",
      "2025-06-18 23:00:45 : Creating diamond database 6 of 8\n",
      "2025-06-18 23:00:45 : Creating diamond database 7 of 8\n",
      "2025-06-18 23:00:45 : Creating diamond database 8 of 8\n",
      "\n",
      "Running diamond all-versus-all\n",
      "------------------------------\n",
      "Using 6 thread(s)\n",
      "2025-06-18 23:00:45 : This may take some time....\n",
      "2025-06-18 23:00:45 : Done 0 of 64\n",
      "2025-06-18 23:05:55 : Done 10 of 64\n",
      "2025-06-18 23:09:54 : Done 20 of 64\n",
      "2025-06-18 23:14:15 : Done 30 of 64\n",
      "2025-06-18 23:17:14 : Done 40 of 64\n",
      "2025-06-18 23:19:39 : Done 50 of 64\n",
      "2025-06-18 23:22:37 : Done all-versus-all sequence search\n",
      "\n",
      "Running OrthoFinder algorithm\n",
      "-----------------------------\n",
      "2025-06-18 23:22:38 : Initial processing of each species\n",
      "2025-06-18 23:22:44 : Initial processing of species 4 complete\n",
      "2025-06-18 23:22:46 : Initial processing of species 5 complete\n",
      "2025-06-18 23:22:46 : Initial processing of species 1 complete\n",
      "2025-06-18 23:22:47 : Initial processing of species 2 complete\n",
      "2025-06-18 23:22:47 : Initial processing of species 0 complete\n",
      "2025-06-18 23:22:49 : Initial processing of species 6 complete\n",
      "2025-06-18 23:22:53 : Initial processing of species 3 complete\n",
      "2025-06-18 23:22:54 : Initial processing of species 7 complete\n",
      "2025-06-18 23:22:58 : Connected putative homologues\n",
      "2025-06-18 23:23:00 : Written final scores for species 4 to graph file\n",
      "2025-06-18 23:23:00 : Written final scores for species 5 to graph file\n",
      "2025-06-18 23:23:00 : Written final scores for species 1 to graph file\n",
      "2025-06-18 23:23:00 : Written final scores for species 2 to graph file\n",
      "2025-06-18 23:23:00 : Written final scores for species 0 to graph file\n",
      "2025-06-18 23:23:01 : Written final scores for species 6 to graph file\n",
      "2025-06-18 23:23:01 : Written final scores for species 3 to graph file\n",
      "2025-06-18 23:23:01 : Written final scores for species 7 to graph file\n",
      "2025-06-18 23:23:16 : Ran MCL\n",
      "\n",
      "Writing orthogroups to file\n",
      "---------------------------\n",
      "\n",
      "2025-06-18 23:23:18 : Done orthogroups\n",
      "\n",
      "Analysing Orthogroups\n",
      "=====================\n",
      "2025-06-18 23:23:18 : Starting MSA/Trees\n",
      "Species tree: Using 1722 orthogroups with minimum of 87.5% of species having single-copy genes in any orthogroup\n",
      "\n",
      "Inferring multiple sequence alignments for species tree\n",
      "-------------------------------------------------------\n",
      "2025-06-18 23:23:20 : Done 0 of 1722\n",
      "2025-06-18 23:23:33 : Done 100 of 1722\n",
      "2025-06-18 23:23:49 : Done 200 of 1722\n",
      "2025-06-18 23:24:02 : Done 300 of 1722\n",
      "2025-06-18 23:24:21 : Done 400 of 1722\n",
      "2025-06-18 23:24:34 : Done 500 of 1722\n",
      "2025-06-18 23:24:49 : Done 600 of 1722\n",
      "2025-06-18 23:25:01 : Done 700 of 1722\n",
      "2025-06-18 23:25:13 : Done 800 of 1722\n",
      "2025-06-18 23:25:28 : Done 900 of 1722\n",
      "2025-06-18 23:25:40 : Done 1000 of 1722\n",
      "2025-06-18 23:25:53 : Done 1100 of 1722\n",
      "2025-06-18 23:26:04 : Done 1200 of 1722\n",
      "2025-06-18 23:26:15 : Done 1300 of 1722\n",
      "2025-06-18 23:26:27 : Done 1400 of 1722\n",
      "2025-06-18 23:26:35 : Done 1500 of 1722\n",
      "2025-06-18 23:26:42 : Done 1600 of 1722\n",
      "2025-06-18 23:26:49 : Done 1700 of 1722\n",
      "\n",
      "Inferring remaining multiple sequence alignments and gene trees\n",
      "---------------------------------------------------------------\n",
      "2025-06-18 23:26:55 : Done 0 of 7803\n",
      "2025-06-18 23:54:17 : Done 1000 of 7803\n",
      "2025-06-19 00:02:52 : Done 2000 of 7803\n",
      "2025-06-19 00:08:11 : Done 3000 of 7803\n",
      "2025-06-19 00:12:15 : Done 4000 of 7803\n",
      "2025-06-19 00:15:32 : Done 5000 of 7803\n",
      "2025-06-19 00:17:37 : Done 6000 of 7803\n",
      "2025-06-19 00:18:17 : Done 7000 of 7803\n",
      "2025-06-19 00:18:43 : Done MSA/Trees\n",
      "\n",
      "Best outgroup(s) for species tree\n",
      "---------------------------------\n",
      "2025-06-19 00:18:43 : Starting STRIDE\n",
      "2025-06-19 00:18:46 : Done STRIDE\n",
      "Observed 72 well-supported, non-terminal duplications. 45 support the best roots and 27 contradict them.\n",
      "Best outgroups for species tree:\n",
      "  japonicum, mekongi\n",
      "  mansoni\n",
      "\n",
      "WARNING: Multiple potential species tree roots were identified, only one will be analyed.\n",
      "\n",
      "Reconciling gene trees and species tree\n",
      "---------------------------------------\n",
      "Outgroup: japonicum, mekongi\n",
      "2025-06-19 00:18:46 : Starting Recon and orthologues\n",
      "2025-06-19 00:18:46 : Starting OF Orthologues\n",
      "2025-06-19 00:18:46 : Done 0 of 8365\n",
      "2025-06-19 00:18:47 : Done 1000 of 8365\n",
      "2025-06-19 00:18:48 : Done 2000 of 8365\n",
      "2025-06-19 00:18:48 : Done 3000 of 8365\n",
      "2025-06-19 00:18:49 : Done 4000 of 8365\n",
      "2025-06-19 00:18:49 : Done 5000 of 8365\n",
      "2025-06-19 00:18:50 : Done 6000 of 8365\n",
      "2025-06-19 00:18:50 : Done 7000 of 8365\n",
      "2025-06-19 00:18:51 : Done 8000 of 8365\n",
      "2025-06-19 00:18:52 : Done OF Orthologues\n",
      "\n",
      "Writing results files\n",
      "=====================\n",
      "OrthoFinder assigned 120995 genes (95.5% of total) to 10116 orthogroups. Fifty percent of genes were in orthogroups with 15 or more genes (G50 was 15) and were contained in the largest 3337 orthogroups (O50 was 3337). There were 6135 orthogroups with all species present and 496 of these consisted entirely of single-copy genes.\n",
      "2025-06-19 00:18:53 : Done orthologues\n",
      "\n",
      "Results:\n",
      "    /home/andrejj/code/hse/bio/project_common/bioinf-project-common/data/proteom/OrthoFinder/Results_Jun18/\n",
      "\n",
      "CITATION:\n",
      " When publishing work that uses OrthoFinder please cite:\n",
      " Emms D.M. & Kelly S. (2019), Genome Biology 20:238\n",
      "\n",
      " If you use the species tree in your work then please also cite:\n",
      " Emms D.M. & Kelly S. (2017), MBE 34(12): 3267-3278\n",
      " Emms D.M. & Kelly S. (2018), bioRxiv https://doi.org/10.1101/267914\n"
     ]
    }
   ],
   "source": [
    "! python local_programs/OrthoFinder_source/orthofinder.py -f data/proteom/ -t 6 -a 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 500,
   "id": "625bd393",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>bovis</th>\n",
       "      <th>haematobium</th>\n",
       "      <th>intercalatum</th>\n",
       "      <th>japonicum</th>\n",
       "      <th>mansoni</th>\n",
       "      <th>mattheei</th>\n",
       "      <th>mekongi</th>\n",
       "      <th>spindalis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OG0000000</td>\n",
       "      <td>CAH8288778.1, CAH8432533.1, CAH8436195.1, CAH8...</td>\n",
       "      <td>MS3_00000172.1_mrna, MS3_00000206.1_mrna, MS3_...</td>\n",
       "      <td>CAH8289572.1, CAH8289622.1, CAH8292534.1, CAH8...</td>\n",
       "      <td>KAH8852528.1, KAH8875471.1, KAH8876554.1</td>\n",
       "      <td>XP_018644951.1, XP_018645907.1, XP_018646405.1...</td>\n",
       "      <td>CAH8287097.1, CAH8287203.1, CAH8428810.1, CAH8...</td>\n",
       "      <td>KAK4467234.1, KAK4467267.1, KAK4467282.1, KAK4...</td>\n",
       "      <td>CAI2723349.1, CAI2723511.1, CAI2723623.1, CAI2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OG0000001</td>\n",
       "      <td>CAH8429419.1, CAH8452260.1, CAH8452539.1, CAH8...</td>\n",
       "      <td>MS3_00000028.1_mrna, MS3_00000030.1_mrna, MS3_...</td>\n",
       "      <td>CAH8290361.1, CAH8293701.1, CAH8293905.1, CAH8...</td>\n",
       "      <td>KAH8859351.1</td>\n",
       "      <td>XP_018644623.1, XP_018645506.1, XP_018646482.1...</td>\n",
       "      <td>CAH8287211.1, CAH8431624.1, CAH8442757.1, CAH8...</td>\n",
       "      <td>KAK4467952.1, KAK4467960.1, KAK4467973.1, KAK4...</td>\n",
       "      <td>CAI2722898.1, CAI2723109.1, CAI2723267.1, CAI2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG0000002</td>\n",
       "      <td>CAH8289611.1, CAH8432098.1, CAH8464629.1, CAH8...</td>\n",
       "      <td>MS3_00000047.1_mrna, MS3_00000069.1_mrna, MS3_...</td>\n",
       "      <td>CAH8293534.1, CAH8443764.1, CAH8443766.1, CAH8...</td>\n",
       "      <td>KAH8848639.1, KAH8852700.1, KAH8860267.1, KAH8...</td>\n",
       "      <td>XP_018644447.1, XP_018644448.1, XP_018644518.1...</td>\n",
       "      <td>CAH8287158.1, CAH8287294.1, CAH8287308.1, CAH8...</td>\n",
       "      <td>KAK4467259.1, KAK4467315.1, KAK4467347.1, KAK4...</td>\n",
       "      <td>CAI2728672.1, CAI2729273.1, CAI2730080.1, CAI2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OG0000003</td>\n",
       "      <td>CAH8435310.1, CAH8484787.1, CAH8500220.1, CAH8...</td>\n",
       "      <td>MS3_00000062.1_mrna, MS3_00000116.1_mrna, MS3_...</td>\n",
       "      <td>CAH8294001.1, CAH8294006.1, CAH8428806.1, CAH8...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_018645393.1, XP_018655013.1</td>\n",
       "      <td>CAH8287428.1, CAH8287431.1, CAH8287470.1, CAH8...</td>\n",
       "      <td>KAK4467238.1, KAK4467248.1, KAK4467263.1, KAK4...</td>\n",
       "      <td>CAI2722797.1, CAI2722928.1, CAI2722929.1, CAI2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OG0000004</td>\n",
       "      <td>CAH8290394.1, CAH8429702.1, CAH8434049.1, CAH8...</td>\n",
       "      <td>MS3_00000127.1_mrna, MS3_00000181.1_mrna, MS3_...</td>\n",
       "      <td>CAH8290911.1, CAH8293995.1, CAH8434768.1, CAH8...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_018645934.1, XP_018647423.1, XP_018649176.1...</td>\n",
       "      <td>CAH8430565.1, CAH8432021.1, CAH8434563.1, CAH8...</td>\n",
       "      <td>KAK4467292.1, KAK4468373.1, KAK4474221.1, KAK4...</td>\n",
       "      <td>CAI2722400.1, CAI2722967.1, CAI2722973.1, CAI2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Orthogroup                                              bovis  \\\n",
       "0  OG0000000  CAH8288778.1, CAH8432533.1, CAH8436195.1, CAH8...   \n",
       "1  OG0000001  CAH8429419.1, CAH8452260.1, CAH8452539.1, CAH8...   \n",
       "2  OG0000002  CAH8289611.1, CAH8432098.1, CAH8464629.1, CAH8...   \n",
       "3  OG0000003  CAH8435310.1, CAH8484787.1, CAH8500220.1, CAH8...   \n",
       "4  OG0000004  CAH8290394.1, CAH8429702.1, CAH8434049.1, CAH8...   \n",
       "\n",
       "                                         haematobium  \\\n",
       "0  MS3_00000172.1_mrna, MS3_00000206.1_mrna, MS3_...   \n",
       "1  MS3_00000028.1_mrna, MS3_00000030.1_mrna, MS3_...   \n",
       "2  MS3_00000047.1_mrna, MS3_00000069.1_mrna, MS3_...   \n",
       "3  MS3_00000062.1_mrna, MS3_00000116.1_mrna, MS3_...   \n",
       "4  MS3_00000127.1_mrna, MS3_00000181.1_mrna, MS3_...   \n",
       "\n",
       "                                        intercalatum  \\\n",
       "0  CAH8289572.1, CAH8289622.1, CAH8292534.1, CAH8...   \n",
       "1  CAH8290361.1, CAH8293701.1, CAH8293905.1, CAH8...   \n",
       "2  CAH8293534.1, CAH8443764.1, CAH8443766.1, CAH8...   \n",
       "3  CAH8294001.1, CAH8294006.1, CAH8428806.1, CAH8...   \n",
       "4  CAH8290911.1, CAH8293995.1, CAH8434768.1, CAH8...   \n",
       "\n",
       "                                           japonicum  \\\n",
       "0           KAH8852528.1, KAH8875471.1, KAH8876554.1   \n",
       "1                                       KAH8859351.1   \n",
       "2  KAH8848639.1, KAH8852700.1, KAH8860267.1, KAH8...   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "\n",
       "                                             mansoni  \\\n",
       "0  XP_018644951.1, XP_018645907.1, XP_018646405.1...   \n",
       "1  XP_018644623.1, XP_018645506.1, XP_018646482.1...   \n",
       "2  XP_018644447.1, XP_018644448.1, XP_018644518.1...   \n",
       "3                     XP_018645393.1, XP_018655013.1   \n",
       "4  XP_018645934.1, XP_018647423.1, XP_018649176.1...   \n",
       "\n",
       "                                            mattheei  \\\n",
       "0  CAH8287097.1, CAH8287203.1, CAH8428810.1, CAH8...   \n",
       "1  CAH8287211.1, CAH8431624.1, CAH8442757.1, CAH8...   \n",
       "2  CAH8287158.1, CAH8287294.1, CAH8287308.1, CAH8...   \n",
       "3  CAH8287428.1, CAH8287431.1, CAH8287470.1, CAH8...   \n",
       "4  CAH8430565.1, CAH8432021.1, CAH8434563.1, CAH8...   \n",
       "\n",
       "                                             mekongi  \\\n",
       "0  KAK4467234.1, KAK4467267.1, KAK4467282.1, KAK4...   \n",
       "1  KAK4467952.1, KAK4467960.1, KAK4467973.1, KAK4...   \n",
       "2  KAK4467259.1, KAK4467315.1, KAK4467347.1, KAK4...   \n",
       "3  KAK4467238.1, KAK4467248.1, KAK4467263.1, KAK4...   \n",
       "4  KAK4467292.1, KAK4468373.1, KAK4474221.1, KAK4...   \n",
       "\n",
       "                                           spindalis  \n",
       "0  CAI2723349.1, CAI2723511.1, CAI2723623.1, CAI2...  \n",
       "1  CAI2722898.1, CAI2723109.1, CAI2723267.1, CAI2...  \n",
       "2  CAI2728672.1, CAI2729273.1, CAI2730080.1, CAI2...  \n",
       "3  CAI2722797.1, CAI2722928.1, CAI2722929.1, CAI2...  \n",
       "4  CAI2722400.1, CAI2722967.1, CAI2722973.1, CAI2...  "
      ]
     },
     "execution_count": 500,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "orthogroups = pd.read_csv('data/proteom/OrthoFinder/Results_Jun18/Orthogroups/Orthogroups.tsv', sep='\\t')\n",
    "orthogroups.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7152da4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>bovis</th>\n",
       "      <th>haematobium</th>\n",
       "      <th>intercalatum</th>\n",
       "      <th>japonicum</th>\n",
       "      <th>mansoni</th>\n",
       "      <th>mattheei</th>\n",
       "      <th>mekongi</th>\n",
       "      <th>spindalis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OG0000000</td>\n",
       "      <td>[gene-SBOV1_LOCUS2675, gene-SBOV1_LOCUS4335, g...</td>\n",
       "      <td>[MS3_00001738, MS3_00005418, MS3_00000214, MS3...</td>\n",
       "      <td>[gene-SINT2_LOCUS1577, gene-SINT2_LOCUS2576, g...</td>\n",
       "      <td>[gene-KSF78_0004650, gene-KSF78_0001145, gene-...</td>\n",
       "      <td>[gene-Smp_025180, gene-Smp_166950, gene-Smp_19...</td>\n",
       "      <td>[gene-SMTH1_LOCUS3830, gene-SMTH1_LOCUS3089, g...</td>\n",
       "      <td>[gene-MN116_000469, gene-MN116_000157, gene-MN...</td>\n",
       "      <td>[gene-SSPN1_LOCUS1690, gene-SSPN1_LOCUS8011, g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OG0000001</td>\n",
       "      <td>[gene-SBOV1_LOCUS796, gene-SBOV1_LOCUS8840, ge...</td>\n",
       "      <td>[MS3_00000895, MS3_00001105, MS3_00003570, MS3...</td>\n",
       "      <td>[gene-SINT2_LOCUS1120, gene-SINT2_LOCUS1638, g...</td>\n",
       "      <td>[gene-KSF78_0006907]</td>\n",
       "      <td>[gene-Smp_122020, gene-Smp_120330, gene-Smp_18...</td>\n",
       "      <td>[gene-SMTH1_LOCUS10701, gene-SMTH1_LOCUS5153, ...</td>\n",
       "      <td>[gene-MN116_000217, gene-MN116_000421, gene-MN...</td>\n",
       "      <td>[gene-SSPN1_LOCUS1012, gene-SSPN1_LOCUS8827, g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG0000002</td>\n",
       "      <td>[gene-SBOV1_LOCUS6926, gene-SBOV1_LOCUS2070, g...</td>\n",
       "      <td>[MS3_00009365, MS3_00003207, MS3_00008121, MS3...</td>\n",
       "      <td>[gene-SINT2_LOCUS5976, gene-SINT2_LOCUS8677, g...</td>\n",
       "      <td>[gene-KSF78_0005439, gene-KSF78_0002617, gene-...</td>\n",
       "      <td>[gene-Smp_099090, gene-Smp_166270, gene-Smp_17...</td>\n",
       "      <td>[gene-SMTH1_LOCUS3510, gene-SMTH1_LOCUS7315, g...</td>\n",
       "      <td>[gene-MN116_000208, gene-MN116_000460, gene-MN...</td>\n",
       "      <td>[gene-SSPN1_LOCUS6331, gene-SSPN1_LOCUS8419, g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OG0000003</td>\n",
       "      <td>[gene-SBOV1_LOCUS4395, gene-SBOV1_LOCUS1862, g...</td>\n",
       "      <td>[MS3_00001180, MS3_00000960, MS3_00000190, MS3...</td>\n",
       "      <td>[gene-SINT2_LOCUS2648, gene-SINT2_LOCUS4418, g...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[gene-Smp_000120, gene-Smp_146890]</td>\n",
       "      <td>[gene-SMTH1_LOCUS6453, gene-SMTH1_LOCUS9462, g...</td>\n",
       "      <td>[gene-MN116_000176, gene-MN116_000154, gene-MN...</td>\n",
       "      <td>[gene-SSPN1_LOCUS1113, gene-SSPN1_LOCUS5190, g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OG0000004</td>\n",
       "      <td>[gene-SBOV1_LOCUS708, gene-SBOV1_LOCUS1939, ge...</td>\n",
       "      <td>[MS3_00003078, MS3_00000542, MS3_00002760, MS3...</td>\n",
       "      <td>[gene-SINT2_LOCUS3231, gene-SINT2_LOCUS6008, g...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[gene-Smp_191250, gene-Smp_112270, gene-Smp_19...</td>\n",
       "      <td>[gene-SMTH1_LOCUS10409, gene-SMTH1_LOCUS2112, ...</td>\n",
       "      <td>[gene-MN116_000398, gene-MN116_000067, gene-MN...</td>\n",
       "      <td>[gene-SSPN1_LOCUS9325, gene-SSPN1_LOCUS1080, g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9519</th>\n",
       "      <td>OG0009519</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[gene-SSPN1_LOCUS4940]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9520</th>\n",
       "      <td>OG0009520</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[gene-SSPN1_LOCUS6955]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9521</th>\n",
       "      <td>OG0009521</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[gene-SSPN1_LOCUS7761, gene-SSPN1_LOCUS7762]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9522</th>\n",
       "      <td>OG0009522</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[gene-SSPN1_LOCUS8430]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9523</th>\n",
       "      <td>OG0009523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[gene-SSPN1_LOCUS9033, gene-SSPN1_LOCUS9030]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9524 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Orthogroup                                              bovis  \\\n",
       "0     OG0000000  [gene-SBOV1_LOCUS2675, gene-SBOV1_LOCUS4335, g...   \n",
       "1     OG0000001  [gene-SBOV1_LOCUS796, gene-SBOV1_LOCUS8840, ge...   \n",
       "2     OG0000002  [gene-SBOV1_LOCUS6926, gene-SBOV1_LOCUS2070, g...   \n",
       "3     OG0000003  [gene-SBOV1_LOCUS4395, gene-SBOV1_LOCUS1862, g...   \n",
       "4     OG0000004  [gene-SBOV1_LOCUS708, gene-SBOV1_LOCUS1939, ge...   \n",
       "...         ...                                                ...   \n",
       "9519  OG0009519                                                NaN   \n",
       "9520  OG0009520                                                NaN   \n",
       "9521  OG0009521                                                NaN   \n",
       "9522  OG0009522                                                NaN   \n",
       "9523  OG0009523                                                NaN   \n",
       "\n",
       "                                            haematobium  \\\n",
       "0     [MS3_00001738, MS3_00005418, MS3_00000214, MS3...   \n",
       "1     [MS3_00000895, MS3_00001105, MS3_00003570, MS3...   \n",
       "2     [MS3_00009365, MS3_00003207, MS3_00008121, MS3...   \n",
       "3     [MS3_00001180, MS3_00000960, MS3_00000190, MS3...   \n",
       "4     [MS3_00003078, MS3_00000542, MS3_00002760, MS3...   \n",
       "...                                                 ...   \n",
       "9519                                                NaN   \n",
       "9520                                                NaN   \n",
       "9521                                                NaN   \n",
       "9522                                                NaN   \n",
       "9523                                                NaN   \n",
       "\n",
       "                                           intercalatum  \\\n",
       "0     [gene-SINT2_LOCUS1577, gene-SINT2_LOCUS2576, g...   \n",
       "1     [gene-SINT2_LOCUS1120, gene-SINT2_LOCUS1638, g...   \n",
       "2     [gene-SINT2_LOCUS5976, gene-SINT2_LOCUS8677, g...   \n",
       "3     [gene-SINT2_LOCUS2648, gene-SINT2_LOCUS4418, g...   \n",
       "4     [gene-SINT2_LOCUS3231, gene-SINT2_LOCUS6008, g...   \n",
       "...                                                 ...   \n",
       "9519                                                NaN   \n",
       "9520                                                NaN   \n",
       "9521                                                NaN   \n",
       "9522                                                NaN   \n",
       "9523                                                NaN   \n",
       "\n",
       "                                              japonicum  \\\n",
       "0     [gene-KSF78_0004650, gene-KSF78_0001145, gene-...   \n",
       "1                                  [gene-KSF78_0006907]   \n",
       "2     [gene-KSF78_0005439, gene-KSF78_0002617, gene-...   \n",
       "3                                                   NaN   \n",
       "4                                                   NaN   \n",
       "...                                                 ...   \n",
       "9519                                                NaN   \n",
       "9520                                                NaN   \n",
       "9521                                                NaN   \n",
       "9522                                                NaN   \n",
       "9523                                                NaN   \n",
       "\n",
       "                                                mansoni  \\\n",
       "0     [gene-Smp_025180, gene-Smp_166950, gene-Smp_19...   \n",
       "1     [gene-Smp_122020, gene-Smp_120330, gene-Smp_18...   \n",
       "2     [gene-Smp_099090, gene-Smp_166270, gene-Smp_17...   \n",
       "3                    [gene-Smp_000120, gene-Smp_146890]   \n",
       "4     [gene-Smp_191250, gene-Smp_112270, gene-Smp_19...   \n",
       "...                                                 ...   \n",
       "9519                                                NaN   \n",
       "9520                                                NaN   \n",
       "9521                                                NaN   \n",
       "9522                                                NaN   \n",
       "9523                                                NaN   \n",
       "\n",
       "                                               mattheei  \\\n",
       "0     [gene-SMTH1_LOCUS3830, gene-SMTH1_LOCUS3089, g...   \n",
       "1     [gene-SMTH1_LOCUS10701, gene-SMTH1_LOCUS5153, ...   \n",
       "2     [gene-SMTH1_LOCUS3510, gene-SMTH1_LOCUS7315, g...   \n",
       "3     [gene-SMTH1_LOCUS6453, gene-SMTH1_LOCUS9462, g...   \n",
       "4     [gene-SMTH1_LOCUS10409, gene-SMTH1_LOCUS2112, ...   \n",
       "...                                                 ...   \n",
       "9519                                                NaN   \n",
       "9520                                                NaN   \n",
       "9521                                                NaN   \n",
       "9522                                                NaN   \n",
       "9523                                                NaN   \n",
       "\n",
       "                                                mekongi  \\\n",
       "0     [gene-MN116_000469, gene-MN116_000157, gene-MN...   \n",
       "1     [gene-MN116_000217, gene-MN116_000421, gene-MN...   \n",
       "2     [gene-MN116_000208, gene-MN116_000460, gene-MN...   \n",
       "3     [gene-MN116_000176, gene-MN116_000154, gene-MN...   \n",
       "4     [gene-MN116_000398, gene-MN116_000067, gene-MN...   \n",
       "...                                                 ...   \n",
       "9519                                                NaN   \n",
       "9520                                                NaN   \n",
       "9521                                                NaN   \n",
       "9522                                                NaN   \n",
       "9523                                                NaN   \n",
       "\n",
       "                                              spindalis  \n",
       "0     [gene-SSPN1_LOCUS1690, gene-SSPN1_LOCUS8011, g...  \n",
       "1     [gene-SSPN1_LOCUS1012, gene-SSPN1_LOCUS8827, g...  \n",
       "2     [gene-SSPN1_LOCUS6331, gene-SSPN1_LOCUS8419, g...  \n",
       "3     [gene-SSPN1_LOCUS1113, gene-SSPN1_LOCUS5190, g...  \n",
       "4     [gene-SSPN1_LOCUS9325, gene-SSPN1_LOCUS1080, g...  \n",
       "...                                                 ...  \n",
       "9519                             [gene-SSPN1_LOCUS4940]  \n",
       "9520                             [gene-SSPN1_LOCUS6955]  \n",
       "9521       [gene-SSPN1_LOCUS7761, gene-SSPN1_LOCUS7762]  \n",
       "9522                             [gene-SSPN1_LOCUS8430]  \n",
       "9523       [gene-SSPN1_LOCUS9033, gene-SSPN1_LOCUS9030]  \n",
       "\n",
       "[9524 rows x 9 columns]"
      ]
     },
     "execution_count": 571,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# заменим протеины на соответствующие им гены\n",
    "\n",
    "import pathlib\n",
    "import numpy as np\n",
    "\n",
    "organism_protein_gene = dict()\n",
    "\n",
    "for organism, config in data_config.items():\n",
    "    organism_protein_gene[organism] = dict()\n",
    "    if organism in gff_files:\n",
    "        current_gff_file = gff_files[organism]\n",
    "        if pathlib.Path(current_gff_file).suffix == '.gff':\n",
    "            with open(current_gff_file, 'r') as gff:\n",
    "                last_gene = None\n",
    "                for row in gff:\n",
    "                    if row[0] == '#':\n",
    "                        continue\n",
    "                    row_splitted = row.split()\n",
    "                    if row_splitted[2] == 'gene':\n",
    "                        last_gene = row_splitted[8].split(';')[0]\n",
    "                        last_gene = last_gene.split('ID=')[1]\n",
    "                    elif 'protein_id' in row_splitted[-1]:\n",
    "                        current_protein = row_splitted[-1].split(';protein_id=')[-1].split(';')[0]\n",
    "                        organism_protein_gene[organism][current_protein] = last_gene\n",
    "\n",
    "def replace_protein_to_gene(organism, proteins):\n",
    "    if isinstance(proteins, str):\n",
    "        proteins_splitted = proteins.split(', ')\n",
    "        res = []\n",
    "        for protein in proteins_splitted:\n",
    "            if 'MS3_' in protein:\n",
    "                res.append(protein.split('.')[0])\n",
    "            else:\n",
    "                if protein in organism_protein_gene[organism]:\n",
    "                    res.append(organism_protein_gene[organism][protein])\n",
    "        res = list(set(res))\n",
    "        return res\n",
    "    else:\n",
    "        return np.nan\n",
    "\n",
    "\n",
    "orthogroups_replaced = orthogroups.copy()\n",
    "\n",
    "for organism in data_config:\n",
    "    orthogroups_replaced[organism] = orthogroups_replaced[organism].apply(lambda x : replace_protein_to_gene(organism, x))\n",
    "\n",
    "orthogroups_replaced"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef9885c4",
   "metadata": {},
   "source": [
    "### 2. От каждого организма в группе выбрать 4-5 (как минимум 1 на члена группы)  генов (ортологов) в промоторах которых есть  квадруплексы и для этих квадруплексов построить выравнивание. И указать где есть вариации (в петле, числе повторов, выпетливания и т.п.) Если таких генов меньше, то для всех генов, что нашлись."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 716,
   "id": "6f4802f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spindalis\n",
      "japonicum\n",
      "bovis\n",
      "mattheei\n",
      "haematobium\n",
      "mansoni\n",
      "intercalatum\n",
      "mekongi\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>bovis</th>\n",
       "      <th>haematobium</th>\n",
       "      <th>intercalatum</th>\n",
       "      <th>japonicum</th>\n",
       "      <th>mansoni</th>\n",
       "      <th>mattheei</th>\n",
       "      <th>mekongi</th>\n",
       "      <th>spindalis</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6675</th>\n",
       "      <td>OG0006675</td>\n",
       "      <td>[gene-SBOV1_LOCUS189]</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-SINT2_LOCUS158]</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-Smp_163240]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903</th>\n",
       "      <td>OG0000903</td>\n",
       "      <td>[gene-SBOV1_LOCUS9352]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-SMTH1_LOCUS3843]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-SSPN1_LOCUS3328]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>OG0001177</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-SMTH1_LOCUS8163]</td>\n",
       "      <td>[gene-MN116_008764]</td>\n",
       "      <td>[gene-SSPN1_LOCUS7051]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6382</th>\n",
       "      <td>OG0006382</td>\n",
       "      <td>[gene-SBOV1_LOCUS5984]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-SMTH1_LOCUS6746]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2555</th>\n",
       "      <td>OG0002555</td>\n",
       "      <td>[gene-SBOV1_LOCUS2303]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-SMTH1_LOCUS2608]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3218</th>\n",
       "      <td>OG0003218</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3219</th>\n",
       "      <td>OG0003219</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3220</th>\n",
       "      <td>OG0003220</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3221</th>\n",
       "      <td>OG0003221</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3211</th>\n",
       "      <td>OG0003211</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9524 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Orthogroup                   bovis haematobium           intercalatum  \\\n",
       "6675  OG0006675   [gene-SBOV1_LOCUS189]        None  [gene-SINT2_LOCUS158]   \n",
       "903   OG0000903  [gene-SBOV1_LOCUS9352]        None                     []   \n",
       "1177  OG0001177                      []        None                     []   \n",
       "6382  OG0006382  [gene-SBOV1_LOCUS5984]        None                     []   \n",
       "2555  OG0002555  [gene-SBOV1_LOCUS2303]        None                     []   \n",
       "...         ...                     ...         ...                    ...   \n",
       "3218  OG0003218                      []        None                     []   \n",
       "3219  OG0003219                      []        None                     []   \n",
       "3220  OG0003220                      []        None                     []   \n",
       "3221  OG0003221                      []        None                     []   \n",
       "3211  OG0003211                      []        None                     []   \n",
       "\n",
       "     japonicum            mansoni                mattheei  \\\n",
       "6675      None  [gene-Smp_163240]                      []   \n",
       "903       None                 []  [gene-SMTH1_LOCUS3843]   \n",
       "1177      None                 []  [gene-SMTH1_LOCUS8163]   \n",
       "6382      None                 []  [gene-SMTH1_LOCUS6746]   \n",
       "2555      None                 []  [gene-SMTH1_LOCUS2608]   \n",
       "...        ...                ...                     ...   \n",
       "3218      None                 []                      []   \n",
       "3219      None                 []                      []   \n",
       "3220      None                 []                      []   \n",
       "3221      None                 []                      []   \n",
       "3211      None                 []                      []   \n",
       "\n",
       "                  mekongi               spindalis  count  \n",
       "6675                   []                      []      3  \n",
       "903                    []  [gene-SSPN1_LOCUS3328]      3  \n",
       "1177  [gene-MN116_008764]  [gene-SSPN1_LOCUS7051]      3  \n",
       "6382                   []                      []      2  \n",
       "2555                   []                      []      2  \n",
       "...                   ...                     ...    ...  \n",
       "3218                   []                      []      0  \n",
       "3219                   []                      []      0  \n",
       "3220                   []                      []      0  \n",
       "3221                   []                      []      0  \n",
       "3211                   []                      []      0  \n",
       "\n",
       "[9524 rows x 10 columns]"
      ]
     },
     "execution_count": 716,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "orthogroups_replaced_g4 = orthogroups_replaced.copy()\n",
    "\n",
    "# genes_with_g4_by_organism = dict()\n",
    "\n",
    "for organism in data_config:\n",
    "    res = None\n",
    "    print(organism)\n",
    "    if organism in genes_with_g4_by_organism:\n",
    "        res = genes_with_g4_by_organism[organism]\n",
    "    else:\n",
    "        res = get_genes_with_promoters_contains(organism, 'g4')\n",
    "        genes_with_g4_by_organism[organism] = res\n",
    "\n",
    "    if res is None:\n",
    "        orthogroups_replaced_g4[organism] = None\n",
    "        continue\n",
    "\n",
    "    cur_set = set(genes_with_g4_by_organism[organism]['gene'])\n",
    "\n",
    "    def filter_by_g4(genes):\n",
    "        if not isinstance(genes, float):\n",
    "            ans = []\n",
    "            for gene in genes:\n",
    "                if gene in cur_set:\n",
    "                    ans.append(gene)\n",
    "            return ans\n",
    "        return np.nan\n",
    "        \n",
    "    orthogroups_replaced_g4[organism] = orthogroups_replaced_g4[organism].apply(filter_by_g4)\n",
    "\n",
    "def contains_gene_count(row):\n",
    "    ans = 0\n",
    "    for key, value in row.items():\n",
    "        if isinstance(value, list):\n",
    "            if len(value) > 0:\n",
    "                ans += 1\n",
    "    return ans\n",
    "\n",
    "orthogroups_replaced_g4['count'] = orthogroups_replaced_g4.apply(contains_gene_count, axis=1)\n",
    "orthogroups_replaced_g4.sort_values(by='count', inplace=True, ascending=False)\n",
    "orthogroups_replaced_g4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 731,
   "id": "1931a66e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>bovis</th>\n",
       "      <th>haematobium</th>\n",
       "      <th>intercalatum</th>\n",
       "      <th>japonicum</th>\n",
       "      <th>mansoni</th>\n",
       "      <th>mattheei</th>\n",
       "      <th>mekongi</th>\n",
       "      <th>spindalis</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6675</th>\n",
       "      <td>OG0006675</td>\n",
       "      <td>[gene-SBOV1_LOCUS189]</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-SINT2_LOCUS158]</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-Smp_163240]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903</th>\n",
       "      <td>OG0000903</td>\n",
       "      <td>[gene-SBOV1_LOCUS9352]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-SMTH1_LOCUS3843]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-SSPN1_LOCUS3328]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>OG0001177</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-SMTH1_LOCUS8163]</td>\n",
       "      <td>[gene-MN116_008764]</td>\n",
       "      <td>[gene-SSPN1_LOCUS7051]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6382</th>\n",
       "      <td>OG0006382</td>\n",
       "      <td>[gene-SBOV1_LOCUS5984]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-SMTH1_LOCUS6746]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2555</th>\n",
       "      <td>OG0002555</td>\n",
       "      <td>[gene-SBOV1_LOCUS2303]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-SMTH1_LOCUS2608]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3164</th>\n",
       "      <td>OG0003164</td>\n",
       "      <td>[gene-SBOV1_LOCUS6033]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-MN116_008196]</td>\n",
       "      <td>[]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7370</th>\n",
       "      <td>OG0007370</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-SINT2_LOCUS61]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-SMTH1_LOCUS84]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5554</th>\n",
       "      <td>OG0005554</td>\n",
       "      <td>[gene-SBOV1_LOCUS4577]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene-SMTH1_LOCUS5171]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Orthogroup                   bovis haematobium           intercalatum  \\\n",
       "6675  OG0006675   [gene-SBOV1_LOCUS189]        None  [gene-SINT2_LOCUS158]   \n",
       "903   OG0000903  [gene-SBOV1_LOCUS9352]        None                     []   \n",
       "1177  OG0001177                      []        None                     []   \n",
       "6382  OG0006382  [gene-SBOV1_LOCUS5984]        None                     []   \n",
       "2555  OG0002555  [gene-SBOV1_LOCUS2303]        None                     []   \n",
       "3164  OG0003164  [gene-SBOV1_LOCUS6033]        None                     []   \n",
       "7370  OG0007370                      []        None   [gene-SINT2_LOCUS61]   \n",
       "5554  OG0005554  [gene-SBOV1_LOCUS4577]        None                     []   \n",
       "\n",
       "     japonicum            mansoni                mattheei  \\\n",
       "6675      None  [gene-Smp_163240]                      []   \n",
       "903       None                 []  [gene-SMTH1_LOCUS3843]   \n",
       "1177      None                 []  [gene-SMTH1_LOCUS8163]   \n",
       "6382      None                 []  [gene-SMTH1_LOCUS6746]   \n",
       "2555      None                 []  [gene-SMTH1_LOCUS2608]   \n",
       "3164      None                 []                      []   \n",
       "7370      None                 []    [gene-SMTH1_LOCUS84]   \n",
       "5554      None                 []  [gene-SMTH1_LOCUS5171]   \n",
       "\n",
       "                  mekongi               spindalis  count  \n",
       "6675                   []                      []      3  \n",
       "903                    []  [gene-SSPN1_LOCUS3328]      3  \n",
       "1177  [gene-MN116_008764]  [gene-SSPN1_LOCUS7051]      3  \n",
       "6382                   []                      []      2  \n",
       "2555                   []                      []      2  \n",
       "3164  [gene-MN116_008196]                      []      2  \n",
       "7370                   []                      []      2  \n",
       "5554                   []                      []      2  "
      ]
     },
     "execution_count": 731,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orthogroups_replaced_g4_selected = orthogroups_replaced_g4[:8]\n",
    "orthogroups_replaced_g4_selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 772,
   "id": "e7fa87cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "for_g4_alignment = []\n",
    "\n",
    "for _, row in orthogroups_replaced_g4_selected.iterrows():\n",
    "    table = []\n",
    "    for organism in data_config:\n",
    "        genes = row[organism]\n",
    "        if isinstance(genes, list):\n",
    "            for gene in genes:\n",
    "                table.append((organism, gene))\n",
    "\n",
    "    for_g4_alignment.append(pd.DataFrame(table))\n",
    "    for_g4_alignment[-1].columns = ['organism', 'gene']\n",
    "\n",
    "g4_by_gene = dict()\n",
    "\n",
    "for rec in for_g4_alignment:\n",
    "    for _, row in rec.iterrows():\n",
    "\n",
    "        search = genes_with_g4_by_organism[row['organism']]\n",
    "        for _, s_row in search.iterrows():\n",
    "            if s_row['gene'] == row['gene']:\n",
    "                g4_by_gene[s_row['gene']] = (s_row['record_id'], s_row['start'], s_row['end'])\n",
    "\n",
    "for i in range(len(for_g4_alignment)):\n",
    "    for_g4_alignment[i]['record_id'] = for_g4_alignment[i]['gene'].apply(lambda x : g4_by_gene[x][0])\n",
    "\n",
    "for i in range(len(for_g4_alignment)):\n",
    "    for_g4_alignment[i]['start'] = for_g4_alignment[i]['gene'].apply(lambda x : g4_by_gene[x][1])\n",
    "\n",
    "for i in range(len(for_g4_alignment)):\n",
    "    for_g4_alignment[i]['end'] = for_g4_alignment[i]['gene'].apply(lambda x : g4_by_gene[x][2])\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "def get_data(organism, record, start, end):\n",
    "    fna_file = fna_files[organism]\n",
    "    for seq_record in SeqIO.parse(fna_file, \"fasta\"):\n",
    "        if seq_record.id == record:\n",
    "            # Извлекаем последовательность (переводим в 0-based для среза)\n",
    "            sequence = str(seq_record.seq[start-1:end]).upper()\n",
    "            return sequence\n",
    "    assert False\n",
    "\n",
    "for i in range(len(for_g4_alignment)):\n",
    "    for_g4_alignment[i]['data'] = for_g4_alignment[i].apply(lambda row: get_data(row['organism'], row['record_id'], row['start'], row['end']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 785,
   "id": "029555c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def has_g4_sequence(dna_sequence):\n",
    "    pattern = r\"G{3,}[ATGC]{1,7}G{3,}[ATGC]{1,7}G{3,}[ATGC]{1,7}G{3,}\"\n",
    "    if bool(re.search(pattern, dna_sequence)):\n",
    "        return 1\n",
    "    complement = dna_sequence.translate(str.maketrans(\"ATGC\", \"TACG\"))\n",
    "    reverse_complement = complement[::-1]\n",
    "\n",
    "    if re.search(pattern, reverse_complement):\n",
    "        return -1\n",
    "\n",
    "    return 0\n",
    "\n",
    "def translate(dna_sequence):\n",
    "    if has_g4_sequence(dna_sequence) == -1:\n",
    "        complement = dna_sequence.translate(str.maketrans(\"ATGC\", \"TACG\"))\n",
    "        reverse_complement = complement[::-1]\n",
    "        return reverse_complement\n",
    "    return dna_sequence\n",
    "\n",
    "for i in range(len(for_g4_alignment)):\n",
    "    for_g4_alignment[i]['data'] = for_g4_alignment[i]['data'].apply(translate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 786,
   "id": "212710e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>record_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bovis</td>\n",
       "      <td>gene-SBOV1_LOCUS189</td>\n",
       "      <td>OX103953.1</td>\n",
       "      <td>8284772</td>\n",
       "      <td>8285771</td>\n",
       "      <td>TACAATATTTAGACATAAGTATTATGCAATATAAATGTATTAAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mansoni</td>\n",
       "      <td>gene-Smp_163240</td>\n",
       "      <td>NC_031495.1</td>\n",
       "      <td>4337210</td>\n",
       "      <td>4338209</td>\n",
       "      <td>GTTTGTTTGTTTTAGTCATCATCATTTGCCTCATGGATACAATTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>intercalatum</td>\n",
       "      <td>gene-SINT2_LOCUS158</td>\n",
       "      <td>OX103876.1</td>\n",
       "      <td>6341243</td>\n",
       "      <td>6342242</td>\n",
       "      <td>ATATAGAGAGAGAGGGAAAAAAAGCGTATGAAACAAATGGTTAGTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       organism                 gene    record_id    start      end  \\\n",
       "0         bovis  gene-SBOV1_LOCUS189   OX103953.1  8284772  8285771   \n",
       "1       mansoni      gene-Smp_163240  NC_031495.1  4337210  4338209   \n",
       "2  intercalatum  gene-SINT2_LOCUS158   OX103876.1  6341243  6342242   \n",
       "\n",
       "                                                data  \n",
       "0  TACAATATTTAGACATAAGTATTATGCAATATAAATGTATTAAAAT...  \n",
       "1  GTTTGTTTGTTTTAGTCATCATCATTTGCCTCATGGATACAATTTA...  \n",
       "2  ATATAGAGAGAGAGGGAAAAAAAGCGTATGAAACAAATGGTTAGTT...  "
      ]
     },
     "execution_count": 786,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for_g4_alignment[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 817,
   "id": "4e967839",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Running allignment for data/g4_genes_alignment_0.meg\n",
      "INFO:root:Running allignment for data/g4_genes_alignment_1.meg\n",
      "INFO:root:Running allignment for data/g4_genes_alignment_2.meg\n",
      "INFO:root:Running allignment for data/g4_genes_alignment_3.meg\n",
      "INFO:root:Running allignment for data/g4_genes_alignment_4.meg\n",
      "INFO:root:Running allignment for data/g4_genes_alignment_5.meg\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allignment for data/g4_genes_for_alignment_5.fasta done\n",
      "INFO:root:Running allignment for data/g4_genes_alignment_6.meg\n",
      "INFO:root:Allignment for data/g4_genes_for_alignment_3.fasta done\n",
      "INFO:root:Running allignment for data/g4_genes_alignment_7.meg\n",
      "INFO:root:Allignment for data/g4_genes_for_alignment_4.fasta done\n",
      "INFO:root:Allignment for data/g4_genes_for_alignment_2.fasta done\n",
      "INFO:root:Allignment for data/g4_genes_for_alignment_1.fasta done\n",
      "INFO:root:Allignment for data/g4_genes_for_alignment_6.fasta done\n",
      "INFO:root:Allignment for data/g4_genes_for_alignment_7.fasta done\n",
      "INFO:root:Allignment for data/g4_genes_for_alignment_0.fasta done\n"
     ]
    }
   ],
   "source": [
    "g4_alignment_results = []\n",
    "\n",
    "def run_alignment(file):\n",
    "    logging.info(f'Running allignment for {file[1]}')\n",
    "    os.system(f'rm -f {file[1]}')\n",
    "    subprocess.run([f'./local_programs/mega/megacc', '-a', 'data/muscle_align.mao', '-d', file[0], '-o', file[1]],     stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)\n",
    "\n",
    "    logging.info(f'Allignment for {file[0]} done')\n",
    "\n",
    "files = []\n",
    "\n",
    "for i in range(len(for_g4_alignment)):\n",
    "    current_fasta_file = f'data/g4_genes_for_alignment_{i}.fasta'\n",
    "    current_meg_file = f'data/g4_genes_alignment_{i}.meg'\n",
    "    files.append((current_fasta_file, current_meg_file))\n",
    "    with open(current_fasta_file, 'w') as fasta_file:\n",
    "        for _, row in for_g4_alignment[i].iterrows():\n",
    "            seq_id, seq = f'{row['organism']}:{row['gene']}', row['data']\n",
    "            fasta_file.write(f\">{seq_id}\\n{seq}\\n\")\n",
    "    g4_alignment_results.append((i, current_meg_file))\n",
    "\n",
    "\n",
    "with concurrent.futures.ThreadPoolExecutor(max_workers=6) as executor:\n",
    "    executor.map(run_alignment, files)\n",
    "    executor.shutdown(wait=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f59a6abc",
   "metadata": {},
   "source": [
    "### 3. От каждого организма в группе выбрать 4-5 (как минимум 1 на члена группы)  генов (ортологов) в промоторах которых есть  Z-ДНК и для этих ДНК построить выравнивание для промоторов. И указать где есть вариации (в петле, числе повторов, выпетливания и т.п.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 839,
   "id": "ee2858ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spindalis\n",
      "japonicum\n",
      "bovis\n",
      "mattheei\n",
      "haematobium\n",
      "mansoni\n",
      "intercalatum\n",
      "mekongi\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>bovis</th>\n",
       "      <th>haematobium</th>\n",
       "      <th>intercalatum</th>\n",
       "      <th>japonicum</th>\n",
       "      <th>mansoni</th>\n",
       "      <th>mattheei</th>\n",
       "      <th>mekongi</th>\n",
       "      <th>spindalis</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>OG0000007</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-SINT2_LOCUS9448]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-MN116_000127]</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1688</th>\n",
       "      <td>OG0001688</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-MN116_008339]</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5796</th>\n",
       "      <td>OG0005796</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-MN116_001274]</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4561</th>\n",
       "      <td>OG0004561</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-SINT2_LOCUS88]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1905</th>\n",
       "      <td>OG0001905</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-SINT2_LOCUS3494]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3218</th>\n",
       "      <td>OG0003218</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3219</th>\n",
       "      <td>OG0003219</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3220</th>\n",
       "      <td>OG0003220</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3221</th>\n",
       "      <td>OG0003221</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3211</th>\n",
       "      <td>OG0003211</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9524 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Orthogroup bovis haematobium            intercalatum japonicum mansoni  \\\n",
       "7     OG0000007  None        None  [gene-SINT2_LOCUS9448]      None    None   \n",
       "1688  OG0001688  None        None                      []      None    None   \n",
       "5796  OG0005796  None        None                      []      None    None   \n",
       "4561  OG0004561  None        None    [gene-SINT2_LOCUS88]      None    None   \n",
       "1905  OG0001905  None        None  [gene-SINT2_LOCUS3494]      None    None   \n",
       "...         ...   ...         ...                     ...       ...     ...   \n",
       "3218  OG0003218  None        None                      []      None    None   \n",
       "3219  OG0003219  None        None                      []      None    None   \n",
       "3220  OG0003220  None        None                      []      None    None   \n",
       "3221  OG0003221  None        None                      []      None    None   \n",
       "3211  OG0003211  None        None                      []      None    None   \n",
       "\n",
       "     mattheei              mekongi spindalis  count  \n",
       "7        None  [gene-MN116_000127]      None      2  \n",
       "1688     None  [gene-MN116_008339]      None      1  \n",
       "5796     None  [gene-MN116_001274]      None      1  \n",
       "4561     None                   []      None      1  \n",
       "1905     None                   []      None      1  \n",
       "...       ...                  ...       ...    ...  \n",
       "3218     None                   []      None      0  \n",
       "3219     None                   []      None      0  \n",
       "3220     None                   []      None      0  \n",
       "3221     None                   []      None      0  \n",
       "3211     None                   []      None      0  \n",
       "\n",
       "[9524 rows x 10 columns]"
      ]
     },
     "execution_count": 839,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "orthogroups_replaced_zdna = orthogroups_replaced.copy()\n",
    "\n",
    "# genes_with_zdna_by_organism = dict()\n",
    "\n",
    "for organism in data_config:\n",
    "    res = None\n",
    "    print(organism)\n",
    "    if organism in genes_with_zdna_by_organism:\n",
    "        res = genes_with_zdna_by_organism[organism]\n",
    "    else:\n",
    "        res = get_genes_with_promoters_contains(organism, 'zdna')\n",
    "        genes_with_zdna_by_organism[organism] = res\n",
    "\n",
    "    if res is None:\n",
    "        orthogroups_replaced_zdna[organism] = None\n",
    "        continue\n",
    "\n",
    "    cur_set = set(genes_with_zdna_by_organism[organism]['gene'])\n",
    "\n",
    "    def filter_by_zdna(genes):\n",
    "        if not isinstance(genes, float):\n",
    "            ans = []\n",
    "            for gene in genes:\n",
    "                if gene in cur_set:\n",
    "                    ans.append(gene)\n",
    "            return ans\n",
    "        return np.nan\n",
    "        \n",
    "    orthogroups_replaced_zdna[organism] = orthogroups_replaced_zdna[organism].apply(filter_by_zdna)\n",
    "\n",
    "def contains_gene_count(row):\n",
    "    ans = 0\n",
    "    for key, value in row.items():\n",
    "        if isinstance(value, list):\n",
    "            if len(value) > 0:\n",
    "                ans += 1\n",
    "    return ans\n",
    "\n",
    "orthogroups_replaced_zdna['count'] = orthogroups_replaced_zdna.apply(contains_gene_count, axis=1)\n",
    "orthogroups_replaced_zdna.sort_values(by='count', inplace=True, ascending=False)\n",
    "orthogroups_replaced_zdna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 840,
   "id": "3f3ed193",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>bovis</th>\n",
       "      <th>haematobium</th>\n",
       "      <th>intercalatum</th>\n",
       "      <th>japonicum</th>\n",
       "      <th>mansoni</th>\n",
       "      <th>mattheei</th>\n",
       "      <th>mekongi</th>\n",
       "      <th>spindalis</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>OG0000007</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-SINT2_LOCUS9448]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[gene-MN116_000127]</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Orthogroup bovis haematobium            intercalatum japonicum mansoni  \\\n",
       "7  OG0000007  None        None  [gene-SINT2_LOCUS9448]      None    None   \n",
       "\n",
       "  mattheei              mekongi spindalis  count  \n",
       "7     None  [gene-MN116_000127]      None      2  "
      ]
     },
     "execution_count": 840,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orthogroups_replaced_zdna_selected = orthogroups_replaced_zdna[:1]\n",
    "orthogroups_replaced_zdna_selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36e36d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "for_zdna_alignment = []\n",
    "\n",
    "for _, row in orthogroups_replaced_zdna_selected.iterrows():\n",
    "    table = []\n",
    "    for organism in data_config:\n",
    "        genes = row[organism]\n",
    "        if isinstance(genes, list):\n",
    "            for gene in genes:\n",
    "                table.append((organism, gene))\n",
    "\n",
    "    for_zdna_alignment.append(pd.DataFrame(table))\n",
    "    for_zdna_alignment[-1].columns = ['organism', 'gene']\n",
    "\n",
    "zdna_by_gene = dict()\n",
    "\n",
    "for rec in for_zdna_alignment:\n",
    "    for _, row in rec.iterrows():\n",
    "\n",
    "        search = genes_with_zdna_by_organism[row['organism']]\n",
    "        for _, s_row in search.iterrows():\n",
    "            if s_row['gene'] == row['gene']:\n",
    "                zdna_by_gene[s_row['gene']] = (s_row['record_id'], s_row['start'], s_row['end'])\n",
    "\n",
    "for i in range(len(for_zdna_alignment)):\n",
    "    for_zdna_alignment[i]['record_id'] = for_zdna_alignment[i]['gene'].apply(lambda x : zdna_by_gene[x][0])\n",
    "\n",
    "for i in range(len(for_zdna_alignment)):\n",
    "    for_zdna_alignment[i]['start'] = for_zdna_alignment[i]['gene'].apply(lambda x : zdna_by_gene[x][1])\n",
    "\n",
    "for i in range(len(for_zdna_alignment)):\n",
    "    for_zdna_alignment[i]['end'] = for_zdna_alignment[i]['gene'].apply(lambda x : zdna_by_gene[x][2])\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "def get_data(organism, record, start, end):\n",
    "    fna_file = fna_files[organism]\n",
    "    for seq_record in SeqIO.parse(fna_file, \"fasta\"):\n",
    "        if seq_record.id == record:\n",
    "            # Извлекаем последовательность (переводим в 0-based для среза)\n",
    "            sequence = str(seq_record.seq[start-1:end]).upper()\n",
    "            return sequence\n",
    "    assert False\n",
    "\n",
    "for i in range(len(for_zdna_alignment)):\n",
    "    for_zdna_alignment[i]['data'] = for_zdna_alignment[i].apply(lambda row: get_data(row['organism'], row['record_id'], row['start'], row['end']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 844,
   "id": "229ac45f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Running allignment for data/zdna_genes_alignment_0.meg\n",
      "INFO:root:Allignment for data/zdna_genes_for_alignment_0.fasta done\n"
     ]
    }
   ],
   "source": [
    "zdna_alignment_results = []\n",
    "\n",
    "def run_alignment(file):\n",
    "    logging.info(f'Running allignment for {file[1]}')\n",
    "    os.system(f'rm -f {file[1]}')\n",
    "    subprocess.run([f'./local_programs/mega/megacc', '-a', 'data/muscle_align.mao', '-d', file[0], '-o', file[1]],     stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)\n",
    "\n",
    "    logging.info(f'Allignment for {file[0]} done')\n",
    "\n",
    "files = []\n",
    "\n",
    "for i in range(len(for_zdna_alignment)):\n",
    "    current_fasta_file = f'data/zdna_genes_for_alignment_{i}.fasta'\n",
    "    current_meg_file = f'data/zdna_genes_alignment_{i}.meg'\n",
    "    files.append((current_fasta_file, current_meg_file))\n",
    "    with open(current_fasta_file, 'w') as fasta_file:\n",
    "        for _, row in for_g4_alignment[i].iterrows():\n",
    "            seq_id, seq = f'{row['organism']}:{row['gene']}', row['data']\n",
    "            fasta_file.write(f\">{seq_id}\\n{seq}\\n\")\n",
    "    g4_alignment_results.append((i, current_meg_file))\n",
    "\n",
    "\n",
    "with concurrent.futures.ThreadPoolExecutor(max_workers=6) as executor:\n",
    "    executor.map(run_alignment, files)\n",
    "    executor.shutdown(wait=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
